
        {
            "cells": [
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["# Plant Systems Biology\n",
"\n",
"1. [Plant Systems Biology](#Plant-Systems-Biology)\n",
"2. [Modeling growth for a defined cell number](#Modeling-growth-for-a-defined-cell-number)\n",
"1. [Insert Growth Data and Display as Chart](#Insert-Growth-Data-and-Display-as-Chart)\n",
"2. [Calculation of growth rate and doubling time for cell cultures](#Calculation-of-growth-rate-and-doubling-time-for-cell-cultures)\n",
"3. [Fitting biological growth curves](#Fitting-biological-growth-curves)\u003cbr\u003e\n",
"    1. [Theory](#Theory)\u003cbr\u003e\n",
"    2. [Model selection](#Model-selection)\u003cbr\u003e\n",
"    3. [Exponential Fit](#Exponential-Fit)\u003cbr\u003e\n",
"4. [Calculate Doubling Time](#Calculate-Doubling-Time)\n",
"5. [References](#References)\n",
"\n",
"## Plant Systems Biology\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"The general paradigm of Systems Biology clearly applies to plants, as they represent complex biological systems. \n",
"The functioning of a plant as a biological system is the result of a combination of multiple intertwined and dynamic interactions between its components. \n",
"In addition, most plants are sessile systems that have to face fluctuating environmental conditions, including biotic and abiotic stresses\u003csup\u003e\u003ca href=\"#8\"\u003e8\u003c/a\u003e\u003c/sup\u003e.\n",
"The process of a biological system responding to changes in environmental conditions is termed acclimation. These molecular physiological responses represent a complex \n",
"dynamic adjustment of the interplay between genes, proteins and metabolites that allows the organism to acclimate to the changing environment. \n",
"The ability to acclimate ensures the survival of all living organisms and is therefore fundamental for the understanding of biological systems. \n",
"Detailed knowledge about how plants acclimate to a changing environment is crucial especially in times of global climate changes, \n",
"as plants are of great importance for our quality of life as a key source of food, shelter, fiber, medicine, and fuel\u003csup\u003e\u003ca href=\"#9\"\u003e9\u003c/a\u003e\u003c/sup\u003e.\n",
"\n",
"The prominent model plant \u003ci\u003eArabidopsis thaliana\u003c/i\u003e is well suited for plant Systems Biology studies because sophisticated experimental tools and extensive data \n",
"collections are readily available\u003csup\u003e\u003ca href=\"#10\"\u003e10\u003c/a\u003e\u003c/sup\u003e. However, the importance of a model organism is not only coined by the availability of molecular \n",
"tools to manipulate the organism, but also by its agricultural and economic impact like in the cases of tobacco, rice, maize or \n",
"barley\u003csup\u003e\u003ca href=\"#11\"\u003e11\u003c/a\u003e\u003c/sup\u003e. Also microalgae are of special economic interest due to their potential as biofuel producers\u003csup\u003e\u003ca href=\"#12\"\u003e12\u003c/a\u003e\u003c/sup\u003e. \n",
"Additionally, the use of organisms with lower biological complexity facilitates the feasibility of System Biology studies and is an important factor to consider \n",
"for the choice of a suitable model organism in Systems Biology.\n",
"\n",
"The eukaryotic green alga \u003ci\u003eChlamydomonas reinhardtii\u003c/i\u003e is particularly well suited for plant Systems Biology approaches. \n",
"This unicellular freshwater and soil-dwelling alga has a single, cup-shaped chloroplast with a photosynthetic apparatus that is similar to \n",
"that of higher plants\u003csup\u003e\u003ca href=\"#13\"\u003e13\u003c/a\u003e,\u003ca href=\"#14\"\u003e14\u003c/a\u003e\u003c/sup\u003e. Hence, results gained on photosynthesis processes in \u003ci\u003eChlamydomonas\u003c/i\u003e \n",
"are likely to be transferable to higher plants. The nuclear, mitochondrial, and chloroplast genomes have been sequenced and tools for manipulating them \n",
"are available\u003csup\u003e\u003ca href=\"#14\"\u003e14\u003c/a\u003e\u003c/sup\u003e. \u003ci\u003eChlamydomonas\u003c/i\u003e cells have a size of ~10 ¬µm and grow under photo-, mixo-, and heterotrophic conditions \n",
"with a generation time of ~5-8 h\u003csup\u003e\u003ca href=\"#15\"\u003e15\u003c/a\u003e\u003c/sup\u003e. \u003ci\u003eChlamydomonas\u003c/i\u003e can be maintained under controlled conditions and environmental \n",
"changes can be applied homogeneously and rapidly to all cells in a liquid culture. In contrast to multicellular organisms there are no influences by \n",
"tissue heterogeneity. Even the influence of different cell cycle stages may be ruled out by performing experiments with asynchronous cell cultures \n",
"\u003csup\u003e\u003ca href=\"#16\"\u003e16\u003c/a\u003e,\u003ca href=\"#17\"\u003e17\u003c/a\u003e\u003c/sup\u003e. Finally, gene families in Chlamydomonas have fewer members than those in higher plants thus facilitating the \n",
"interpretation of results involving many genes/proteins\u003csup\u003e\u003ca href=\"#14\"\u003e14\u003c/a\u003e\u003c/sup\u003e.\n",
"\u003c/div\u003e\n",
"\n",
"## Modeling growth for a defined cell number\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"In order to solves real world task more convenient, F# provides a huge collection of additional programming libraries. \n",
"Anything that extends beyond the basics must be written by a user. If the chunk of code is useful to multiple different users, \n",
"it\u0027s often put into a library to make it easily reusable. A library is a collection of related pieces of code that have been compiled \n",
"and stored together in a single file and can than be used an included. The most important libraries in F# for bioinformatics are:\n",
"\n",
"\u003cul\u003e\n",
"\u003cul\u003e\n",
"    \u003cli\u003e\u003ca href=\"https://csbiology.github.io/BioFSharp/\"\u003eBioFSharp\u003c/a\u003e: Open source bioinformatics and computational biology toolbox written in F#\u003c/li\u003e\n",
"    \u003cli\u003e\u003ca href=\"https://csbiology.github.io/FSharp.Stats/\"\u003eFSharp.Stats\u003c/a\u003e: F# project for statistical computing \u003c/li\u003e\n",
"    \u003cli\u003e\u003ca href=\"https://github.com/plotly/Plotly.NET\"\u003ePlotly.NET\u003c/a\u003e: .NET interface for plotly.js written in F# üìà \u003c/li\u003e\n",
"\u003c/ul\u003e\n",
"\n",
"The first real world use case of F# in Systems Biology is to model growth for a defined cell number to see possible overexpression effects. \n",
"Biologists often utilize growth experiments to analyze basic properties of a given organism or cellular model. For a solid comparison of data \n",
"obtain from different experiment and to investigate the speciÔ¨Åc eÔ¨Äect of a given experimental set up, modeling the growth is needed after recording the data. \n",
"\n",
"This notebook introduces two basic ways to model growth of \u003ci\u003eChlamydomonas reinhardtii\u003c/i\u003e using F#.\n",
"\n",
"Now, let\u0027s get started by loading our libraries first.\n",
"\u003c/div\u003e\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 1, "outputs": [], 
           "source": ["#r \"nuget: FSharp.Stats, 0.4.0\"\n",
"#r \"nuget: Plotly.NET, 2.0.0-beta6\"\n",
"//#load \"..\\IfSharp\\Paket.Generated.Refs.fsx\"\n",
"//#load \"..\\IfSharp\\FSharp.Plotly.fsx\"\n",
"open Plotly.NET\n",
"open FSharp.Stats\n",
"open FSharp.Stats.Fitting.NonLinearRegression\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["## Insert Growth Data and Display as Chart\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"A normal cell culture experiment with measurements for the growth curve will return data like the following.\n",
"Multiple cell counts (y_Count) each related to a specific timepoint (x_Hours).\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 2, "outputs": [
          {
           "data": {
            "text/html": ["\u003cdiv id=\"c6edc6ee-eecf-4f88-a519-856feb6467cf\" style=\"width: 600px; height: 600px;\"\u003e\u003c!-- Plotly chart will be drawn inside this DIV --\u003e\u003c/div\u003e",
"\u003cscript type=\"text/javascript\"\u003e",
"",
"            var renderPlotly_c6edc6eeeecf4f88a519856feb6467cf = function() {",
"            var fsharpPlotlyRequire = requirejs.config({context:\u0027fsharp-plotly\u0027,paths:{plotly:\u0027https://cdn.plot.ly/plotly-latest.min\u0027}}) || require;",
"            fsharpPlotlyRequire([\u0027plotly\u0027], function(Plotly) {",
"",
"            var data = [{\"type\":\"scatter\",\"x\":[0.0,19.5,25.5,43.0,48.5,67.75],\"y\":[1659000.0,4169000.0,6585400.0,16608400.0,17257800.0,18041000.0],\"mode\":\"markers\",\"marker\":{}}];",
"            var layout = {\"title\":\"Growth curve of \u003ci\u003eClamydomonas reinhardtii\u003c/i\u003e cell cultures\",\"xaxis\":{\"title\":\"Number of cells\"},\"yaxis\":{\"title\":\"Time [hours]\"}};",
"            var config = {};",
"            Plotly.newPlot(\u0027c6edc6ee-eecf-4f88-a519-856feb6467cf\u0027, data, layout, config);",
"});",
"            };",
"            if ((typeof(requirejs) !==  typeof(Function)) || (typeof(requirejs.config) !== typeof(Function))) {",
"                var script = document.createElement(\"script\");",
"                script.setAttribute(\"src\", \"https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js\");",
"                script.onload = function(){",
"                    renderPlotly_c6edc6eeeecf4f88a519856feb6467cf();",
"                };",
"                document.getElementsByTagName(\"head\")[0].appendChild(script);",
"            }",
"            else {",
"                renderPlotly_c6edc6eeeecf4f88a519856feb6467cf();",
"            }",
"\u003c/script\u003e",
""]
        },
           "execution_count": 2,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// Code-Block 1\n",
"\n",
"let exmp_x_Hours = [|0.; 19.5; 25.5; 43.; 48.5; 51.25; 67.75|]\n",
"let exmp_y_Count = [|1659000.; 4169000.; 6585400.; 16608400.; 17257800.; nan; 18041000.|]\n",
"\n",
"/// filter out any nans. These could be introduced through missing measurements.\n",
"let exmp_x_Hours_Filtered,exmp_y_Count_Filtered =\n",
"    Array.zip exmp_x_Hours exmp_y_Count\n",
"    |\u003e Array.filter (fun (x,y) -\u003e isNan y = false \u0026\u0026 isNan x = false )\n",
"    |\u003e Array.unzip\n",
"\n",
"// Such data can easily be display with the following code block.\n",
"// Chart.Point takes a sequence of x-axis-points and a series of y-axis-points as input\n",
"let example_Chart_1 = \n",
"    Chart.Point(exmp_x_Hours_Filtered,exmp_y_Count_Filtered)\n",
"    // some minor styling with title and axis-titles.\n",
"    |\u003e Chart.withTitle \"Growth curve of \u003ci\u003eClamydomonas reinhardtii\u003c/i\u003e cell cultures\"\n",
"    |\u003e Chart.withX_AxisStyle (\"Number of cells\")\n",
"    |\u003e Chart.withY_AxisStyle (\"Time [hours]\")\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["## Calculation of growth rate and doubling time for cell cultures\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"The normal growth of an in vitro cell culture is defined through three phases. The lag phase in which the cells still acclimate to the growth conditions, the exponential growth, also called log phase, during which cell growth is exponential due to the proliferation of cells into two daughter cells, and the stationary phase in which the growth rate and the death rate are equal. The stationary phase is typically initiated due to limitations in growth conditions, e.g. depletion of essential nutrients or accumulation of toxic/inhibitory excretions/products. The doubling time defines a time interval in which the quantity of cells doubles and is calculated as seen in Equation 1.\n",
"\n",
"\u003ci\u003eEquation 1: Calculation of the doubling time. Growth rate is calculates as shown in Equation 2.\u003c/i\u003e\n",
"\u003ci\u003eEquation 1: Calculation of the doubling time. Growth rate is calculates as shown in Equation 2.\u003c/i\u003e\n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cimg src=\"https://latex.codecogs.com/gif.latex?doubling\u0026space;Time\u0026space;=\u0026space;\\frac{ln(2)}{growthRate}\" title=\"doubling Time = \\frac{ln(2)}{growthRate}\" style=\"margin: 1rem auto 0; display: block\"/\u003e\n",
"\u003c/div\u003e\n",
"\n",
"Growth rate can then be calculated as shown in Equation 2.\n",
"\n",
"\u003ci\u003eEquation 2: Calculation of the growth rate. With N(t) = the number of cells at time t, N(0) = number of cells at time 0, gr = growth rate, and t = time.\u003c/i\u003e\n",
"\u003ci\u003eEquation 2: Calculation of the growth rate. With N(t) = the number of cells at time t, N(0) = number of cells at time 0, gr = growth rate, and t = time.\u003c/i\u003e\n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cimg src=\"https://latex.codecogs.com/gif.latex?gr=\\frac{ln(\\frac{N(t)}{N(0)})}{t}\" title=\"gr=\\frac{ln(\\frac{N(t)}{N(0)})}{t}\" style=\"margin: 1rem auto 0; display: block\"/\u003e\n",
"\u003c/div\u003e\n",
"\u003c/div\u003e\n",
"\u003chr\u003e\n",
"\n",
"## Fitting biological growth curves\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\n",
"### Theory\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"To derive parameters required for e.g. the doubling time calculation, the measured growth data points have to be modelled. \n",
"In order to obtain a continuous function with known coefficients, a suitable model function is fitted onto the existing data. \n",
"Many models exist, each one of them optimized for a specific task\u003csup\u003e\u003ca href=\"#18\"\u003e18\u003c/a\u003e\u003c/sup\u003e.\n",
"\n",
"Linear model function example: \u003cimg src=\"https://latex.codecogs.com/gif.latex?f(x)\u0026space;=\u0026space;mx\u0026space;\u0026plus;\u0026space;b\" title=\"f(x) = mx + b\" /\u003e\n",
" \n",
"When a model is fitted onto the data, there are endless possibilities to choose coefficients of the model function. \n",
"In the case above there are two coefficients to be identified: The slope m and the y-intercept b. But how can the best fitting coefficients be determined?\n",
"\n",
"Therefore a quality measure called \u003cb\u003eResidual Sum of Squares (RSS)\u003c/b\u003e is used. It describes the discrepancy of the measured points \n",
"and the corresponding estimation model. If the discrepancy is small, the RSS is small too.\n",
"\n",
"In regression analysis the optimal set of coefficients (m and b) that \u003ca href= \"https://mathworld.wolfram.com/LeastSquaresFitting.html\"\u003e minimizes the RSS is searched\u003c/a\u003e.\n",
"\n",
"If there is no straightforward way to identify the RSS-minimizing coefficient set, then the problem is part of nonlinear regression. \n",
"Here, initial coefficients are guessed and the RSS is calculated. Thereafter, the coefficients are modified in tiny steps. \n",
"If the RSS decreases, the direction of the coefficient change seems to be correct. \n",
"By \u003ca href= \"https://books.google.de/books?id=rs51DwAAQBAJ\u0026pg=PA422\u0026lpg=PA422\u0026dq=rss+minimizing+solver\u0026source=bl\u0026ots=qZ0Y4cYtM-\u0026sig=ACfU3U0rHGWCmTo_kv5wqYMmSo8ZKyj5Pg\u0026hl=de\u0026sa=X\u0026ved=2ahUKEwjKtdf-oaHoAhUUwsQBHX07DTwQ6AEwBHoECAkQAQ#v=onepage\u0026q=rss%20minimizing%20solver\u0026f=false\"\u003e \n",
"iteratively changing coefficients\u003c/a\u003e, the optimal coefficient set is determined when no further change leads to an decrease in RSS. \n",
"Algorithms, that perform such a \u0027gradient descent\u0027 methods to solve nonlinear regression tasks are called \u003cb\u003esolver\u003c/b\u003e \n",
"(e.g. Gauss-Newton algorithm or Levenberg‚ÄìMarquardt algorithm). \u003ca href= \"https://www.youtube.com/watch?v=sDv4f4s2SB8\"\u003e Introduction to RSS and optimization problems. \u003c/a\u003e\n",
"\u003c/div\u003e\n",
"\n",
"### Model selection\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"     \n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"Under certain circumstances, more than one solution may arise out of a optimization process. \n",
"If the solutions are based on the same data and the same fitting model, the function minimizing the RSS can be selected as best estimator. \n",
"\u003c/div\u003e\n",
"\n",
"### Exponential Fit\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"Since cellular growth behaves in an exponential manner, it seems to make sense to use an exponential fitting function. \n",
"\n",
"\u003cimg src=\"https://latex.codecogs.com/gif.latex?f(x)\u0026space;=\u0026space;ae^{bx}\" title=\"f(x) = ae^{bx}\" /\u003e\n",
"\u003cimg src=\"https://latex.codecogs.com/gif.latex?f(x)\u0026space;=\u0026space;ae^{bx}\" title=\"f(x) = ae^{bx}\" /\u003e\n",
"\n",
"As seen below, the resulting \u003ca href=\"https://da.khanacademy.org/science/biology/ecology/population-growth-and-regulation/a/exponential-logistic-growth\"\u003e exponential fit\u003c/a\u003e does not represent the data sufficiently, even though it is the best fit, that a exponential model can provide. This is caused by the lag- and stationary phase, both not following an exponential increase. \n",
"In order to use an exponential function as model, it would be necessary to discard data points from lag- and stationary phases and model the remaining data points. \n",
"\n",
" \n",
"There are two main problems regarding this workflow: \n",
"\u003col\u003e\n",
"    \u003cli value=\"(1)\"\u003e The assignment of points to lag-, log-, and stationary phases is a nontrivial task.\u003c/li\u003e\n",
"    \u003cli value=\"(2)\"\u003e The exponential phase only lasts a short period of time and therefore the number of points that can be assigned to the log phase is (very) low.\u003c/li\u003e \n",
"\u003c/ol\u003e\n",
"Consequential the fitted function is not robust against variance introduced during cell count measurements.\n",
"\n",
"\u003c/div\u003e\n",
"\u003c/div\u003e\n",
"\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 4, "outputs": [
          {
           "data": {
            "text/html": ["\u003cdiv id=\"5660b7fa-c253-4db0-b0dc-f6c1b6960abc\" style=\"width: 900px; height: 600px;\"\u003e\u003c!-- Plotly chart will be drawn inside this DIV --\u003e\u003c/div\u003e",
"\u003cscript type=\"text/javascript\"\u003e",
"",
"            var renderPlotly_5660b7fac2534db0b0dcf6c1b6960abc = function() {",
"            var fsharpPlotlyRequire = requirejs.config({context:\u0027fsharp-plotly\u0027,paths:{plotly:\u0027https://cdn.plot.ly/plotly-latest.min\u0027}}) || require;",
"            fsharpPlotlyRequire([\u0027plotly\u0027], function(Plotly) {",
"",
"            var data = [{\"type\":\"scatter\",\"x\":[0.0,19.5,25.5,43.0,48.5,67.75],\"y\":[1659000.0,4169000.0,6585400.0,16608400.0,17257800.0,18041000.0],\"mode\":\"markers\",\"marker\":{},\"name\":\"data points\"},{\"type\":\"scatter\",\"x\":[0.0,1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,10.0,11.0,12.0,13.0,14.0,15.0,16.0,17.0,18.0,19.0,20.0,21.0,22.0,23.0,24.0,25.0,26.0,27.0,28.0,29.0,30.0,31.0,32.0,33.0,34.0,35.0,36.0,37.0,38.0,39.0,40.0,41.0,42.0,43.0,44.0,45.0,46.0,47.0,48.0,49.0,50.0,51.0,52.0,53.0,54.0,55.0,56.0,57.0,58.0,59.0,60.0,61.0,62.0,63.0,64.0,65.0,66.0,67.0],\"y\":[4518669.361090467,4622152.9711177405,4728006.495092799,4836284.207234704,4947041.624715182,5060335.53612391,5176224.030585701,5294766.5275445115,5416023.807229531,5540058.041819007,5666932.827317743,5796713.216164645,5929465.750587027,6065258.496718765,6204161.079499813,6346244.7183749555,6491582.263810117,6640248.234644941,6792318.856300795,6947872.099863786,7106987.722062834,7269747.306163292,7436234.3037971,7606534.077750878,7780733.945733956,7958923.225148724,8141193.278886302,8327637.562170978,8518351.670477455,8713433.38854548,8912982.740516933,9117102.041221175,9325895.948634863,9539471.517543178,9757938.25442996,9981408.173624923,10209995.854736688,10443818.50140112,10682996.001375096,10927650.988006478,11177908.903111847,11433898.061294217,11695749.715733707,11963598.125484915,12237580.624315485,12517837.691121181,12804513.021953547,13097753.603697112,13397709.789433882,13704535.375533808,14018387.680510705,14339427.625684083,14667819.817688258,15003732.632871015,15347338.303625118,15698813.006696932,16058336.953517433,16426094.482601909,16802274.154065747,17187068.846304763,17580675.85488963,17983296.99372512,18395138.69852607,18816412.132662997,19247333.295431815,19688123.132803045,20139007.65070729,20600218.030915134],\"mode\":\"lines\",\"marker\":{},\"name\":\"exponential fit\"}];",
"            var layout = {\"title\":\"Growth curve of \u003ci\u003eClamydomonas reinhardtii\u003c/i\u003e cell cultures\",\"xaxis\":{\"title\":\"time [hours]\",\"mirror\":\"all\",\"showline\":true,\"showgrid\":false},\"yaxis\":{\"title\":\"number of cells\",\"mirror\":\"all\",\"showline\":true,\"showgrid\":false},\"width\":900.0,\"height\":600.0};",
"            var config = {};",
"            Plotly.newPlot(\u00275660b7fa-c253-4db0-b0dc-f6c1b6960abc\u0027, data, layout, config);",
"});",
"            };",
"            if ((typeof(requirejs) !==  typeof(Function)) || (typeof(requirejs.config) !== typeof(Function))) {",
"                var script = document.createElement(\"script\");",
"                script.setAttribute(\"src\", \"https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js\");",
"                script.onload = function(){",
"                    renderPlotly_5660b7fac2534db0b0dcf6c1b6960abc();",
"                };",
"                document.getElementsByTagName(\"head\")[0].appendChild(script);",
"            }",
"            else {",
"                renderPlotly_5660b7fac2534db0b0dcf6c1b6960abc();",
"            }",
"\u003c/script\u003e",
""]
        },
           "execution_count": 4,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// Code-Block 2\n",
"\n",
"// An template exponential function has the form f(x) = a * exp(b * x) with the two unknowns a and b. \n",
"\n",
"// The model we need already exists in FSharp.Stats and can be taken from the \"Table\" module.\n",
"let expModel = Table.expModel\n",
"\n",
"// the solver needs additional information like the initial coefficient guesses or the coefficient accuracy\n",
"// FSharp.Stats assists by estimating the required parameters based on the original input data\n",
"let solverOptions = Table.expSolverOptions exmp_x_Hours_Filtered exmp_y_Count_Filtered\n",
"\n",
"// The Gauss-Newton solver is used to find the optimal coefficients for an exponential function (expModel)\n",
"// The result is a vector, containing parameter a and b as floats.\n",
"let coefficientsExp = GaussNewton.estimatedParams expModel solverOptions exmp_x_Hours_Filtered exmp_y_Count_Filtered\n",
"\n",
"// The determined coefficients can be inserted into the exponential template function\n",
"let fittingExpFunction x = coefficientsExp.[0] * System.Math.Exp(coefficientsExp.[1] * x)\n",
"\n",
"// create a chart with \n",
"let exp_Chart_1 = \n",
"    [|0. .. (Array.last exmp_x_Hours_Filtered)|]\n",
"    |\u003e Array.map (fun xValue -\u003e xValue,fittingExpFunction xValue) // gives tuples of (xValue,yValue)\n",
"    |\u003e Chart.Line\n",
"    |\u003e Chart.withTraceName \"exponential fit\"\n",
"\n",
"// styling of the chart axis\n",
"let templateAxis title = Axis.LinearAxis.init(Title=title,Showgrid=false,Showline=true,Mirror=StyleParam.Mirror.All)       \n",
"\n",
"let exponentialFitChart =\n",
"    [\n",
"        example_Chart_1\n",
"        |\u003e Chart.withTraceName \"data points\"\n",
"        exp_Chart_1\n",
"    ]\n",
"    |\u003e Chart.Combine\n",
"    |\u003e Chart.withX_Axis (templateAxis \"time [hours]\")\n",
"    |\u003e Chart.withY_Axis (templateAxis \"number of cells\")\n",
"    |\u003e Chart.withSize (900.,600.)\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 6, "outputs": [], 
           "source": ["    \n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["## Logistic regression fit\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"As seen above, the model selection is a crucial step for obtaining reasonable functions and to derive function properties with \n",
"which further studies are examined. The selected model should match the theoretical (time) course of the studied signal, but under \n",
"consideration of Occams razor principle. It states, that a approriate model with a low number of coefficients should be preferred over a \n",
"model with many coefficients, since the excessive use of coefficients leads to overfitting.\n",
"\n",
"A better model that can be used in growth curve fitting, is a \u003ca href= \"https://en.wikipedia.org/wiki/Logistic_function\"\u003elogistic function\u003c/a\u003e. \n",
"It is defined by a minimum, a maximum, and a sigmoidal transition between those two. Thereby, the lag, log, and stationary phase are covered.\n",
"\n",
"The function has the form: \u003cimg src=\"https://latex.codecogs.com/gif.latex?f(x)=\\frac{L}{1\u0026plus;e^{-k(x-x_{0})}}\u0026plus;N\" title=\"f(x)=\\frac{L}{1+e^{-k(x-x_{0})}}+N\" /\u003e\n",
"\n",
"where:\n",
"        \n",
"__L__ = curve maximum\n",
"\n",
"__k__ = steepness\n",
"\n",
"__x0__ = xValue of sigmoid\u0027s midpoint\n",
"\n",
"__N__ = curve minimum\n",
"\n",
"In the following, we will go through the necessary steps to calculate the doubling time with the help of a logistic fit. \n",
"This is more complex than the exponential fit, but the given problem requires a more sophisticated method.\n",
"\u003c/div\u003e\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 7, "outputs": [
          {
           "data": {
            "text/html": ["\u003cdiv id=\"9e0b0bd8-a129-4e74-95d4-3486a97a3be8\" style=\"width: 600px; height: 600px;\"\u003e\u003c!-- Plotly chart will be drawn inside this DIV --\u003e\u003c/div\u003e",
"\u003cscript type=\"text/javascript\"\u003e",
"",
"            var renderPlotly_9e0b0bd8a1294e7495d43486a97a3be8 = function() {",
"            var fsharpPlotlyRequire = requirejs.config({context:\u0027fsharp-plotly\u0027,paths:{plotly:\u0027https://cdn.plot.ly/plotly-latest.min\u0027}}) || require;",
"            fsharpPlotlyRequire([\u0027plotly\u0027], function(Plotly) {",
"",
"            var data = [{\"type\":\"scatter\",\"x\":[0.0,19.5,25.5,43.0,48.5,51.25,67.75],\"y\":[1659000.0,4169000.0,6585400.0,16608400.0,17257800.0,\"NaN\",18041000.0],\"mode\":\"markers\",\"marker\":{},\"name\":\"data points\"},{\"type\":\"scatter\",\"x\":[0.0,1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,10.0,11.0,12.0,13.0,14.0,15.0,16.0,17.0,18.0,19.0,20.0,21.0,22.0,23.0,24.0,25.0,26.0,27.0,28.0,29.0,30.0,31.0,32.0,33.0,34.0,35.0,36.0,37.0,38.0,39.0,40.0,41.0,42.0,43.0,44.0,45.0,46.0,47.0,48.0,49.0,50.0,51.0,52.0,53.0,54.0,55.0,56.0,57.0,58.0,59.0,60.0,61.0,62.0,63.0,64.0,65.0,66.0,67.0],\"y\":[1876999.9944503456,1906848.803758243,1940719.94600666,1979136.6527863236,2022684.7682919612,2072018.657528991,2127867.1953612776,2191039.6666429946,2262431.3467984484,2343028.4584132466,2433912.111973107,2536260.7380834147,2651350.406100207,2780552.304140117,2925326.5351030803,3087211.2739702384,3267806.249678159,3468749.482613939,3691686.2545178668,3938229.4447078994,4209910.670986393,4508122.159045681,4834049.954644421,5188599.992463952,5572319.616866827,5985318.341981698,6427192.819717921,6896961.981899393,7393018.92695839,7913106.11522494,8454319.639774118,9013146.6628907,9585537.60242349,10167011.536717804,10752789.949740313,11337950.843477342,11917592.900596859,12486998.189407893,13041782.077243831,13578020.526271453,14092347.520753514,14582018.580428544,15044939.646447834,15479663.619258612,15885359.140182646,16261757.663983287,16609085.456281662,16927986.9842058,17219445.445331562,17484705.117901877,17725199.01706121,17942484.168003645,18138185.766270846,18313950.64696391,18471409.846787687,18612149.60542433,18737689.88846836,18849469.38820629,18948835.935241107,19037041.301068366,19115239.46237758,19184487.511291552,19245748.516880408,19299895.76189491,19347717.88844201,19389924.583880767,19427152.522214647,19459971.346733198],\"mode\":\"lines\",\"marker\":{},\"name\":\"logistic fit\"}];",
"            var layout = {\"yaxis\":{\"title\":\"cell count\",\"mirror\":\"all\",\"showline\":true,\"showgrid\":false},\"xaxis\":{\"title\":\"time [Hours]\",\"mirror\":\"all\",\"showline\":true,\"showgrid\":false}};",
"            var config = {};",
"            Plotly.newPlot(\u00279e0b0bd8-a129-4e74-95d4-3486a97a3be8\u0027, data, layout, config);",
"});",
"            };",
"            if ((typeof(requirejs) !==  typeof(Function)) || (typeof(requirejs.config) !== typeof(Function))) {",
"                var script = document.createElement(\"script\");",
"                script.setAttribute(\"src\", \"https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js\");",
"                script.onload = function(){",
"                    renderPlotly_9e0b0bd8a1294e7495d43486a97a3be8();",
"                };",
"                document.getElementsByTagName(\"head\")[0].appendChild(script);",
"            }",
"            else {",
"                renderPlotly_9e0b0bd8a1294e7495d43486a97a3be8();",
"            }",
"\u003c/script\u003e",
""]
        },
           "execution_count": 7,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// Code-Block 3\n",
"\n",
"// The model we need already exists in FSharp.Stats and can be taken from the \"Table\" module.\n",
"let modelLogistic = Table.LogisticFunctionVarYAscending\n",
"\n",
"// To fit the logistic function, the solver requires more parameters. Some of them are stored in the solverOption type\n",
"let lineSolverOptions initialParamGuess = {\n",
"    // defines the stepwidth of the x_value change\n",
"    MinimumDeltaValue       = 0.00001\n",
"    // defines the stepwidth of the parameter change\n",
"    MinimumDeltaParameters  = 0.00001\n",
"    // defines the number of iterations until the solver converges to a solution\n",
"    MaximumIterations       = 10000\n",
"    // initial parameters to start the solving algorithms\n",
"    // vector containing all coefficients of the function: vector [L;k;x0;N]\n",
"    InitialParamGuess       = initialParamGuess\n",
"    }\n",
"\n",
"// Generation of solverOptions with varying steepnesses\n",
"let initialGuess = \n",
"    // maximum measured intensity/cell count\n",
"    let l  = exmp_y_Count_Filtered |\u003e Array.max\n",
"    // estimate of the xValue of sigmoid\u0027s midpoint\n",
"    let x0 = (exmp_x_Hours_Filtered |\u003e Array.max) / 2.\n",
"    // minimum measured intensity/cell count.\n",
"    let n  = exmp_y_Count_Filtered |\u003e Array.min \n",
"    \n",
"    //since steepness in unknown, a variety of steepnesses is provided \n",
"    let steepnessRange = [|0.01 .. 0.01 .. 1.|]\n",
"    \n",
"    steepnessRange\n",
"    |\u003e Array.map (fun steepness -\u003e lineSolverOptions [|l; steepness; x0; n|])\n",
"\n",
"// Estimate coefficients for a possible solution based on residual sum \n",
"// of squares. Besides the solverOptions, an upper and lower bound for \n",
"// the parameters are required. It is recommended to define them depending \n",
"// on the initial param guess.\n",
"// It reports coefficients tupled with a corresponding RSS value.\n",
"let estimateCoefficientsRSS =\n",
"    initialGuess\n",
"    |\u003e Array.map (fun solvOpt -\u003e\n",
"        let lowerBound =\n",
"            solvOpt.InitialParamGuess\n",
"            |\u003e Array.map (fun param -\u003e param - (abs param) * 0.2)\n",
"            |\u003e vector\n",
"        let upperBound =\n",
"            solvOpt.InitialParamGuess\n",
"            |\u003e Array.map (fun param -\u003e param + (abs param) * 0.2)\n",
"            |\u003e vector\n",
"        // all parameters are given to the solver and the set of coefficients that minimize the RSS are reported\n",
"        LevenbergMarquardtConstrained.estimatedParamsWithRSS \n",
"            modelLogistic         // logistic function model\n",
"            solvOpt               // solver options with optimization parameters and initial guess\n",
"            0.001                 //\n",
"            10.                   //\n",
"            lowerBound            // lower bound of coefficient space\n",
"            upperBound            // upper bound of coefficient space\n",
"            exmp_x_Hours_Filtered // x data\n",
"            exmp_y_Count_Filtered // y data\n",
"    )\n",
"    |\u003e Array.filter (fun (coeffs,rss) -\u003e not(coeffs |\u003e Vector.exists System.Double.IsNaN)) // discard solutions with nan as coefficients\n",
"    |\u003e Array.minBy snd // minimize all possible solutions based on RSS quality measure\n",
"    |\u003e fun (solCoeffs,rss) -\u003e\n",
"        printfn \"Chosen Estimate: %A\" solCoeffs\n",
"        printfn \"Equation: (%.1f / (1. + exp(-%.3f * (x - %.3f)))) + %.1f\" solCoeffs.[0] solCoeffs.[1] solCoeffs.[2] solCoeffs.[3]\n",
"        solCoeffs\n",
"        \n",
"// Create fitting function from optimal coefficients\n",
"let fittingLogisticFunction = modelLogistic.GetFunctionValue estimateCoefficientsRSS\n",
"\n",
"// Code-Block 4\n",
"\n",
"// fit of the optimized logistic function over all x Values\n",
"let fittedY = \n",
"     [|0. .. exmp_x_Hours |\u003e Array.max|]\n",
"     |\u003e Array.map (fun x -\u003e x, fittingLogisticFunction x) //tupled (xValue,yValue)\n",
"\n",
"let fittedLogisticFunc =\n",
"    [\n",
"        // raw chart\n",
"        Chart.Point (exmp_x_Hours, exmp_y_Count) |\u003e Chart.withTraceName\"data points\"\n",
"        // logistic fit\n",
"        Chart.Line fittedY                       |\u003e Chart.withTraceName \"logistic fit\"\n",
"    ]\n",
"    |\u003e Chart.Combine\n",
"    |\u003e Chart.withY_Axis (templateAxis \"cell count\")\n",
"    |\u003e Chart.withX_Axis (templateAxis \"time [Hours]\")\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["## Calculate Doubling Time\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"To calculate the doubling time it is necessary to determine the growth rate (gr) for \u003ci\u003eequation 1\u003c/i\u003e.\n",
"To get gr we make use of the first and second derivative of the logistic function. They can be calculated by hand or with help \n",
"of \u003ca href=\"https://www.ableitungsrechner.net/\"\u003ederivative calculator\u003c/a\u003e.\n",
"\n",
"The first derivative of the logistic function is: \n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cimg src=\"https://latex.codecogs.com/gif.latex?\\dfrac{kl\\mathrm{e}^{k\\left(x-m\\right)}}{\\left(\\mathrm{e}^{k\\left(x-m\\right)}\u0026plus;1\\right)^2}\" title=\"\\dfrac{kl\\mathrm{e}^{k\\left(x-m\\right)}}{\\left(\\mathrm{e}^{k\\left(x-m\\right)}+1\\right)^2}\" style=\"margin: 1rem auto 0; display: block\" /\u003e\n",
"\u003c/div\u003e\n",
"\n",
"The second derivative of the logistic function is: \n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cimg src=\"https://latex.codecogs.com/gif.latex?-\\dfrac{k^2l\\left(\\mathrm{e}^{k\\left(x-m\\right)}-1\\right)\\mathrm{e}^{k\\left(x-m\\right)}}{\\left(\\mathrm{e}^{k\\left(x-m\\right)}\u0026plus;1\\right)^3}\" title=\"-\\dfrac{k^2l\\left(\\mathrm{e}^{k\\left(x-m\\right)}-1\\right)\\mathrm{e}^{k\\left(x-m\\right)}}{\\left(\\mathrm{e}^{k\\left(x-m\\right)}+1\\right)^3}\" style=\"margin: 1rem auto 0; display: block\" /\u003e\n",
"\u003c/div\u003e\n",
"\n",
"\u003c/div\u003e\n",
"\u003c/div\u003e\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 9, "outputs": [
          {
           "data": {
            "text/html": ["\u003cdiv id=\"84e72074-3dc3-4dbd-9717-e2106f81e7a8\" style=\"width: 600px; height: 600px;\"\u003e\u003c!-- Plotly chart will be drawn inside this DIV --\u003e\u003c/div\u003e",
"\u003cscript type=\"text/javascript\"\u003e",
"",
"            var renderPlotly_84e720743dc34dbd9717e2106f81e7a8 = function() {",
"            var fsharpPlotlyRequire = requirejs.config({context:\u0027fsharp-plotly\u0027,paths:{plotly:\u0027https://cdn.plot.ly/plotly-latest.min\u0027}}) || require;",
"            fsharpPlotlyRequire([\u0027plotly\u0027], function(Plotly) {",
"",
"            var data = [{\"type\":\"scatter\",\"x\":[0.0,0.5,1.0,1.5,2.0,2.5,3.0,3.5,4.0,4.5,5.0,5.5,6.0,6.5,7.0,7.5,8.0,8.5,9.0,9.5,10.0,10.5,11.0,11.5,12.0,12.5,13.0,13.5,14.0,14.5,15.0,15.5,16.0,16.5,17.0,17.5,18.0,18.5,19.0,19.5,20.0,20.5,21.0,21.5,22.0,22.5,23.0,23.5,24.0,24.5,25.0,25.5,26.0,26.5,27.0,27.5,28.0,28.5,29.0,29.5,30.0,30.5,31.0,31.5,32.0,32.5,33.0,33.5,34.0,34.5,35.0,35.5,36.0,36.5,37.0,37.5,38.0,38.5,39.0,39.5,40.0,40.5,41.0,41.5,42.0,42.5,43.0,43.5,44.0,44.5,45.0,45.5,46.0,46.5,47.0,47.5,48.0,48.5,49.0,49.5,50.0,50.5,51.0,51.5,52.0,52.5,53.0,53.5,54.0,54.5,55.0,55.5,56.0,56.5,57.0,57.5,58.0,58.5,59.0,59.5,60.0,60.5,61.0,61.5,62.0,62.5,63.0,63.5,64.0,64.5,65.0,65.5,66.0,66.5,67.0,67.5],\"y\":[3551.72073781862,3777.8937928557475,4017.5939884897025,4271.510135682126,4540.349121165499,4824.833681373931,5125.699644023785,5443.692566850858,5779.563696868922,6134.065167350835,6507.944343690394,6901.937223558289,7316.760791543541,7753.104224047656,8211.618836905875,8692.906666448578,9197.507574956866,9725.884774261194,10278.408667213518,10855.338916652016,11456.804666078437,12082.78285646795,12733.074610388201,13407.279688929255,14104.769069879436,14824.655748152707,15565.7639226529,16326.596808394195,17105.303399426222,17899.644607270533,18706.95931107266,19524.130978861704,20347.555652825693,21173.112233120584,21996.136141158582,22811.39759009122,23613.08583150179,24394.800875896042,25149.554291704862,25869.780763029623,26547.362118853394,27173.66552353144,27739.597427205317,28235.67470267929,28652.114130625687,28978.941028214023,29206.117341173976,29323.688934333117,29321.951125250212,29191.63072139305,28924.081963388286,28511.492874150248,27947.097604140232,27225.38949293906,26342.32878988609,25295.53834926789,24084.48019692742,22710.605710490403,21177.47231196401,19490.820073581566,17658.602502082904,15690.966988116355,13600.181956821849,11400.509578465888,9108.024915968317,6740.384500774866,4316.549426835727,1856.470014710723,-619.259193618629,-3089.7640000779484,-5534.263664379553,-7932.441157232413,-10264.799276940164,-12512.992008865795,-14660.121641946898,-16690.993697231726,-18592.323573028352,-20352.890856597245,-21963.63937351772,-23417.72312552703,-24710.500199182283,-25839.47842213916,-26804.217934515178,-27606.19688781282,-28248.647166289007,-28736.36735170647,-29075.520148126878,-29273.421190235298,-29338.325628526418,-29279.218174992675,-29105.611462218963,-28827.356672713122,-28454.469483870198,-27996.973489459204,-27464.76243442453,-26867.481860536518,-26214.4301217513,-25514.47819796514,-24776.007315438856,-24006.863067460177,-23214.324511772546,-22405.08659120378,-21585.254168501306,-20760.345972763498,-19935.30681035712,-19114.526485885548,-18301.86399780475,-17500.675709219206,-16713.846339239484,-15943.821767522164,-15192.642789112146,-14461.97909461132,-13753.162879269985,-13067.221602062373,-12404.909521178128,-11766.737725287096,-11153.002460552805,-10563.811622181935,-9999.109337056869,-9458.698611624133,-8942.262057710292,-8449.380739348731,-7979.551207055741,-7532.200803278857,-7106.701334879522,-6702.381216351653,-6318.53619178153,-5954.4387449982305,-5609.346306540114,-5282.508363481189,-4973.1725742573435,-4680.589985771152,-4404.019444539841,-4142.731287731809,-3896.0103938140696,-3663.15866636487],\"mode\":\"lines\",\"marker\":{},\"name\":\"fst derivative\"},{\"type\":\"scatter\",\"x\":[0.0,0.5,1.0,1.5,2.0,2.5,3.0,3.5,4.0,4.5,5.0,5.5,6.0,6.5,7.0,7.5,8.0,8.5,9.0,9.5,10.0,10.5,11.0,11.5,12.0,12.5,13.0,13.5,14.0,14.5,15.0,15.5,16.0,16.5,17.0,17.5,18.0,18.5,19.0,19.5,20.0,20.5,21.0,21.5,22.0,22.5,23.0,23.5,24.0,24.5,25.0,25.5,26.0,26.5,27.0,27.5,28.0,28.5,29.0,29.5,30.0,30.5,31.0,31.5,32.0,32.5,33.0,33.5,34.0,34.5,35.0,35.5,36.0,36.5,37.0,37.5,38.0,38.5,39.0,39.5,40.0,40.5,41.0,41.5,42.0,42.5,43.0,43.5,44.0,44.5,45.0,45.5,46.0,46.5,47.0,47.5,48.0,48.5,49.0,49.5,50.0,50.5,51.0,51.5,52.0,52.5,53.0,53.5,54.0,54.5,55.0,55.5,56.0,56.5,57.0,57.5,58.0,58.5,59.0,59.5,60.0,60.5,61.0,61.5,62.0,62.5,63.0,63.5,64.0,64.5,65.0,65.5,66.0,66.5,67.0,67.5],\"y\":[27997.55042954533,29829.404525640653,31777.698635375098,33849.3677386603,36051.69582236704,38392.32438248587,40879.25968297677,43520.87848776192,46325.93194554013,49303.54726723811,52463.22679299692,55814.84399971804,59368.635951566655,63135.19164478068,67125.43564513883,71350.60636216526,75822.2282494707,80552.07716668647,85552.13808667105,90834.55428382658,96411.56709760898,102295.44533222217,108498.40333209233,115032.50676655008,121909.5651702533,129141.01032282092,136737.75961698365,144710.06366481725,153067.33753220277,161817.97517870058,170969.14691974907,180526.5800265311,190494.32294153806,200874.49401943758,211667.01620660108,222869.3396499081,234476.15487499651,246479.09989147526,258866.46535929176,271622.902773225,284729.1414727595,298161.72113780427,311892.7472559478,325889.67780686385,340115.15006065735,354526.85688059445,369077.4822040281,383714.70539362234,398381.2838501108,413015.22260764573,427550.03855107696,441915.1253712289,456036.2233961357,469835.99601109535,483234.71154110396,496151.02627630095,508502.86086426536,520208.35869069904,531186.9112664765,541360.232199047,550653.4582275247,558996.2532243154,566323.8891840827,572578.2771846725,577708.9212323071,581673.7688686969,584439.9344411602,585984.2739804646,586293.794597164,585365.8860416064,583208.3673730345,579839.3473102232,575286.9025296156,569588.583672136,562790.7638651008,554947.8489384135,546121.3720349341,536378.9978569549,525793.4632861186,514441.4815545595,502402.6365785915,489758.29259001836,476590.54195036035,462981.21117335884,449010.94188913994,434758.35994001426,420299.3421755003,405706.38696809753,391048.0911334307,376388.73291289236,361787.9580389245,347300.56369901344,332976.3734642296,318860.19494845276,304991.85109310335,291406.2754925406,278133.66204060736,265199.65933650127,252625.6006834418,240428.7610924011,228622.63341450322,217217.2165233366,206219.30931190067,195632.80512412256,185458.98208009748,175696.78555590566,166343.0998272111,157393.0065700895,148840.0285262795,140676.3571805336,132893.06376510544,125480.29330303526,118427.4417318069,111723.31641729723,105356.28058054965,99314.3823229372,93585.46905494288,88157.2882161166,83017.57522449184,78154.12961816373,73554.88035466065,69207.94121948647,65101.65726755868,61224.643183492204,57565.81440158831,54114.41177633268,50860.02054114887,47792.5842386825,44902.414251290305,42180.19550667658,39616.988881527206,37204.23077610313,34933.73028547341,32797.66434865201,30788.571215506843,28899.342532990813],\"mode\":\"lines\",\"marker\":{},\"name\":\"snd derivative\"}];",
"            var layout = {\"yaxis\":{\"title\":\"slope or curvature\",\"mirror\":\"all\",\"showline\":true,\"showgrid\":false},\"xaxis\":{\"title\":\"time [hours]\",\"mirror\":\"all\",\"showline\":true,\"showgrid\":false}};",
"            var config = {};",
"            Plotly.newPlot(\u002784e72074-3dc3-4dbd-9717-e2106f81e7a8\u0027, data, layout, config);",
"});",
"            };",
"            if ((typeof(requirejs) !==  typeof(Function)) || (typeof(requirejs.config) !== typeof(Function))) {",
"                var script = document.createElement(\"script\");",
"                script.setAttribute(\"src\", \"https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js\");",
"                script.onload = function(){",
"                    renderPlotly_84e720743dc34dbd9717e2106f81e7a8();",
"                };",
"                document.getElementsByTagName(\"head\")[0].appendChild(script);",
"            }",
"            else {",
"                renderPlotly_84e720743dc34dbd9717e2106f81e7a8();",
"            }",
"\u003c/script\u003e",
""]
        },
           "execution_count": 9,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// Code-Block 5\n",
"\n",
"// calculate fst derivative of logistic function\n",
"let fstDerivative x = \n",
"    let l  = estimateCoefficientsRSS.[0]\n",
"    let k  = estimateCoefficientsRSS.[1]\n",
"    let x0 = estimateCoefficientsRSS.[2]\n",
"    let n  = estimateCoefficientsRSS.[3]\n",
"    let exp = System.Math.Exp(k*(x-x0))\n",
"    - (k**2.*l*(exp - 1.)* exp ) / (exp + 1.)**3.\n",
"\n",
"// calculate snd derivative of logistic function \n",
"let sndDerivative x = \n",
"    let l  = estimateCoefficientsRSS.[0]\n",
"    let k  = estimateCoefficientsRSS.[1]\n",
"    let x0 = estimateCoefficientsRSS.[2]\n",
"    let n  = estimateCoefficientsRSS.[3]\n",
"    let exp = System.Math.Exp(k*(x-x0))\n",
"    (k*l*exp) / (exp + 1.)**2.\n",
"\n",
"// calculate derivatives to corresponding x values\n",
"let yValuesOfDerivative fkt = \n",
"    [|0. .. 0.5 .. exmp_x_Hours |\u003e Array.max|]\n",
"    |\u003e Array.map (fun x -\u003e x,fkt x)\n",
"    \n",
"let fitAllLogisticFunc =\n",
"    [\n",
"        Chart.Line (yValuesOfDerivative fstDerivative)|\u003e Chart.withTraceName \"fst derivative\"\n",
"        Chart.Line (yValuesOfDerivative sndDerivative)|\u003e Chart.withTraceName \"snd derivative\"\n",
"    ]\n",
"    |\u003e Chart.Combine\n",
"    |\u003e Chart.withY_Axis (templateAxis \"slope or curvature\")\n",
"    |\u003e Chart.withX_Axis (templateAxis \"time [hours]\")\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"We define the region between the maximal and minimal curvature (second derivative) as the time period to derive the growth rate from. \n",
"An alternative is to just use the slope at the midpoint because this is the point of maximal slope (minimal generation time), but since \n",
"this calculation would be only dependent from this particular point we go for the more conservative approach.\n",
"    \n",
"When the xValues of the maximal curvatures are identified (either by calculus or by plotting the derivatives) the generation time calculation\n",
" is straight forward (\u003ca href=\"https://en.wikipedia.org/wiki/Doubling_time\"\u003eWikipedia - Doubling time\u003c/a\u003e).\n",
"\u003c/div\u003e\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 11, "outputs": [
          {
           "data": {
            "text/plain": ["\"The doubling time is 13.12 hours.\""]
        },
           "execution_count": 11,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// Code-Block 6\n",
"\n",
"// The exponential phase is considered to be between the maximal positive curvature\n",
"// and the minimal negative curvature of the fitting functions (other interpretations possible).\n",
"let xValuesOfMaximalCurvature = [| 23.5 ; 44.0 |]\n",
"\n",
"let calculateDoublingTimeLogistic fittingFunction =\n",
"\n",
"    //https://en.wikipedia.org/wiki/Doubling_time\n",
"    // -\u003e Cell culture doubling time\n",
"    let growthRate nCells0 nCellsT t =\n",
"        log(nCellsT/nCells0)\n",
"        |\u003e fun x -\u003e x/t\n",
"\n",
"    let doublingTime growthRate =\n",
"        (log(2.))/growthRate\n",
"\n",
"    /// get the corresponding cell counts around the midpoint of the logistic function\n",
"    let rootsYY = xValuesOfMaximalCurvature |\u003e Array.map fittingFunction\n",
"    \n",
"    /// calculate the time difference between both roots\n",
"    let diff = xValuesOfMaximalCurvature.[1] - xValuesOfMaximalCurvature.[0]\n",
"    \n",
"    /// get the minimum measured cell count and the maximum measured cell count\n",
"    let min,max = rootsYY.[0], rootsYY.[1]\n",
"\n",
"    let doublingTime =\n",
"        growthRate min max diff\n",
"        |\u003e doublingTime\n",
"\n",
"    doublingTime\n",
"    \n",
"let doublingTime = \n",
"    calculateDoublingTimeLogistic fittingLogisticFunction\n",
"    \n",
"sprintf \"The doubling time is %.2f hours.\" doublingTime\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["\u003cnav class=\"level is-mobile\"\u003e\n",
"\u003cnav class=\"level is-mobile\"\u003e\n",
"    \u003cdiv class=\"level-left\"\u003e\n",
"        \u003cdiv class=\"level-item\"\u003e\n",
"            \u003cbutton class=\"button is-primary is-outlined\" onclick=\"location.href=\u0027/JP01_FSharpExcercises.html\u0027;\"\u003e\u0026#171; JP01\u003c/button\u003e\n",
"        \u003c/div\u003e\n",
"    \u003c/div\u003e\n",
"    \u003cdiv class=\"level-right\"\u003e\n",
"        \u003cdiv class=\"level-item\"\u003e\n",
"            \u003cbutton class=\"button is-primary is-outlined\" onclick=\"location.href=\u0027/JP03_Mass_spectrometry_based_proteomics.html\u0027;\"\u003eJP03 \u0026#187;\u003c/button\u003e\n",
"        \u003c/div\u003e\n",
"    \u003c/div\u003e\n",
"\u003c/nav\u003e\n",
"\n",
"## References\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Plant-Systems-Biology\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\u003cbr\u003e\n",
"\u003cbr\u003e\n",
"\n",
"\u003col\u003e\n",
"\u003col\u003e\n",
"\u003cli Value=\"8\" Id=\"8\"\u003eRuffel, S., Krouk, G. \u0026 Coruzzi, G. M. A systems view of responses to nutritional cues in Arabidopsis: toward a paradigm shift for predictive network modeling. Plant physiology 152, 445‚Äì452; 10.1104/pp.109.148502 (2010).\u003c/li\u003e\n",
"\u003cli Id=\"9\"\u003e Minorsky, P. V. Achieving the in Silico Plant. Systems Biology and the Future of Plant Biological Research. Plant physiology 132, 404‚Äì409; 10.1104/pp.900076 (2003). \u003cdiv\u003e\n",
"\u003cli Id=\"10\"\u003eVan Norman, Jaimie M \u0026 Benfey, P. N. Arabidopsis thaliana as a model organism in systems biology. Wiley interdisciplinary reviews. Systems biology and medicine 1, 372‚Äì379; 10.1002/wsbm.25 (2009).\u003c/li\u003e\n",
"\u003cli Id=\"11\"\u003eP√£curar, D. I. Model organisms - a journey from the dawn of biological research to the post-genomic era. Romanian Society of Biological Sciences, 4087‚Äì4094 (2009).\u003c/li\u003e\n",
"\u003cli Id=\"12\"\u003eCagnon, C. et al. Development of a forward genetic screen to isolate oil mutants in the green microalga Chlamydomonas reinhardtii. Biotechnology for biofuels 6, 178; 10.1186/1754-6834-6-178 (2013).\u003c/li\u003e\n",
"\u003cli Id=\"13\"\u003eEberhard, S., Finazzi, G. \u0026 Wollman, F.-A. The dynamics of photosynthesis. Annual review of genetics 42, 463‚Äì515; 10.1146/annurev.genet.42.110807.091452 (2008).\u003c/li\u003e\n",
"\u003cli Id=\"14\"\u003eMerchant, S. S. et al. The Chlamydomonas genome reveals the evolution of key animal and plant functions. Science (New York, N.Y.) 318, 245‚Äì250; 10.1126/science.1143609 (2007).\u003c/li\u003e\n",
"\u003cli Id=\"15\"\u003eHarris, E. H. The chlamydomonas sourcebook. 2nd ed. (Academic, London, 2008).\u003c/li\u003e\n",
"\u003cli Id=\"16\"\u003eBruggeman, F. J. \u0026 Westerhoff, H. V. The nature of systems biology. Trends in microbiology 15, 45‚Äì50; 10.1016/j.tim.2006.11.003 (2007).\u003c/li\u003e\n",
"\u003cli Id=\"17\"\u003eHarris, E. H. CHLAMYDOMONAS AS A MODEL ORGANISM. Annual review of plant physiology and plant molecular biology 52, 363‚Äì406; 10.1146/annurev.arplant.52.1.363 (2001).\u003c/li\u003e\n",
"\u003cli Id=\"18\"\u003eKaplan, S. et al. Comparison of growth curves using non-linear regression function in Japanese squail. Journal of Applied Animal Research 46, 112-117; 10.1080/09712119.2016.1268965 (2018).\u003c/li\u003e.\n",
"\u003c/ol\u003e\n",
"\n"]
          }],
            "metadata": {
            "kernelspec": {"display_name": ".NET (F#)", "language": "F#", "name": ".net-fsharp"},
            "langauge_info": {
        "file_extension": ".fs",
        "mimetype": "text/x-fsharp",
        "name": "C#",
        "pygments_lexer": "fsharp",
        "version": "4.5"
        }
        },
            "nbformat": 4,
            "nbformat_minor": 1
        }
        

