<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <title>JP03 Mass Spectrometry Based Proteomics
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="author" content="">
    <link rel="shortcut icon" type="image/x-icon" href="/BIO-BTE-06-L-7//img/favicon.ico">
    <script src="https://code.jquery.com/jquery-3.4.1.min.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <!-- <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/MathJax.js?config=TeX-MML-AM_CHTML"></script> -->

    <link type="text/css" rel="stylesheet" href="/BIO-BTE-06-L-7/content/fsdocs-default.css" />
    <link type="text/css" rel="stylesheet" href="/BIO-BTE-06-L-7/content/fsdocs-custom.css" />
    <script type="text/javascript" src="/BIO-BTE-06-L-7/content/fsdocs-tips.js"></script>
    
    <link type="text/css" rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" />
    <script defer type="text/javascript" src="https://kit.fontawesome.com/0d3e0ea7a6.js"></script>
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <!-- END SEARCH BOX: this adds support for the search box -->
    
</head> 

<body>
    <div class="columns is-fullheight m-0">
        <div id="fsdocs-menu" class="column is-2 is-paddingless">
            <aside class="menu">
                <div class="container is-paddingless is-marginless has-background-dark">
                    <a href="/BIO-BTE-06-L-7/">
                        <figure class="image is-128x128 container">
                            <img src="/BIO-BTE-06-L-7/img/logo.png">
                        </figure>
                    </a>
                    <h2 class="title has-text-centered has-text-white pb-4">BIO-BTE-06-L-7</h2>
                </div>
                <div class="content">
                    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
                    <div id="header">
                        <div class="field">
                            <div class="control has-icons-left has-icons-right searchbox" id="fsdocs-searchbox">
                                <input data-search-input="" id="search-by" type="search" placeholder="Search..." class="input"/>
                                <span class="icon is-small is-left">
                                    <i class="fas fa-search"></i>
                                </span>
                            </div>
                        </div>
                    </div>

                    <!-- END SEARCH BOX: this adds support for the search box -->
                    <ul class="navbar-nav">
                        <li class="nav-header">
  Documentation
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP05_Isotopic_distribution.html">
    JP05 Isotopic Distribution

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP00_Systems_Biology_FSharp_Introduction.html">
    JP00 Systems Biology

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP04_Digestion_and_mass_calculation.html">
    JP04 Digestion and mass calculation

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP01_FSharpExcercises.html">
    JP01 FSharp Introduction Excercises

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP02_Plant_Systems_Biology.html">
    JP02 Plant Systems Biology

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP03_Mass_spectrometry_based_proteomics.html">
    JP03 Mass Spectrometry Based Proteomics

  </a>
</li>
                        <li class="nav-header">Links</li>
                        <li class="nav-item" >
                            <a class="nav-link" href="https://csb.bio.uni-kl.de/teaching/">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fas fa-graduation-cap"></i>
                                    </span>
                                    <span>Other CSB courses</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item" >
                            <a class="nav-link" href="https://twitter.com/cs_biology">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fab fa-twitter"></i>
                                    </span>
                                    <span>Get in touch on twitter</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item" >
                            <a class="nav-link" href="https://github.com/CSBiology?type=source">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fab fa-github"></i>
                                    </span>
                                    <span>Check out our open source projects</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://github.com/CSBiology/BIO-BTE-06-L-7">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fas fa-code-branch"></i>
                                    </span>
                                    <span>Source repository</span>
                                </span>
                            </a>
                        </li>
                    </ul>
                </div>
            </aside>
        </div>
        <div class="column is-10 is-paddingless has-text-justified	">
            <div class="section pt-4">
                <div class="container" id="fsdocs-content">
                    <h1><a name="JP03-Mass-Spectrometry-Based-Proteomics" class="anchor" href="#JP03-Mass-Spectrometry-Based-Proteomics">JP03 Mass Spectrometry Based Proteomics</a></h1>
<p><a href="https://mybinder.org/v2/gh/CSBiology/BIO-BTE-06-L-7/gh-pages?filepath=JP03_Mass_spectrometry_based_proteomics.ipynb"><img src="https://mybinder.org/badge_logo.svg" alt="Binder" /></a></p>
<ol>
<li><a href="#Mass-spectrometry-(MS)-based-proteomic">Mass spectrometry (MS)-based proteomic</a></li>
<li><a href="#References">References</a></li>
</ol>
<h2><a name="Mass-spectrometry-MS-based-proteomic" class="anchor" href="#Mass-spectrometry-MS-based-proteomic">Mass spectrometry (MS)-based proteomic</a></h2>
<a href="#Mass-Spectrometry-Based-Proteomics" style="display: inline-block"><sup>&#8593;back</sup></a><br>
<div class="container">
<div Id= "figure2" Style="float: right; display: inline-block; color: #44546a; width: 60%; padding: 15px; max-width: 600px">
    <img src="img/OmicsWorkflow.png" Style="width: 100%">
    <div Style="padding-left: 1rem; padding-right: 1rem; text-align: justify; font-size:0.8rem">
        <b>Figure 2. Summary of a typical proteomics workflow following the bottom-up principle.</b> During sample preparation proteins are extracted from the samples and digested into peptides using proteases typically trypsin. An optional fractionation or enrichment may be applied at either the protein or peptide level to enhance the scope of identification. Peptides are separated by high-performance liquid chromatography (HPLC) and afterward transferred into the vacuum of the mass spectrometer mostly using electrospray ionization (ESI). Cycles of full MS including all peptide at a time followed by MS/MS of selected peptides are measured. The consecutive measured MS and MS/MS spectra are then used to computationally identify and quantify the peptide sequence and infer the protein.
    </div>
</div>  
<p>The proteome is understood as an entire complement of proteins in one cell, tissue or a whole organism.
Proteomics as a scientific field deals with the qualitative and quantitative analysis of protein expression patterns.
Thus, proteomics relies primarily on the ability to unambiguously identify proteins, followed by accurate quantification.
Mass spectrometry-based proteomics refers to the large-scale analysis of
proteins using mass spectrometry, an analytical method to determine the mass of molecules. In the case of proteomics, the target
molecules are whole proteins or peptides.</p>
<p>The typical proteomics workflow is as follows (<a href="#figure2">Figure 2</a>): first, proteins are isolated from
cells or tissues by lysis followed by biochemical fractionation or affinity selection. MS on whole proteins (top-down proteomics)
is less sensitive and more difficult to handle when compared to MS on peptides (bottom-up proteomics), as the mass of the intact protein by
itself is insufficient for protein identification<sup><a href="#18">18</a>,<a href="#19">19</a></sup>. Therefore, the bottom-up approach is
standardly used, which comprises the enzymatic degradation of proteins to peptides using an endopeptidase (trypsin in most cases)
<sup><a href="#20">20</a></sup>. Trypsin is advantageous as it generates peptides with C-terminally protonated amino acids to foster
the detectability of a full ion ladder (series) in a subsequent, optional fragmentation step for sequencing. The peptides are separated by one
or more steps of liquid chromatography (LC) and afterwards transferred into the vacuum of the mass spectrometer, where a mass spectrum
of the peptides eluting at this time from the LC is taken (MS<sup>1</sup> spectrum or ‘normal mass spectrum’). A prioritized list of these peptides
for fragmentation is either automatically generated or provided by the operator. These peptide ions are then fragmented by energetic collision with
an inert gas and recorded as tandem MS spectra (MS/MS or MS<sup>2</sup> spectra). The consecutive MS and MS/MS
spectra are then used to computationally identify the peptide’s sequence and quantify its abundance<sup><a href="#21">22</a>,
<a href="#22">22</a></sup>.</p>
</div>
<hr>
<nav class="level is-mobile">
    <div class="level-left">
        <div class="level-item">
            <button class="button is-primary is-outlined" onclick="location.href='/JP02_Plant_Systems_Biology.html';">&#171; JP02</button>
        </div>
    </div>
    <div class="level-right">
        <div class="level-item">
            <button class="button is-primary is-outlined" onclick="location.href='/JP04_Digestion_and_mass_calculation.html';">JP04 &#187;</button>
        </div>
    </div>
</nav>
<h2><a name="References" class="anchor" href="#References">References</a></h2>
<a href="#Mass-Spectrometry-Based-Proteomics" style="display: inline-block"><sup>&#8593;back</sup></a><br>
<ol>
<li Value="18" Id="18"> Breuker, K., Jin, M., Han, X., Jiang, H. & McLafferty, F. W. Top-down identification and characterization of biomolecules by mass spectrometry. Journal of the American Society for Mass Spectrometry 19, 1045–1053; 10.1016/j.jasms.2008.05.013 (2008).</li>
<li Id="19"> Reid, G. E. & McLuckey, S. A. 'Top down' protein characterization via tandem mass spectrometry. Journal of mass spectrometry : JMS 37, 663–675; 10.1002/jms.346 (2002).</li>
<li Id="20"> Olsen, J. V., Ong, S.-E. & Mann, M. Trypsin cleaves exclusively C-terminal to arginine and lysine residues. Molecular & cellular proteomics : MCP 3, 608–614; 10.1074/mcp.T400003-MCP200 (2004).</li>
<li Id="21"> Walther, T. C. & Mann, M. Mass spectrometry-based proteomics in cell biology. J. Cell Biol. 190, 491–500; 10.1083/jcb.201004052 (2010).</li>
<li Id="22"> Aebersold, R. & Mann, M. Mass spectrometry-based proteomics. Nature 422, 198–207; 10.1038/Nature01511 (2003).</li>
</ol>

                </div>
            </div>
        </div>
        
    </div>

    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <script type="text/javascript">var fsdocs_search_baseurl = '/BIO-BTE-06-L-7/'</script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/lunr.js/2.3.8/lunr.min.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.min.js"></script>
    <script type="text/javascript" src="/BIO-BTE-06-L-7/content/fsdocs-search.js"></script>
    <!-- END SEARCH BOX: this adds support for the search box -->
</body>

</html>