
        {
            "cells": [
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["# JP03 Mass Spectrometry Based Proteomics\n",
"\n",
"1. [Mass spectrometry (MS)-based proteomic](#Mass-spectrometry-(MS)-based-proteomic)\n",
"6. [References](#References)\n",
"\n",
"## Mass spectrometry (MS)-based proteomic\n",
"\u003ca href=\"#Mass-Spectrometry-Based-Proteomics\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Mass-Spectrometry-Based-Proteomics\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"\n",
"\u003cdiv Id= \"figure2\" Style=\"float: right; display: inline-block; color: #44546a; width: 60%; padding: 15px; max-width: 600px\"\u003e\n",
"\u003cdiv Id= \"figure2\" Style=\"float: right; display: inline-block; color: #44546a; width: 60%; padding: 15px; max-width: 600px\"\u003e\n",
"    \u003cimg src=\"img/OmicsWorkflow.png\" Style=\"width: 100%\"\u003e\n",
"    \u003cdiv Style=\"padding-left: 1rem; padding-right: 1rem; text-align: justify; font-size:0.8rem\"\u003e\n",
"        \u003cb\u003eFigure 2. Summary of a typical proteomics workflow following the bottom-up principle.\u003c/b\u003e During sample preparation proteins are extracted from the samples and digested into peptides using proteases typically trypsin. An optional fractionation or enrichment may be applied at either the protein or peptide level to enhance the scope of identification. Peptides are separated by high-performance liquid chromatography (HPLC) and afterward transferred into the vacuum of the mass spectrometer mostly using electrospray ionization (ESI). Cycles of full MS including all peptide at a time followed by MS/MS of selected peptides are measured. The consecutive measured MS and MS/MS spectra are then used to computationally identify and quantify the peptide sequence and infer the protein.\n",
"    \u003c/div\u003e\n",
"\u003c/div\u003e  \n",
"\n",
"The proteome is understood as an entire complement of proteins in one cell, tissue or a whole organism. \n",
"Proteomics as a scientific field deals with the qualitative and quantitative analysis of protein expression patterns. \n",
"Thus, proteomics relies primarily on the ability to unambiguously identify proteins, followed by accurate quantification. \n",
"Mass spectrometry-based proteomics refers to the large-scale analysis of\n",
"proteins using mass spectrometry, an analytical method to determine the mass of molecules. In the case of proteomics, the target \n",
"molecules are whole proteins or peptides.\n",
"\n",
"The typical proteomics workflow is as follows (\u003ca href=\"#figure2\"\u003eFigure 2\u003c/a\u003e): first, proteins are isolated from \n",
"cells or tissues by lysis followed by biochemical fractionation or affinity selection. MS on whole proteins (top-down proteomics) \n",
"is less sensitive and more difficult to handle when compared to MS on peptides (bottom-up proteomics), as the mass of the intact protein by \n",
"itself is insufficient for protein identification\u003csup\u003e\u003ca href=\"#18\"\u003e18\u003c/a\u003e,\u003ca href=\"#19\"\u003e19\u003c/a\u003e\u003c/sup\u003e. Therefore, the bottom-up approach is \n",
"standardly used, which comprises the enzymatic degradation of proteins to peptides using an endopeptidase (trypsin in most cases)\n",
"\u003csup\u003e\u003ca href=\"#20\"\u003e20\u003c/a\u003e\u003c/sup\u003e. Trypsin is advantageous as it generates peptides with C-terminally protonated amino acids to foster\n",
" the detectability of a full ion ladder (series) in a subsequent, optional fragmentation step for sequencing. The peptides are separated by one\n",
"  or more steps of liquid chromatography (LC) and afterwards transferred into the vacuum of the mass spectrometer, where a mass spectrum\n",
"   of the peptides eluting at this time from the LC is taken (MS\u003csup\u003e1\u003c/sup\u003e spectrum or ‘normal mass spectrum’). A prioritized list of these peptides\n",
"    for fragmentation is either automatically generated or provided by the operator. These peptide ions are then fragmented by energetic collision with\n",
"     an inert gas and recorded as tandem MS spectra (MS/MS or MS\u003csup\u003e2\u003c/sup\u003e spectra). The consecutive MS and MS/MS\n",
"      spectra are then used to computationally identify the peptide’s sequence and quantify its abundance\u003csup\u003e\u003ca href=\"#21\"\u003e22\u003c/a\u003e,\n",
"      \u003ca href=\"#22\"\u003e22\u003c/a\u003e\u003c/sup\u003e.\n",
"\n",
"\u003c/div\u003e\n",
"\u003c/div\u003e\n",
"\n",
"\u003chr\u003e\n",
"\u003chr\u003e\n",
"\n",
"\u003cnav class=\"level is-mobile\"\u003e\n",
"\u003cnav class=\"level is-mobile\"\u003e\n",
"    \u003cdiv class=\"level-left\"\u003e\n",
"        \u003cdiv class=\"level-item\"\u003e\n",
"            \u003cbutton class=\"button is-primary is-outlined\" onclick=\"location.href=\u0027/JP02_Plant_Systems_Biology.html\u0027;\"\u003e\u0026#171; JP02\u003c/button\u003e\n",
"        \u003c/div\u003e\n",
"    \u003c/div\u003e\n",
"    \u003cdiv class=\"level-right\"\u003e\n",
"        \u003cdiv class=\"level-item\"\u003e\n",
"            \u003cbutton class=\"button is-primary is-outlined\" onclick=\"location.href=\u0027/JP04_Digestion_and_mass_calculation.html\u0027;\"\u003eJP04 \u0026#187;\u003c/button\u003e\n",
"        \u003c/div\u003e\n",
"    \u003c/div\u003e\n",
"\u003c/nav\u003e\n",
"\n",
"## References\n",
"\u003ca href=\"#Mass-Spectrometry-Based-Proteomics\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Mass-Spectrometry-Based-Proteomics\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\u003col\u003e\n",
"\u003col\u003e\n",
"\u003cli Value=\"18\" Id=\"18\"\u003e Breuker, K., Jin, M., Han, X., Jiang, H. \u0026 McLafferty, F. W. Top-down identification and characterization of biomolecules by mass spectrometry. Journal of the American Society for Mass Spectrometry 19, 1045–1053; 10.1016/j.jasms.2008.05.013 (2008).\u003c/li\u003e\n",
"\n",
"\u003cli Id=\"19\"\u003e Reid, G. E. \u0026 McLuckey, S. A. \u0027Top down\u0027 protein characterization via tandem mass spectrometry. Journal of mass spectrometry : JMS 37, 663–675; 10.1002/jms.346 (2002).\u003c/li\u003e\n",
"\u003cli Id=\"19\"\u003e Reid, G. E. \u0026 McLuckey, S. A. \u0027Top down\u0027 protein characterization via tandem mass spectrometry. Journal of mass spectrometry : JMS 37, 663–675; 10.1002/jms.346 (2002).\u003c/li\u003e\n",
"\n",
"\u003cli Id=\"20\"\u003e Olsen, J. V., Ong, S.-E. \u0026 Mann, M. Trypsin cleaves exclusively C-terminal to arginine and lysine residues. Molecular \u0026 cellular proteomics : MCP 3, 608–614; 10.1074/mcp.T400003-MCP200 (2004).\u003c/li\u003e\n",
"\u003cli Id=\"20\"\u003e Olsen, J. V., Ong, S.-E. \u0026 Mann, M. Trypsin cleaves exclusively C-terminal to arginine and lysine residues. Molecular \u0026 cellular proteomics : MCP 3, 608–614; 10.1074/mcp.T400003-MCP200 (2004).\u003c/li\u003e\n",
"\n",
"\u003cli Id=\"21\"\u003e Walther, T. C. \u0026 Mann, M. Mass spectrometry-based proteomics in cell biology. J. Cell Biol. 190, 491–500; 10.1083/jcb.201004052 (2010).\u003c/li\u003e\n",
"\u003cli Id=\"21\"\u003e Walther, T. C. \u0026 Mann, M. Mass spectrometry-based proteomics in cell biology. J. Cell Biol. 190, 491–500; 10.1083/jcb.201004052 (2010).\u003c/li\u003e\n",
"\n",
"\u003cli Id=\"22\"\u003e Aebersold, R. \u0026 Mann, M. Mass spectrometry-based proteomics. Nature 422, 198–207; 10.1038/Nature01511 (2003).\u003c/li\u003e\n",
"\u003cli Id=\"22\"\u003e Aebersold, R. \u0026 Mann, M. Mass spectrometry-based proteomics. Nature 422, 198–207; 10.1038/Nature01511 (2003).\u003c/li\u003e\n",
"\u003c/ol\u003e\n",
"\n"]
          }],
            "metadata": {
            "kernelspec": {"display_name": ".NET (F#)", "language": "F#", "name": ".net-fsharp"},
            "langauge_info": {
        "file_extension": ".fs",
        "mimetype": "text/x-fsharp",
        "name": "C#",
        "pygments_lexer": "fsharp",
        "version": "4.5"
        }
        },
            "nbformat": 4,
            "nbformat_minor": 1
        }
        

