
        {
            "cells": [
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["# JP05 Isotopic Distribution\n",
"\n",
"1. [Isotopic Distribution](#Isotopic-Distribution)\n",
"    1. [Simulating Isotopic Clusters for peptides](#Simulating-Isotopic-Clusters-for-peptides)\n",
"    2. [Simulating Isotopic Clusters for peptides with stable isotope labeled variant](#Simulating-Isotopic-Clusters-for-peptides-with-stable-isotope-labeled-variant)\n",
"2. [References](#References)\n",
"\n",
"## Isotopic Distribution\n",
"\u003ca href=\"#Isotopic-Distribution\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Isotopic-Distribution\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"Peptide signals exhibit a characteristic shape in the mass spectrum that depend on their isotopic profile, which is defined by \n",
"the number of naturally occurring isotopes in the peptide. The occurrence probabilities of natural isotopes are reflected in the mass \n",
"spectrum by the relative heights of the peak series belonging to the respective peptide. The frequency at which natural isotopes occur \n",
"is known and can be used to compute the isotope distribution of a molecule. The isotopic distribution for a given peptide molecule \n",
"C\u003csub\u003ev\u003c/sub\u003eH\u003csub\u003ew\u003c/sub\u003eN\u003csub\u003ex\u003c/sub\u003eO\u003csub\u003ey\u003c/sub\u003eS\u003csub\u003ez\u003c/sub\u003e is described by the following product of polynomials:\n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cimg src=\"https://latex.codecogs.com/gif.latex?\\large\u0026space;\\newline(\u0026space;{}^{12}\\textrm{C}\u0026space;\u0026plus;\u0026space;{}^{13}\\textrm{C})^{v}\u0026space;\\times\u0026space;({}^{1}\\textrm{H}\u0026plus;{}^{2}\\textrm{H})^{w}\u0026space;\\times\u0026space;({}^{14}\\textrm{N}\u0026plus;{}^{15}\\textrm{N})^{x}\\times({}^{16}\\textrm{O}\u0026plus;{}^{17}\\textrm{O}\u0026space;\u0026plus;\u0026space;{}^{18}\\textrm{O})^{y}\\newline\\times({}^{32}\\textrm{S}\u0026plus;{}^{33}\\textrm{S}\u0026plus;{}^{34}\\textrm{S}\u0026plus;{}^{36}\\textrm{S})^{z}\" title=\"\\large \\newline( {}^{12}\\textrm{C} + {}^{13}\\textrm{C})^{v} \\times ({}^{1}\\textrm{H}+{}^{2}\\textrm{H})^{w} \\times ({}^{14}\\textrm{N}+{}^{15}\\textrm{N})^{x}\\times({}^{16}\\textrm{O}+{}^{17}\\textrm{O} + {}^{18}\\textrm{O})^{y}\\newline\\times({}^{32}\\textrm{S}+{}^{33}\\textrm{S}+{}^{34}\\textrm{S}+{}^{36}\\textrm{S})^{z}\" Style=\"margin: 1rem auto 1rem; display: block\"/\u003e\n",
"\u003c/div\u003e\n",
"\n",
"Symbolic expansion of the polynomials results in many product terms, which correspond to different isotopic variants of a molecule. \n",
"Even for molecules of a medium size, the straightforward expansion of the polynomials leads to an explosion regarding the number of product terms. \n",
"Due to this complexity, there was a need to develop algorithms for efficient computation. The different strategies comprise pruning the \n",
"polynomials to discard terms with coefficients below a threshold\u003csup\u003e\u003ca href=\"#25\"\u003e25\u003c/a\u003e\u003c/sup\u003e combined with a recursive \n",
"computation\u003csup\u003e\u003ca href=\"#26\"\u003e26\u003c/a\u003e\u003c/sup\u003e, and Fourier Transformation for a more efficient convolution of the isotope distributions of \n",
"individual elements\u003csup\u003e\u003ca href=\"#27\"\u003e27\u003c/a\u003e\u003c/sup\u003e, or rely on dynamic programming\u003csup\u003e\u003ca href=\"#28\"\u003e28\u003c/a\u003e\u003c/sup\u003e.\n",
"\n",
"\u003cdiv Style=\"text-align: justify ; margin-top: 2rem ; margin-bottom: 2rem ; line-height: 1.3 ; width: 85% ; margin-left: auto ; margin-right: auto ; padding: 10px ; border: 2px dotted #708090 ; color: #708090\"\u003e\n",
"\u003cdiv Style=\"text-align: justify ; margin-top: 2rem ; margin-bottom: 2rem ; line-height: 1.3 ; width: 85% ; margin-left: auto ; margin-right: auto ; padding: 10px ; border: 2px dotted #708090 ; color: #708090\"\u003e\n",
"MIDAs\u003csup\u003e\u003ca href=\"#29\"\u003e29\u003c/a\u003e\u003c/sup\u003e is one of the more elaborate algorithms to predict an isotope cluster based on a given peptide sequence. \n",
"Simulate the isotopic cluster of the peptide sequence ‘PEPTIDES’ and ‘PEPTIDEPEPTIDEPEPTIDEPEPTIDES’ with natural occurring isotope abundances.\n",
"\u003c/div\u003e\n",
"\u003c/div\u003e\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 1, "outputs": [], 
           "source": ["#r \"nuget: BioFSharp, 2.0.0-beta5\"\n",
"#r \"nuget: BioFSharp.IO, 2.0.0-beta5\"\n",
"#r \"nuget: Plotly.NET, 2.0.0-beta6\"\n",
"\n",
"#r \"nuget: Plotly.NET, 2.0.0-beta6\"\n",
"#r \"nuget: Plotly.NET.Interactive, 2.0.0-beta6\"\n",
"\n",
"open Plotly.NET\n",
"open BioFSharp\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["## Simulating Isotopic Clusters for peptides\n",
"\u003ca href=\"#Isotopic-Distribution\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Isotopic-Distribution\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"We will use two artificial peptide sequences and translate them into their elemental composition to simulate their isotopic clusters. \n",
"Therefore, we first define a function that maps from a peptide sequence to its formula: \n",
"\u003c/div\u003e\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 2, "outputs": [], 
           "source": ["// Code-Block 1\n",
"\n",
"// create chemical formula for amino acid and add water to reflect hydrolysed state in mass spectrometer\n",
"let toFormula bioseq =  \n",
"    bioseq\n",
"    |\u003e BioSeq.toFormula\n",
"    // peptides are hydrolysed in the mass spectrometer, so we add H2O\n",
"    |\u003e Formula.add Formula.Table.H2O\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["Next, we will apply our function to receive the elemental composition or chemical formula of the peptides.\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 3, "outputs": [
          {
           "data": {
            "text/plain": ["\"C37.00 H58.00 N8.00 O17.00\""]
        },
           "execution_count": 3,
           "metadata": {},
           "output_type": "execute_result"
          },

          {
           "data": {
            "text/plain": ["\"C139.00 H211.00 N29.00 O59.00\""]
        },
           "execution_count": 3,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// Code-Block 2\n",
"\n",
"// translate single letter code into amino acids and create chemical formula of it.\n",
"let peptide_short = \n",
"    \"PEPTIDES\" \n",
"    |\u003e BioSeq.ofAminoAcidString\n",
"    |\u003e toFormula\n",
"    \n",
"let peptide_long  = \n",
"    \"PEPTIDEPEPTIDEPEPTIDEPEPTIDES\" \n",
"    |\u003e BioSeq.ofAminoAcidString\n",
"    |\u003e toFormula\n",
"    \n",
"let peptide_shortString =\n",
"    peptide_short \n",
"    |\u003e Formula.toString\n",
"\n",
"\n",
"let peptide_longString =\n",
"    peptide_long \n",
"    |\u003e Formula.toString\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"Additionally, we need a function that maps from Formula (and charge) to the isotopic distribution. Here, we \n",
"can use \u003ccode\u003eIsotopicDistribution.MIDA.ofFormula\u003c/code\u003e from the BioFSharp library. However, for convenience \n",
"(to use the same parameter twice), we define our function \u003ccode\u003egenerateIsotopicDistribution\u003c/code\u003e:\n",
"\u003c/div\u003e\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 6, "outputs": [
          {
           "data": {
            "text/plain": ["[(3231.447065, 0.6079646485); (3232.450057, 1.0); (3233.452986, 0.9085467193);",
" (3234.455919, 0.5658743567); (3235.458395, 0.2618251568);",
" (3236.462592, 0.08134469002)]"]
        },
           "execution_count": 6,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// Code-Block 3\n",
"\n",
"// Predicts an isotopic distribution of the given formula at the given charge, \n",
"// normalized by the sum of probabilities, using the MIDAs algorithm\n",
"let generateIsotopicDistribution (charge:int) (f:Formula.Formula) =\n",
"    IsotopicDistribution.MIDA.ofFormula \n",
"        IsotopicDistribution.MIDA.normalizeByMaxProb\n",
"        0.01\n",
"        0.005\n",
"        charge\n",
"        f\n",
"        \n",
"// create pattern for peptide_short\n",
"let isoPattern_peptide_short = \n",
"    generateIsotopicDistribution 1 peptide_short\n",
"\n",
"// create pattern for peptide_long\n",
"let isoPattern_peptide_long = \n",
"    generateIsotopicDistribution 1 peptide_long\n",
"    \n",
"isoPattern_peptide_long\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 7, "outputs": [
          {
           "data": {
            "text/html": ["\u003cdiv id=\"1cf7fe0e-32d5-4763-9b6f-68a65c62cfe5\" style=\"width: 900px; height: 600px;\"\u003e\u003c!-- Plotly chart will be drawn inside this DIV --\u003e\u003c/div\u003e",
"\u003cscript type=\"text/javascript\"\u003e",
"",
"            var renderPlotly_1cf7fe0e32d547639b6f68a65c62cfe5 = function() {",
"            var fsharpPlotlyRequire = requirejs.config({context:\u0027fsharp-plotly\u0027,paths:{plotly:\u0027https://cdn.plot.ly/plotly-latest.min\u0027}}) || require;",
"            fsharpPlotlyRequire([\u0027plotly\u0027], function(Plotly) {",
"",
"            var data = [{\"type\":\"bar\",\"x\":[887.399558199034,888.4024872665019,889.4049942776403,890.4081884772683],\"y\":[1.0,0.43236774529511174,0.12578247843311785,0.024134205673036688],\"marker\":{},\"name\":\"peptide_short\",\"xaxis\":\"x\",\"yaxis\":\"y\"},{\"type\":\"bar\",\"x\":[3231.4470647196554,3232.4500572584698,3233.452986075872,3234.4559191540584,3235.4583946426615,3236.4625915315855],\"y\":[0.6079646485424579,1.0,0.9085467193387605,0.5658743566632869,0.26182515682174157,0.08134469002336972],\"marker\":{},\"name\":\"peptide_long\",\"xaxis\":\"x2\",\"yaxis\":\"y2\"}];",
"            var layout = {\"xaxis\":{\"title\":\"m/z\",\"range\":[885.0,895.0],\"anchor\":\"y\",\"domain\":[0.0,0.45]},\"yaxis\":{\"anchor\":\"x\",\"domain\":[0.050000000000000044,1.0],\"title\":\"intensity\"},\"xaxis2\":{\"title\":\"m/z\",\"range\":[3230.0,3240.0],\"anchor\":\"y2\",\"domain\":[0.5,0.95]},\"yaxis2\":{\"anchor\":\"x2\",\"domain\":[0.050000000000000044,1.0]},\"width\":900.0,\"height\":600.0,\"title\":\"Isotopeclusters\"};",
"            var config = {};",
"            Plotly.newPlot(\u00271cf7fe0e-32d5-4763-9b6f-68a65c62cfe5\u0027, data, layout, config);",
"});",
"            };",
"            if ((typeof(requirejs) !==  typeof(Function)) || (typeof(requirejs.config) !== typeof(Function))) {",
"                var script = document.createElement(\"script\");",
"                script.setAttribute(\"src\", \"https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js\");",
"                script.onload = function(){",
"                    renderPlotly_1cf7fe0e32d547639b6f68a65c62cfe5();",
"                };",
"                document.getElementsByTagName(\"head\")[0].appendChild(script);",
"            }",
"            else {",
"                renderPlotly_1cf7fe0e32d547639b6f68a65c62cfe5();",
"            }",
"\u003c/script\u003e",
""]
        },
           "execution_count": 7,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// Code-Block 4\n",
"\n",
"// create one chart for both, short and long peptide isotopic patterns.     \n",
"let isoPatternChart = \n",
"    [\n",
"        Chart.Column(isoPattern_peptide_short,Name= \"peptide_short\" )\n",
"        |\u003e Chart.withX_AxisStyle (\"m/z\",MinMax=(885.,895.))\n",
"        Chart.Column(isoPattern_peptide_long,Name= \"peptide_long\" )\n",
"        |\u003e Chart.withX_AxisStyle (\"m/z\",MinMax=(3230., 3240.))\n",
"    ]\n",
"    |\u003e Chart.Stack 2\n",
"    |\u003e Chart.withSize (900.,600.)\n",
"    |\u003e Chart.withTitle \"Isotopeclusters\"\n",
"    |\u003e Chart.withY_AxisStyle \"intensity\"\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["## Simulating Isotopic Clusters for peptides with stable isotope labeled variant\n",
"\u003ca href=\"#Isotopic-Distribution\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Isotopic-Distribution\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\u003cdiv class=\"container\"\u003e\n",
"\u003cdiv class=\"container\"\u003e\n",
"In addition to the natural occurring isotopic distribution, the field of proteomics has benefited greatly from the ability to \n",
"introduce stable isotopes into peptide sequences. So called isotopic labeling refers to the introduction of a naturally low-abundance \n",
"isotope of carbon, nitrogen, hydrogen and, in some cases, oxygen, into a peptide sequence. The isotopes commonly used are \u003csup\u003e13\u003c/sup\u003eC, \n",
"\u003csup\u003e15\u003c/sup\u003eN, \u003csup\u003e2\u003c/sup\u003eH (deuterium) and \u003csup\u003e18\u003c/sup\u003eO with natural abundances of 1.10%, 0.366%, 0.015% and 0.200%, \n",
"respectively\u003csup\u003e\u003ca href=\"#30\"\u003e30\u003c/a\u003e\u003c/sup\u003e. Therefore, the introduction of these isotopes into a peptide sequence can be detected by \n",
"most modern mass spectrometers leading to a respective mass shift and the ability to separate the same peptide species within the same run.\n",
" \n",
"\u003cdiv Style=\"text-align: justify ; margin-top: 2rem ; margin-bottom: 2rem ; line-height: 1.3 ; width: 85% ; margin-left: auto ; margin-right: auto ; padding: 10px ; border: 2px dotted #708090 ; color: #708090\"\u003e\n",
"\u003cdiv Style=\"text-align: justify ; margin-top: 2rem ; margin-bottom: 2rem ; line-height: 1.3 ; width: 85% ; margin-left: auto ; margin-right: auto ; padding: 10px ; border: 2px dotted #708090 ; color: #708090\"\u003e\n",
"MIDAs\u003csup\u003e\u003ca href=\"#29\"\u003e29\u003c/a\u003e\u003c/sup\u003e is also able to predict isotope clusters with altered isotope abundances. Simulate the isotopic cluster \n",
"of the peptide sequence ‘PEPTIDES’ and ‘PEPTIDEPEPTIDEPEPTIDEPEPTIDES’ with stable isotopes \u003csup\u003e15\u003c/sup\u003eN labeling. \n",
"\u003c/div\u003e\n",
"\n",
"Therefore, we define a function called \u003ccode\u003elabel\u003c/code\u003e. The function maps from a formula to a formula with exchangen nitrogen isotopes. \n",
"(Attention: Don\u0027t get confused a formula is just a FSharpMap.) \n",
"\n",
"\u003c/div\u003e\n",
"\u003c/div\u003e\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 9, "outputs": [], 
           "source": ["// Code-Block 5\n",
"\n",
"/// returns a function that replaces the nitrogen atoms in a formula\n",
"/// with the 15N isotope\n",
"let label formula =\n",
"    Formula.replaceElement formula Elements.Table.N Elements.Table.Heavy.N15\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 10, "outputs": [
          {
           "data": {
            "text/plain": ["map",
"  [(Di { Symbol = \"C\"",
"         X = { AtomicSymbol = \"C\"",
"               AtomicNumberZ = 6",
"               MassNumber = 12",
"               Mass = 12.0",
"               NatAbundance = 0.98938",
"               RelAtomicMass = 12.01078 }",
"         Xcomp = 0.98938",
"         X1 = { AtomicSymbol = \"C\"",
"                AtomicNumberZ = 6",
"                MassNumber = 13",
"                Mass = 13.00335484",
"                NatAbundance = 0.01078",
"                RelAtomicMass = 12.01078 }",
"         X1comp = 0.01078",
"         Root = -91.77922078 }, 37.0); (Di { Symbol = \"H\"",
"                                             X = { AtomicSymbol = \"H\"",
"                                                   AtomicNumberZ = 1",
"                                                   MassNumber = 1",
"                                                   Mass = 1.007825032",
"                                                   NatAbundance = 0.999885",
"                                                   RelAtomicMass = 1.007947 }",
"                                             Xcomp = 0.999885",
"                                             X1 = { AtomicSymbol = \"D\"",
"                                                    AtomicNumberZ = 1",
"                                                    MassNumber = 2",
"                                                    Mass = 2.014101778",
"                                                    NatAbundance = 0.000115",
"                                                    RelAtomicMass = 1.007947 }",
"                                             X1comp = 0.000115",
"                                             Root = -8694.652174 }, 58.0);",
"   (Di { Symbol = \"N15\"",
"         X = { AtomicSymbol = \"N\"",
"               AtomicNumberZ = 7",
"               MassNumber = 15",
"               Mass = 15.0001089",
"               NatAbundance = 0.00364",
"               RelAtomicMass = 14.0067 }",
"         Xcomp = 0.99636",
"         X1 = { AtomicSymbol = \"N\"",
"                AtomicNumberZ = 7",
"                MassNumber = 14",
"                Mass = 14.003074",
"                NatAbundance = 0.99636",
"                RelAtomicMass = 14.0067 }",
"         X1comp = 0.00364",
"         Root = -273.7252747 }, 8.0);",
"   (Tri",
"      { Symbol = \"O\"",
"        X = { AtomicSymbol = \"O\"",
"              AtomicNumberZ = 8",
"              MassNumber = 16",
"              Mass = 15.99491462",
"              NatAbundance = 0.99757",
"              RelAtomicMass = 15.9994 }",
"        Xcomp = 0.99757",
"        X1 = { AtomicSymbol = \"O\"",
"               AtomicNumberZ = 8",
"               MassNumber = 17",
"               Mass = 16.9991317",
"               NatAbundance = 0.00038",
"               RelAtomicMass = 15.9994 }",
"        X1comp = 0.00038",
"        X2 = { AtomicSymbol = \"O\"",
"               AtomicNumberZ = 8",
"               MassNumber = 18",
"               Mass = 17.9991611",
"               NatAbundance = 0.00205",
"               RelAtomicMass = 15.9994 }",
"        X2comp = 0.00205",
"        Root =",
"              ((-0.09268292682926829, 22.05925932732548),",
"               (-0.09268292682926829, -22.05925932732548)) }, 17.0)]"]
        },
           "execution_count": 10,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// Code-Block 6\n",
"\n",
"let N15_peptide_short = \n",
"    \"PEPTIDES\" \n",
"    |\u003e BioSeq.ofAminoAcidString\n",
"    |\u003e toFormula\n",
"    |\u003e label\n",
"\n",
"let N15_peptide_long  = \n",
"    \"PEPTIDEPEPTIDEPEPTIDEPEPTIDES\" \n",
"    |\u003e BioSeq.ofAminoAcidString\n",
"    |\u003e toFormula\n",
"    |\u003e label\n",
"\n",
"//result: N15_peptide_short\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 12, "outputs": [
          {
           "data": {
            "text/plain": ["map",
"  [(Di { Symbol = \"C\"",
"         X = { AtomicSymbol = \"C\"",
"               AtomicNumberZ = 6",
"               MassNumber = 12",
"               Mass = 12.0",
"               NatAbundance = 0.98938",
"               RelAtomicMass = 12.01078 }",
"         Xcomp = 0.98938",
"         X1 = { AtomicSymbol = \"C\"",
"                AtomicNumberZ = 6",
"                MassNumber = 13",
"                Mass = 13.00335484",
"                NatAbundance = 0.01078",
"                RelAtomicMass = 12.01078 }",
"         X1comp = 0.01078",
"         Root = -91.77922078 }, 139.0); (Di { Symbol = \"H\"",
"                                              X = { AtomicSymbol = \"H\"",
"                                                    AtomicNumberZ = 1",
"                                                    MassNumber = 1",
"                                                    Mass = 1.007825032",
"                                                    NatAbundance = 0.999885",
"                                                    RelAtomicMass = 1.007947 }",
"                                              Xcomp = 0.999885",
"                                              X1 = { AtomicSymbol = \"D\"",
"                                                     AtomicNumberZ = 1",
"                                                     MassNumber = 2",
"                                                     Mass = 2.014101778",
"                                                     NatAbundance = 0.000115",
"                                                     RelAtomicMass = 1.007947 }",
"                                              X1comp = 0.000115",
"                                              Root = -8694.652174 }, 211.0);",
"   (Di { Symbol = \"N15\"",
"         X = { AtomicSymbol = \"N\"",
"               AtomicNumberZ = 7",
"               MassNumber = 15",
"               Mass = 15.0001089",
"               NatAbundance = 0.00364",
"               RelAtomicMass = 14.0067 }",
"         Xcomp = 0.99636",
"         X1 = { AtomicSymbol = \"N\"",
"                AtomicNumberZ = 7",
"                MassNumber = 14",
"                Mass = 14.003074",
"                NatAbundance = 0.99636",
"                RelAtomicMass = 14.0067 }",
"         X1comp = 0.00364",
"         Root = -273.7252747 }, 29.0);",
"   (Tri",
"      { Symbol = \"O\"",
"        X = { AtomicSymbol = \"O\"",
"              AtomicNumberZ = 8",
"              MassNumber = 16",
"              Mass = 15.99491462",
"              NatAbundance = 0.99757",
"              RelAtomicMass = 15.9994 }",
"        Xcomp = 0.99757",
"        X1 = { AtomicSymbol = \"O\"",
"               AtomicNumberZ = 8",
"               MassNumber = 17",
"               Mass = 16.9991317",
"               NatAbundance = 0.00038",
"               RelAtomicMass = 15.9994 }",
"        X1comp = 0.00038",
"        X2 = { AtomicSymbol = \"O\"",
"               AtomicNumberZ = 8",
"               MassNumber = 18",
"               Mass = 17.9991611",
"               NatAbundance = 0.00205",
"               RelAtomicMass = 15.9994 }",
"        X2comp = 0.00205",
"        Root =",
"              ((-0.09268292682926829, 22.05925932732548),",
"               (-0.09268292682926829, -22.05925932732548)) }, 59.0)]"]
        },
           "execution_count": 12,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["//result: N15_peptide_long\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 14, "outputs": [
          {
           "data": {
            "text/plain": ["[(894.3788366, 0.02888603821); (895.3759499, 1.0); (896.3792324, 0.3984467169);",
" (897.38185, 0.1130793873); (898.3845066, 0.02385315906)]"]
        },
           "execution_count": 14,
           "metadata": {},
           "output_type": "execute_result"
          },

          {
           "data": {
            "text/plain": ["[(3259.364178, 0.06384122647); (3260.36211, 0.7008291559); (3261.365076, 1.0);",
" (3262.368107, 0.8402034739); (3263.370813, 0.4705485682);",
" (3264.373788, 0.2169537728); (3265.376745, 0.08062486598)]"]
        },
           "execution_count": 14,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// Code-Block 7\n",
"\n",
"// create pattern for N15_peptide_short\n",
"let N15_isoPattern_peptide_short = \n",
"    generateIsotopicDistribution 1 N15_peptide_short\n",
"\n",
"// create pattern for N15_peptide_long\n",
"let N15_isoPattern_peptid_long = \n",
"    generateIsotopicDistribution 1 N15_peptide_long\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 17, "outputs": [
          {
           "data": {
            "text/html": ["\u003cdiv id=\"7e5e7ddc-c78e-4612-ad0c-ccb3e614d253\" style=\"width: 600px; height: 600px;\"\u003e\u003c!-- Plotly chart will be drawn inside this DIV --\u003e\u003c/div\u003e",
"\u003cscript type=\"text/javascript\"\u003e",
"",
"            var renderPlotly_7e5e7ddcc78e4612ad0cccb3e614d253 = function() {",
"            var fsharpPlotlyRequire = requirejs.config({context:\u0027fsharp-plotly\u0027,paths:{plotly:\u0027https://cdn.plot.ly/plotly-latest.min\u0027}}) || require;",
"            fsharpPlotlyRequire([\u0027plotly\u0027], function(Plotly) {",
"",
"            var data = [{\"type\":\"bar\",\"x\":[887.399558199034,888.4024872665019,889.4049942776403,890.4081884772683],\"y\":[1.0,0.43236774529511174,0.12578247843311785,0.024134205673036688],\"marker\":{},\"name\":\"peptide_short\",\"xaxis\":\"x\",\"yaxis\":\"y\"},{\"type\":\"bar\",\"x\":[894.3788365975154,895.3759499224518,896.3792323902378,897.3818499569743,898.3845066469612],\"y\":[0.0288860382140478,1.0,0.3984467169061343,0.11307938727778763,0.023853159064528624],\"marker\":{},\"name\":\"N15_peptide_short\",\"xaxis\":\"x\",\"yaxis\":\"y\"},{\"type\":\"bar\",\"x\":[3231.4470647196554,3232.4500572584698,3233.452986075872,3234.4559191540584,3235.4583946426615,3236.4625915315855],\"y\":[0.6079646485424579,1.0,0.9085467193387605,0.5658743566632869,0.26182515682174157,0.08134469002336972],\"marker\":{},\"name\":\"peptide_long\",\"xaxis\":\"x2\",\"yaxis\":\"y2\"},{\"type\":\"bar\",\"x\":[3259.3641783135813,3260.362110012513,3261.365076207296,3262.3681072195172,3263.370812501318,3264.3737877709027,3265.3767448967224],\"y\":[0.06384122646754022,0.7008291559420009,1.0,0.8402034738792131,0.4705485682431846,0.21695377275354297,0.08062486597973496],\"marker\":{},\"name\":\"N15_peptide_long\",\"xaxis\":\"x2\",\"yaxis\":\"y2\"}];",
"            var layout = {\"xaxis\":{\"title\":\"m/z\",\"range\":[885.0,905.0],\"anchor\":\"y\",\"domain\":[0.0,0.45]},\"yaxis\":{\"anchor\":\"x\",\"domain\":[0.050000000000000044,1.0],\"title\":\"intensity\"},\"xaxis2\":{\"title\":\"m/z\",\"range\":[3230.0,3270.0],\"anchor\":\"y2\",\"domain\":[0.5,0.95]},\"yaxis2\":{\"anchor\":\"x2\",\"domain\":[0.050000000000000044,1.0]},\"title\":\"Isotopeclusters\"};",
"            var config = {};",
"            Plotly.newPlot(\u00277e5e7ddc-c78e-4612-ad0c-ccb3e614d253\u0027, data, layout, config);",
"});",
"            };",
"            if ((typeof(requirejs) !==  typeof(Function)) || (typeof(requirejs.config) !== typeof(Function))) {",
"                var script = document.createElement(\"script\");",
"                script.setAttribute(\"src\", \"https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js\");",
"                script.onload = function(){",
"                    renderPlotly_7e5e7ddcc78e4612ad0cccb3e614d253();",
"                };",
"                document.getElementsByTagName(\"head\")[0].appendChild(script);",
"            }",
"            else {",
"                renderPlotly_7e5e7ddcc78e4612ad0cccb3e614d253();",
"            }",
"\u003c/script\u003e",
""]
        },
           "execution_count": 17,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// Code-Block 8\n",
"\n",
"// Create two charts. Each with the related N14 and N15 isotopic clusters. Then stack them two one unit.\n",
"let isoPatternChart2 = \n",
"    [\n",
"        [\n",
"            Chart.Column(isoPattern_peptide_short,Name= \"peptide_short\" )\n",
"            Chart.Column(N15_isoPattern_peptide_short,Name= \"N15_peptide_short\" )\n",
"        ] \n",
"        |\u003e Chart.Combine \n",
"        |\u003e Chart.withX_AxisStyle (\"m/z\",MinMax=(885., 905.0))\n",
"\n",
"        [\n",
"            Chart.Column(isoPattern_peptide_long,Name= \"peptide_long\" )\n",
"            Chart.Column(N15_isoPattern_peptid_long,Name= \"N15_peptide_long\" )            \n",
"        ] \n",
"        |\u003e Chart.Combine \n",
"        |\u003e Chart.withX_AxisStyle (\"m/z\",MinMax=(3230.0, 3270.0))\n",
"    ]\n",
"    |\u003e Chart.Stack 2\n",
"    |\u003e Chart.withTitle \"Isotopeclusters\"\n",
"    |\u003e Chart.withY_AxisStyle \"intensity\"\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["\u003chr\u003e\n",
"\u003chr\u003e\n",
"\u003cnav class=\"level is-mobile\"\u003e\n",
"    \u003cdiv class=\"level-left\"\u003e\n",
"        \u003cdiv class=\"level-item\"\u003e\n",
"            \u003cbutton class=\"button is-primary is-outlined\" onclick=\"location.href=\u0027/JP04_Digestion_and_mass_calculation.html\u0027;\"\u003e\u0026#171; JP04\u003c/button\u003e\n",
"        \u003c/div\u003e\n",
"    \u003c/div\u003e\n",
"    \u003cdiv class=\"level-right\"\u003e\n",
"        \u003cdiv class=\"level-item\"\u003e\n",
"            \u003cbutton class=\"button is-primary is-outlined\" onclick=\"location.href=\u0027/JP06_Retention_time_and_scan_time\u0027;\"\u003eJP06 \u0026#187;\u003c/button\u003e\n",
"        \u003c/div\u003e\n",
"    \u003c/div\u003e\n",
"\u003c/nav\u003e\n",
"\n",
"## References\n",
"\u003ca href=\"#Isotopic-Distribution\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\u003ca href=\"#Isotopic-Distribution\" style=\"display: inline-block\"\u003e\u003csup\u003e\u0026#8593;back\u003c/sup\u003e\u003c/a\u003e\u003cbr\u003e\n",
"\n",
"\u003col\u003e\n",
"\u003col\u003e\n",
"\u003cli Value=\"25\" Id=\"25\"\u003e Yergey, J. A. A General-Approach to Calculating Isotopic Distributions for Mass-Spectrometry. Int J Mass Spectrom 52, 337–349; 10.1016/0020-7381(83)85053-0 (1983).\u003c/li\u003e\n",
"\u003cli Id=\"26\"\u003e Claesen, J., Dittwald, P., Burzykowski, T. \u0026 Valkenborg, D. An efficient method to calculate the aggregated isotopic distribution and exact center-masses. Journal of the American Society for Mass Spectrometry 23, 753–763; 10.1007/s13361-011-0326-2 (2012).\u003c/li\u003e\n",
"\u003cli Id=\"27\"\u003e Rockwood, A. L., Vanorden, S. L. \u0026 Smith, R. D. Rapid Calculation of Isotope Distributions. Anal Chem 67, 2699–2704; 10.1021/Ac00111a031 (1995).\u003c/li\u003e\n",
"\u003cli Id=\"28\"\u003e Snider, R. K. Efficient calculation of exact mass isotopic distributions. Journal of the American Society for Mass Spectrometry 18, 1511–1515; 10.1016/j.jasms.2007.05.016 (2007).\u003c/li\u003e\n",
"\u003cli Id=\"29\"\u003e Alves, G. \u0026 Yu, Y. K. Robust accurate identification of peptides (RAId). deciphering MS2 data using a structured library search with de novo based statistics. Bioinformatics 21, 3726–3732; 10.1093/bioinformatics/bti620 (2005).\u003c/li\u003e\n",
"\u003cli Id=\"30\"\u003e Becker, G. W. Stable isotopic labeling of proteins for quantitative proteomic applications. Brief Funct Genomic Proteomic 7, 371–382; 10.1093/bfgp/eln047 (2008).\u003c/li\u003e\n",
"\u003c/ol\u003e\n",
"\n"]
          }],
            "metadata": {
            "kernelspec": {"display_name": ".NET (F#)", "language": "F#", "name": ".net-fsharp"},
            "langauge_info": {
        "file_extension": ".fs",
        "mimetype": "text/x-fsharp",
        "name": "C#",
        "pygments_lexer": "fsharp",
        "version": "4.5"
        }
        },
            "nbformat": 4,
            "nbformat_minor": 1
        }
        

