<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <title>JP10 Peptide Identification
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="author" content="">
    <link rel="shortcut icon" type="image/x-icon" href="/BIO-BTE-06-L-7//img/favicon.ico">
    <script src="https://code.jquery.com/jquery-3.4.1.min.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <!-- <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/MathJax.js?config=TeX-MML-AM_CHTML"></script> -->

    <link type="text/css" rel="stylesheet" href="/BIO-BTE-06-L-7/content/fsdocs-default.css" />
    <link type="text/css" rel="stylesheet" href="/BIO-BTE-06-L-7/content/fsdocs-custom.css" />
    <script type="text/javascript" src="/BIO-BTE-06-L-7/content/fsdocs-tips.js"></script>
    
    <link type="text/css" rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" />
    <script defer type="text/javascript" src="https://kit.fontawesome.com/0d3e0ea7a6.js"></script>
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <!-- END SEARCH BOX: this adds support for the search box -->
    
</head> 

<body>
    <div class="columns is-fullheight m-0">
        <div id="fsdocs-menu" class="column is-2 is-paddingless">
            <aside class="menu">
                <div class="container is-paddingless is-marginless has-background-dark">
                    <a href="/BIO-BTE-06-L-7/">
                        <figure class="image is-128x128 container">
                            <img src="/BIO-BTE-06-L-7/img/logo.png">
                        </figure>
                    </a>
                    <h2 class="title has-text-centered has-text-white pb-4">BIO-BTE-06-L-7</h2>
                </div>
                <div class="content">
                    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
                    <div id="header">
                        <div class="field">
                            <div class="control has-icons-left has-icons-right searchbox" id="fsdocs-searchbox">
                                <input data-search-input="" id="search-by" type="search" placeholder="Search..." class="input"/>
                                <span class="icon is-small is-left">
                                    <i class="fas fa-search"></i>
                                </span>
                            </div>
                        </div>
                    </div>

                    <!-- END SEARCH BOX: this adds support for the search box -->
                    <ul class="navbar-nav">
                        <li class="nav-header">
  Documentation
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP06_Retention_time_and_scan_time.html">
    JP06 Retention time and scan time

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP09_Fragmentation_for_peptide_identification.html">
    JP09 Fragmentation for peptide identification

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP08_Centroidisation.html">
    JP08 Centroidisation

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP05_Isotopic_distribution.html">
    JP05 Isotopic Distribution

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP00_Systems_Biology_FSharp_Introduction.html">
    JP00 Systems Biology

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP11_Quantification.html">
    JP11_Quantification
  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP10_Peptide_Identification.html">
    JP10 Peptide Identification

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP04_Digestion_and_mass_calculation.html">
    JP04 Digestion and mass calculation

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP01_FSharpExcercises.html">
    JP01 FSharp Introduction Excercises

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP07_Signal_detection_and_quantification.html">
    JP07 Signal detection and quantification

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP02_Plant_Systems_Biology.html">
    JP02 Plant Systems Biology

  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-06-L-7/JP03_Mass_spectrometry_based_proteomics.html">
    JP03 Mass Spectrometry Based Proteomics

  </a>
</li>
                        <li class="nav-header">Links</li>
                        <li class="nav-item" >
                            <a class="nav-link" href="https://csb.bio.uni-kl.de/teaching/">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fas fa-graduation-cap"></i>
                                    </span>
                                    <span>Other CSB courses</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item" >
                            <a class="nav-link" href="https://twitter.com/cs_biology">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fab fa-twitter"></i>
                                    </span>
                                    <span>Get in touch on twitter</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item" >
                            <a class="nav-link" href="https://github.com/CSBiology?type=source">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fab fa-github"></i>
                                    </span>
                                    <span>Check out our open source projects</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://github.com/CSBiology/BIO-BTE-06-L-7">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fas fa-code-branch"></i>
                                    </span>
                                    <span>Source repository</span>
                                </span>
                            </a>
                        </li>
                    </ul>
                </div>
            </aside>
        </div>
        <div class="column is-10 is-paddingless has-text-justified	">
            <div class="section pt-4">
                <div class="container" id="fsdocs-content">
                    <h1><a name="JP10-Peptide-Identification" class="anchor" href="#JP10-Peptide-Identification">JP10 Peptide Identification</a></h1>
<p><a href="https://mybinder.org/v2/gh/CSBiology/BIO-BTE-06-L-7/gh-pages?filepath=JP10_Peptide_Identification.ipynb"><img src="https://mybinder.org/badge_logo.svg" alt="Binder" /></a></p>
<ol>
<li><a href="#Understanding-peptide-identification-from-MS2-spectra">Understanding peptide identification from MS<sup>2</sup> spectra</a></li>
<li>
<a href="#Matching-and-scoring-of-Tandem-MS-peptide-identification">Matching and scoring of Tandem MS peptide identification</a>
<ol>
<li><a href="#Step-1:-Data-acquisition-and-preprocessing">Step 1: Data acquisition and preprocessing</a></li>
<li><a href="#Step-2:-Preselecting-the-peptides-of-interest">Step 2: Preselecting the peptides of interest</a></li>
<li><a href="#Step-3+4:-Matching-and-Scoring">Step 3+4: Matching and Scoring</a></li>
</ol>
</li>
</ol>
<h2><a name="Understanding-peptide-identification-from-MS-sup-2-sup-spectra" class="anchor" href="#Understanding-peptide-identification-from-MS-sup-2-sup-spectra">Understanding peptide identification from MS<sup>2</sup> spectra</a></h2>
<a href="#Peptide-Identification" style="display: inline-block"><sup>&#8593;back</sup></a><br>
<div class="container">
Under low-energy fragmentation conditions, peptide fragment patterns are reproducible and, in general, predictable, which allows for an 
amino acid sequence identification according to a fragmentation expectation. Algorithms for peptide identification perform in principle 
three basic tasks:
<b>(i)</b> a raw data preprocessing step is applied to the MS/MS spectra to obtain clean peak information. The same signal filtering 
and background subtraction methods are used as discussed in the section of low-level processing. Peak detection, however, may be performed 
differently. Preprocessing of MS/MS spectra focuses on the extraction of the precise m/z of the peak rather than the accurate peak areas. 
The conversion of a peak profile into the corresponding m/z and intensity values reduces the complexity, its representation is termed centroiding. 
To extract the masses for identification in a simple and fast way, peak fitting approaches are used. These approaches take either the most intense 
point of the peak profile, fit a Lorentzian distribution to the profile, or use a quadratic fit<sup><a href="#37">37</a></sup>. 
<b>(ii)</b> Spectrum information and possible amino acid sequences assignments are evaluated. 
<b>(iii)</b> The quality of the match between spectrum and possible sequences is scored.
<p>Even though the three steps roughly describe the basic principle of algorithms used for peptide sequence identification, most implementations
show differences in individual steps which can lead to major changes in the outcome. Therefore, it has been proven useful to utilize more than
one algorithm for a robust and thorough identification. Due to their major difference in identification strategies and prerequisites,
identification algorithms are normally classified into three categories: (i) <i>de novo</i> peptide sequencing, (ii) peptide sequence-tag (PST)
searching, and (iii) uninterpreted sequence searching. However, in this notebook we focus on the explanation of (iii) uninterpreted sequence
searching, the most frequently used methods.
</div></p>
<h2><a name="Matching-and-scoring-of-Tandem-MS-peptide-identification" class="anchor" href="#Matching-and-scoring-of-Tandem-MS-peptide-identification">Matching and scoring of Tandem MS peptide identification</a></h2>
<a href="#Peptide-Identification" style="display: inline-block"><sup>&#8593;back</sup></a><br>
<div class="container">
<div Id="figure4" Style="float: right ; display: inline-block ; color: #44546a ; width: 60% ; padding: 15px">
    <img src="img/ComputationalProteinIdentification.png" Style="width: 100%">
    <div Style="padding-left: 1rem ; padding-right: 1rem ; text-align: justify ; font-size: 0.8rem"><b>
        Figure 4: Process of computational identification of peptides from their fragment spectra</b>
    </div>
</div>    
<p>Previously we learned, that peptides fragment to create patterns characteristic of a specific amino acid sequence. These patterns are reproducible and, in general, predictable taking the applied fragmentation method into account. This can be used for computational identification of peptides from their fragment spectra. This process can be subdivided into 5 main steps: spectrum preprocessing, selection of possible sequences, generating theoretical spectra, matching and scoring (<a href="#figure5" >Figure 4</a>). The first step is a preprocessing of the experimental spectra and is done to reduce noise. Secondly, all possible amino acid sequences are selected which match the particular precursor peptide mass. The possible peptides can but do not need to be restricted to a particular organism. A theoretical spectrum is predicted for each of these amino acid sequences. Matching and scoring is performed by comparing experimental spectra to their predicted corresponding theoretical spectra. The score function measures the closeness of fit between the experimental acquired and theoretical spectrum. There are many different score functions, which can be considered as the main reason of different identifications considering different identification algorithm. The most natural score function is the cross correlation score (xcorr) used by the commercially available search tool SEQUEST. One of the reasons the xcorr is so sensitive is because it involves a correction factor that assesses the background correlation for each acquired spectrum and the theoretically predicted spectrum from sequences within a database. To compute this correction factor, a measure of similarity is calculated at different offsets between a preprocessed mass spectrum and a theoretical spectrum. The final xcorr is then the correlation at zero offset minus the mean correlation from all the individual offsets.</p>
<p>Thus, the correlation function measures the coherence of two signals by, in effect, translating one signal across the other. This can be mathematically achieved using the formula for cross-correlation in the form for discrete input signals:</p>
<i><b>Equation 5</b></i>
<img src="https://latex.codecogs.com/gif.latex?\large&space;R_{\tau}&space;=&space;\sum_{i=0}^{n-1}x_{i}\cdot y_{i&plus;\tau}" title="\large R_{𝜏} = \sum_{i=0}^{n-1}x_{i}\cdot y_{i+𝜏}" style="margin: 1rem auto 1rem; display: block" />
<p>where x<sub>i</sub> is a peak of the reconstructed spectrum at position i and y<sub>i</sub> is a peak of the experimental spectrum. The displacement value 𝜏 is the amount by which the signal is offset during the translation and is varied over a range of values. If two signals are the same, the correlation function should have its maxima at 𝜏=0, where there is no offset. The average of the offset-correlation is seen as the average background correlation and needs to be subtracted from the correlation at 𝜏=0. It follows:</p>
<i><b>Equation 6</b></i>
<img src="https://latex.codecogs.com/gif.latex?xcorr&space;=&space;R_{0}&space;-&space;\frac{(\sum&space;\begin{matrix}&space;\tau=&plus;offeset\\&space;\tau=-offeset\end{matrix}R_{\tau})}{2*offset&plus;1}" title="xcorr = R_{0} - \frac{(\sum \begin{matrix} \tau=+offeset\\ \tau=-offeset\end{matrix}R_{\tau})}{2*offset+1}" style="margin: 1rem auto 1rem; display: block" />
<p>In practice many theoretical spectra have to be matched again a single experimental spectrum. Therefore, the calculation can be speed up by reformulating Equation 5 and Equation 6 and introduce a preprocessing step, which is independent of the predicted spectra.</p>
<i><b>Equation 7</b></i>
<img src="https://latex.codecogs.com/gif.latex?xcorr&space;=&space;x_{0}\cdot&space;y_{0}&space;-&space;\frac{(\sum&space;\begin{matrix}&space;\tau=&plus;offeset\\&space;\tau=-offeset\end{matrix}x_{0}\cdot&space;y_{\tau})}{2*offset&plus;1}" title="xcorr = x_{0}\cdot y_{0} - \frac{(\sum \begin{matrix} \tau=+offeset\\ \tau=-offeset\end{matrix}x_{0}\cdot y_{\tau})}{2*offset+1}" style="margin: 1rem auto 1rem; display: block" />
<p>For the preprocessed experimental spectrum y' it follows:</p>
<i><b>Equation 8</b></i>
<img src="https://latex.codecogs.com/gif.latex?xcorr&space;=&space;x_{0}\cdot&space;y'" title="xcorr = x_{0}\cdot y'" style="margin: 1rem auto 1rem; display: block" />
<p>where:</p>
<i><b>Equation 9</b></i>
<img src="https://latex.codecogs.com/gif.latex?y'&space;=&space;y_{0}&space;-&space;\frac{(\sum&space;\begin{matrix}&space;\tau=&plus;offeset\\&space;\tau=-offeset\end{matrix}x_{0}\cdot&space;y_{\tau})}{2*offset&plus;1}" title="y' = y_{0} - \frac{(\sum \begin{matrix} \tau=+offeset\\ \tau=-offeset\end{matrix}x_{0}\cdot y_{\tau})}{2*offset+1}" style="margin: 1rem auto 1rem; display: block" />
<div Style="text-align: justify ; margin-top: 2rem ; margin-bottom: 2rem ; line-height: 1.3 ; width: 85% ; margin-left: auto ; margin-right: auto ; padding: 10px ; border: 2px dotted #708090 ; color: #708090">
Matching a measured spectrum against chlamy database
</div>
</div>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="pp">#r</span> <span class="s">&quot;nuget: BioFSharp, 2.0.0-beta5&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: BioFSharp.IO, 2.0.0-beta5&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: Plotly.NET, 2.0.0-beta6&quot;</span>

<span class="pp">#if</span> <span class="id">IPYNB</span>
<span class="inactive">#r</span><span class="inactive"> </span><span class="inactive">&quot;nuget:</span><span class="inactive"> </span><span class="inactive">Plotly.NET,</span><span class="inactive"> </span><span class="inactive">2.0.0-beta6&quot;</span>
<span class="inactive">#r</span><span class="inactive"> </span><span class="inactive">&quot;nuget:</span><span class="inactive"> </span><span class="inactive">Plotly.NET.Interactive,</span><span class="inactive"> </span><span class="inactive">2.0.0-beta6&quot;</span>
<span class="pp">#endif</span> <span class="c">// IPYNB</span>

<span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="id">Plotly</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs2', 2)" onmouseover="showTip(event, 'fs2', 2)" class="id">NET</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs3', 3)" onmouseover="showTip(event, 'fs3', 3)" class="id">BioFSharp</span>
</code></pre>
<h3><a name="Step-1-Data-acquisition-and-preprocessing" class="anchor" href="#Step-1-Data-acquisition-and-preprocessing">Step 1: Data acquisition and preprocessing</a></h3>
<a href="#Peptide-Identification" style="display: inline-block"><sup>&#8593;back</sup></a><br>
<p>We load a single MS<sup>2</sup> spectrum that is saved in a mgf file.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 1</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs4', 4)" onmouseover="showTip(event, 'fs4', 4)" class="id">ms2</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs3', 5)" onmouseover="showTip(event, 'fs3', 5)" class="id">BioFSharp</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs5', 6)" onmouseover="showTip(event, 'fs5', 6)" class="id">IO</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs6', 7)" onmouseover="showTip(event, 'fs6', 7)" class="m">Mgf</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs7', 8)" onmouseover="showTip(event, 'fs7', 8)" class="id">readMgf</span> <span class="pn">(</span><span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">@&quot;/../AuxFiles/ms2sample.mgf&quot;</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs8', 9)" onmouseover="showTip(event, 'fs8', 9)" class="m">List</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs9', 10)" onmouseover="showTip(event, 'fs9', 10)" class="id">head</span>
    
<span onmouseout="hideTip(event, 'fs4', 11)" onmouseover="showTip(event, 'fs4', 11)" class="id">ms2</span>
</code></pre>
<table class="pre"><tr><td><pre><code>{ Parameters =
              map
                [("CHARGE", "2"); ("Id", "sample");
                 ("PRECURSORMZ", "511.2691141");
                 ("RawFile", "20180322MS_SHshift1514002");
                 ("SCANTIME", "51.95223333"); ("TITLE", "Exemplary MS2 Scan");
                 ("centroided", "True")]
  Mass =
        [|112.036653; 120.080594; 130.046346; 153.062829; 157.109862; 158.092818;
          171.076774; 175.067784; 189.087923; 189.338845; 198.083737; 199.072418;
          201.124816; 211.144868; 217.082739; 219.109405; 227.174856; 229.120531;
          233.165312; 238.16277; 242.111545; 253.044327; 255.142393; 256.169969;
          260.12437; 268.091868; 270.07139; 272.085967; 279.059968; 280.158068;
          281.197142; 285.149895; 288.08612; 296.090984; 296.202063; 305.155873;
          314.098751; 315.094288; 322.006424; 322.192782; 323.185918; 332.111085;
          332.315692; 338.132174; 342.195319; 345.718946; 351.209137; 353.131429;
          358.185808; 359.153126; 366.12195; 375.242387; 376.194349; 381.174952;
          384.133931; 385.13565; 393.22419; 394.226541; 399.184537; 400.16638;
          403.14732; 413.239741; 417.196992; 417.300178; 427.184224; 427.323445;
          428.176669; 429.240034; 439.257781; 443.242876; 445.194799; 446.196081;
          447.234097; 453.230571; 453.768446; 457.230402; 462.822253; 464.260295;
          464.586884; 465.26462; 465.500733; 465.652119; 470.259918; 471.209786;
          479.273205; 480.256729; 484.275011; 485.269833; 494.262925; 502.262504;
          502.76579; 511.271366; 512.268119; 513.259484; 514.283629; 516.232954;
          523.264365; 526.253758; 530.272248; 531.274408; ...|]
  Intensity =
             [|253.0; 1118.0; 96.0; 943.0; 290.0; 781.0; 1540.0; 46.0; 12618.0;
               24.0; 233.0; 12520.0; 5369.0; 48.0; 10722.0; 318.0; 345.0; 1643.0;
               24.0; 24.0; 1283.0; 120.0; 72.0; 239.0; 120.0; 120.0; 300.0; 48.0;
               48.0; 72.0; 24.0; 96.0; 421.0; 588.0; 48.0; 1960.0; 15750.0;
               288.0; 24.0; 3545.0; 215.0; 29178.0; 24.0; 211.0; 599.0; 286.0;
               48.0; 24.0; 127.0; 72.0; 72.0; 208.0; 987.0; 190.0; 48.0; 72.0;
               9226.0; 728.0; 1443.0; 439.0; 141.0; 405.0; 6152.0; 23.0; 2033.0;
               24.0; 120.0; 96.0; 48.0; 432.0; 8936.0; 610.0; 577.0; 96.0; 72.0;
               120.0; 164.0; 35545.0; 24.0; 2098.0; 48.0; 24.0; 48.0; 139.0;
               478.0; 72.0; 362.0; 72.0; 208.0; 2302.0; 436.0; 166.0; 657.0;
               365.0; 170.0; 700.0; 24.0; 24.0; 654.0; 72.0; ...|] }</code></pre></td></tr></table>
<p>Here, the spectrum is already centroidized as shown in <a href="JP08_Centroidisation.ipynb">JP08_Centroidisation</a> using the function
<code>msPeakPicking</code>. So we just visualize mass and intensity:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 2</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs10', 12)" onmouseover="showTip(event, 'fs10', 12)" class="id">ms2Chart</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs11', 13)" onmouseover="showTip(event, 'fs11', 13)" class="rt">Chart</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs12', 14)" onmouseover="showTip(event, 'fs12', 14)" class="id">Column</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs4', 15)" onmouseover="showTip(event, 'fs4', 15)" class="id">ms2</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs13', 16)" onmouseover="showTip(event, 'fs13', 16)" class="id">Mass</span><span class="pn">,</span> <span onmouseout="hideTip(event, 'fs4', 17)" onmouseover="showTip(event, 'fs4', 17)" class="id">ms2</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs14', 18)" onmouseover="showTip(event, 'fs14', 18)" class="id">Intensity</span><span class="pn">)</span>
</code></pre>
<div id="45d04da8-4c61-40e6-822b-0efa86da9469" style="width: 600px; height: 600px;"><!-- Plotly chart will be drawn inside this DIV --></div>
<script type="text/javascript">

            var renderPlotly_45d04da84c6140e6822b0efa86da9469 = function() {
            var fsharpPlotlyRequire = requirejs.config({context:'fsharp-plotly',paths:{plotly:'https://cdn.plot.ly/plotly-latest.min'}}) || require;
            fsharpPlotlyRequire(['plotly'], function(Plotly) {

            var data = [{"type":"bar","x":[112.036653,120.080594,130.046346,153.062829,157.109862,158.092818,171.076774,175.067784,189.087923,189.338845,198.083737,199.072418,201.124816,211.144868,217.082739,219.109405,227.174856,229.120531,233.165312,238.16277,242.111545,253.044327,255.142393,256.169969,260.12437,268.091868,270.07139,272.085967,279.059968,280.158068,281.197142,285.149895,288.08612,296.090984,296.202063,305.155873,314.098751,315.094288,322.006424,322.192782,323.185918,332.111085,332.315692,338.132174,342.195319,345.718946,351.209137,353.131429,358.185808,359.153126,366.12195,375.242387,376.194349,381.174952,384.133931,385.13565,393.22419,394.226541,399.184537,400.16638,403.14732,413.239741,417.196992,417.300178,427.184224,427.323445,428.176669,429.240034,439.257781,443.242876,445.194799,446.196081,447.234097,453.230571,453.768446,457.230402,462.822253,464.260295,464.586884,465.26462,465.500733,465.652119,470.259918,471.209786,479.273205,480.256729,484.275011,485.269833,494.262925,502.262504,502.76579,511.271366,512.268119,513.259484,514.283629,516.232954,523.264365,526.253758,530.272248,531.274408,540.271182,541.269683,542.301782,550.245007,551.259267,554.142479,554.274614,556.318098,558.279137,558.3985,558.683698,559.333957,560.310053,565.310057,569.267126,575.531536,576.56209,577.350783,578.342469,584.290103,587.319537,593.299364,593.524966,594.304648,596.317768,596.42057,597.263876,599.31635,600.326716,601.317288,611.295671,611.406695,612.32997,629.321462,630.314525,637.323317,641.517339,654.354213,655.359643,656.416175,660.340675,663.300761,664.182845,665.333326,682.336088,688.250119,690.423762,690.733508,691.427002,695.410243,700.347824,701.328517,707.331578,724.428818,726.439443,732.43487,734.380513,744.453695,763.456062,770.45088,775.452194,787.463086,788.432043,802.950406,803.101507,804.589423,805.457347,806.453247,810.323167,811.65784,813.329727,815.46007,819.463982,828.849964,829.423713,830.442404,847.414598,847.63519,898.460976,906.520859,916.467903],"y":[253.0,1118.0,96.0,943.0,290.0,781.0,1540.0,46.0,12618.0,24.0,233.0,12520.0,5369.0,48.0,10722.0,318.0,345.0,1643.0,24.0,24.0,1283.0,120.0,72.0,239.0,120.0,120.0,300.0,48.0,48.0,72.0,24.0,96.0,421.0,588.0,48.0,1960.0,15750.0,288.0,24.0,3545.0,215.0,29178.0,24.0,211.0,599.0,286.0,48.0,24.0,127.0,72.0,72.0,208.0,987.0,190.0,48.0,72.0,9226.0,728.0,1443.0,439.0,141.0,405.0,6152.0,23.0,2033.0,24.0,120.0,96.0,48.0,432.0,8936.0,610.0,577.0,96.0,72.0,120.0,164.0,35545.0,24.0,2098.0,48.0,24.0,48.0,139.0,478.0,72.0,362.0,72.0,208.0,2302.0,436.0,166.0,657.0,365.0,170.0,700.0,24.0,24.0,654.0,72.0,1132.0,119.0,120.0,48.0,48.0,24.0,24.0,24.0,2500.0,23.0,24.0,228.0,416.0,48.0,72.0,132.0,72.0,68417.0,4148.0,177.0,154.0,72.0,24.0,24.0,24.0,24.0,24.0,24.0,24.0,288.0,507.0,24.0,48.0,1388.0,131.0,48.0,24.0,72.0,48.0,24.0,24.0,24.0,24.0,24.0,158.0,96.0,10941.0,24.0,1317.0,24.0,358.0,72.0,24.0,24.0,24.0,24.0,48.0,24.0,48.0,24.0,48.0,296.0,153.0,96.0,24.0,48.0,27770.0,2648.0,24.0,24.0,24.0,48.0,24.0,24.0,72.0,24.0,96.0,24.0,96.0,72.0,796.0],"marker":{}}];
            var layout = {};
            var config = {};
            Plotly.newPlot('45d04da8-4c61-40e6-822b-0efa86da9469', data, layout, config);
});
            };
            if ((typeof(requirejs) !==  typeof(Function)) || (typeof(requirejs.config) !== typeof(Function))) {
                var script = document.createElement("script");
                script.setAttribute("src", "https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js");
                script.onload = function(){
                    renderPlotly_45d04da84c6140e6822b0efa86da9469();
                };
                document.getElementsByTagName("head")[0].appendChild(script);
            }
            else {
                renderPlotly_45d04da84c6140e6822b0efa86da9469();
            }
</script>

<p>Now, we will preprocess the experimental spectrum from our example. This is on the one hand to reduce noise, but also to make
the measured spectrum more like the once we are able to simulate.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 3</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs17', 22)" onmouseover="showTip(event, 'fs17', 22)" class="id">lowerScanLimit</span> <span class="o">=</span> <span class="n">150.</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs18', 23)" onmouseover="showTip(event, 'fs18', 23)" class="id">upperScanLimit</span> <span class="o">=</span> <span class="n">1000.</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs19', 24)" onmouseover="showTip(event, 'fs19', 24)" class="id">preprocessedIntesities</span> <span class="o">=</span>
    <span class="id">Mz</span><span class="pn">.</span><span class="id">PeakArray</span><span class="pn">.</span><span class="id">zip</span> <span onmouseout="hideTip(event, 'fs4', 25)" onmouseover="showTip(event, 'fs4', 25)" class="id">ms2</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs13', 26)" onmouseover="showTip(event, 'fs13', 26)" class="id">Mass</span> <span onmouseout="hideTip(event, 'fs4', 27)" onmouseover="showTip(event, 'fs4', 27)" class="id">ms2</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs14', 28)" onmouseover="showTip(event, 'fs14', 28)" class="id">Intensity</span>
    <span class="o">|&gt;</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs20', 29)" onmouseover="showTip(event, 'fs20', 29)" class="fn">pa</span> <span class="k">-&gt;</span> <span class="id">Mz</span><span class="pn">.</span><span class="id">PeakArray</span><span class="pn">.</span><span class="id">peaksToNearestUnitDaltonBinVector</span> <span onmouseout="hideTip(event, 'fs20', 30)" onmouseover="showTip(event, 'fs20', 30)" class="id">pa</span> <span onmouseout="hideTip(event, 'fs17', 31)" onmouseover="showTip(event, 'fs17', 31)" class="id">lowerScanLimit</span> <span onmouseout="hideTip(event, 'fs18', 32)" onmouseover="showTip(event, 'fs18', 32)" class="id">upperScanLimit</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs20', 33)" onmouseover="showTip(event, 'fs20', 33)" class="fn">pa</span> <span class="k">-&gt;</span> <span class="id">Mz</span><span class="pn">.</span><span class="id">SequestLike</span><span class="pn">.</span><span class="id">windowNormalizeIntensities</span> <span onmouseout="hideTip(event, 'fs20', 34)" onmouseover="showTip(event, 'fs20', 34)" class="id">pa</span> <span class="n">10</span><span class="pn">)</span>
    
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs21', 35)" onmouseover="showTip(event, 'fs21', 35)" class="id">intensityChart</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs11', 36)" onmouseover="showTip(event, 'fs11', 36)" class="rt">Chart</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs12', 37)" onmouseover="showTip(event, 'fs12', 37)" class="id">Column</span><span class="pn">(</span><span class="pn">[</span><span onmouseout="hideTip(event, 'fs17', 38)" onmouseover="showTip(event, 'fs17', 38)" class="id">lowerScanLimit</span> <span class="o">..</span> <span onmouseout="hideTip(event, 'fs18', 39)" onmouseover="showTip(event, 'fs18', 39)" class="id">upperScanLimit</span><span class="pn">]</span><span class="pn">,</span> <span onmouseout="hideTip(event, 'fs19', 40)" onmouseover="showTip(event, 'fs19', 40)" class="id">preprocessedIntesities</span><span class="pn">)</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<h3><a name="Step-2-Preselecting-the-peptides-of-interest" class="anchor" href="#Step-2-Preselecting-the-peptides-of-interest">Step 2: Preselecting the peptides of interest</a></h3>
<a href="#Peptide-Identification" style="display: inline-block"><sup>&#8593;back</sup></a><br>
<div class="container">
Every MS<sup>2</sup> spectrum is accompanied by a m/z ratio reported by the instrument. Additionally, we can estimate the charge looking 
at the isotopic cluster. We take the peptide "DTDILAAFR" from our previous notebook again. Our example has a measured m/z = 511.2691141 and 
a charge of z=2.
</div>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs22', 44)" onmouseover="showTip(event, 'fs22', 44)" class="id">peptideMass</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs23', 45)" onmouseover="showTip(event, 'fs23', 45)" class="m">Mass</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs24', 46)" onmouseover="showTip(event, 'fs24', 46)" class="id">ofMZ</span> <span class="n">511.2691141</span> <span class="n">2.</span>

<span onmouseout="hideTip(event, 'fs22', 47)" onmouseover="showTip(event, 'fs22', 47)" class="id">peptideMass</span>
</code></pre>
<table class="pre"><tr><td><pre><code>1020.523675</code></pre></td></tr></table>
<p>From our previos notebook <a href="JP04_Digestion_and_mass_calculation.ipynb">JP04 Digestion and mass calculation</a>, we know how to
calculate all peptide masses that we can expect to be present in <i>Chlamydomonas reinhardtii</i>.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 4</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs25', 48)" onmouseover="showTip(event, 'fs25', 48)" class="id">source</span> <span class="o">=</span> <span class="k">__SOURCE_DIRECTORY__</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs26', 49)" onmouseover="showTip(event, 'fs26', 49)" class="id">filePath</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs25', 50)" onmouseover="showTip(event, 'fs25', 50)" class="id">source</span> <span class="o">+</span> <span class="s">@&quot;/../AuxFiles/Chlamy_JGI5_5(Cp_Mp).fasta&quot;</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs27', 51)" onmouseover="showTip(event, 'fs27', 51)" class="id">peptideAndMasses</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs26', 52)" onmouseover="showTip(event, 'fs26', 52)" class="id">filePath</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs5', 53)" onmouseover="showTip(event, 'fs5', 53)" class="id">IO</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs28', 54)" onmouseover="showTip(event, 'fs28', 54)" class="m">FastA</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs29', 55)" onmouseover="showTip(event, 'fs29', 55)" class="id">fromFile</span> <span onmouseout="hideTip(event, 'fs30', 56)" onmouseover="showTip(event, 'fs30', 56)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs31', 57)" onmouseover="showTip(event, 'fs31', 57)" class="id">ofAminoAcidString</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs32', 58)" onmouseover="showTip(event, 'fs32', 58)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs33', 59)" onmouseover="showTip(event, 'fs33', 59)" class="id">toArray</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs34', 60)" onmouseover="showTip(event, 'fs34', 60)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs35', 61)" onmouseover="showTip(event, 'fs35', 61)" class="id">mapi</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs36', 62)" onmouseover="showTip(event, 'fs36', 62)" class="fn">i</span> <span onmouseout="hideTip(event, 'fs37', 63)" onmouseover="showTip(event, 'fs37', 63)" class="fn">fastAItem</span> <span class="k">-&gt;</span>
        <span onmouseout="hideTip(event, 'fs38', 64)" onmouseover="showTip(event, 'fs38', 64)" class="m">Digestion</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs39', 65)" onmouseover="showTip(event, 'fs39', 65)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs40', 66)" onmouseover="showTip(event, 'fs40', 66)" class="id">digest</span> <span onmouseout="hideTip(event, 'fs38', 67)" onmouseover="showTip(event, 'fs38', 67)" class="m">Digestion</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs41', 68)" onmouseover="showTip(event, 'fs41', 68)" class="m">Table</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs42', 69)" onmouseover="showTip(event, 'fs42', 69)" class="id">Trypsin</span> <span onmouseout="hideTip(event, 'fs36', 70)" onmouseover="showTip(event, 'fs36', 70)" class="fn">i</span> <span onmouseout="hideTip(event, 'fs37', 71)" onmouseover="showTip(event, 'fs37', 71)" class="fn">fastAItem</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs43', 72)" onmouseover="showTip(event, 'fs43', 72)" class="id">Sequence</span>
        <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs38', 73)" onmouseover="showTip(event, 'fs38', 73)" class="m">Digestion</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs39', 74)" onmouseover="showTip(event, 'fs39', 74)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs44', 75)" onmouseover="showTip(event, 'fs44', 75)" class="id">concernMissCleavages</span> <span class="n">0</span> <span class="n">0</span>
        <span class="pn">)</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs34', 76)" onmouseover="showTip(event, 'fs34', 76)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs45', 77)" onmouseover="showTip(event, 'fs45', 77)" class="id">concat</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs34', 78)" onmouseover="showTip(event, 'fs34', 78)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs46', 79)" onmouseover="showTip(event, 'fs46', 79)" class="id">map</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs47', 80)" onmouseover="showTip(event, 'fs47', 80)" class="fn">peptide</span> <span class="k">-&gt;</span>
        <span class="c">// calculate mass for each peptide</span>
        <span onmouseout="hideTip(event, 'fs47', 81)" onmouseover="showTip(event, 'fs47', 81)" class="fn">peptide</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs48', 82)" onmouseover="showTip(event, 'fs48', 82)" class="id">PepSequence</span><span class="pn">,</span> <span onmouseout="hideTip(event, 'fs49', 83)" onmouseover="showTip(event, 'fs49', 83)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs50', 84)" onmouseover="showTip(event, 'fs50', 84)" class="id">toMonoisotopicMassWith</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs51', 85)" onmouseover="showTip(event, 'fs51', 85)" class="m">BioItem</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs52', 86)" onmouseover="showTip(event, 'fs52', 86)" class="id">monoisoMass</span> <span onmouseout="hideTip(event, 'fs53', 87)" onmouseover="showTip(event, 'fs53', 87)" class="m">ModificationInfo</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs54', 88)" onmouseover="showTip(event, 'fs54', 88)" class="m">Table</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs55', 89)" onmouseover="showTip(event, 'fs55', 89)" class="id">H2O</span><span class="pn">)</span> <span onmouseout="hideTip(event, 'fs47', 90)" onmouseover="showTip(event, 'fs47', 90)" class="fn">peptide</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs48', 91)" onmouseover="showTip(event, 'fs48', 91)" class="id">PepSequence</span>
        <span class="pn">)</span>

<span onmouseout="hideTip(event, 'fs27', 92)" onmouseover="showTip(event, 'fs27', 92)" class="id">peptideAndMasses</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs34', 93)" onmouseover="showTip(event, 'fs34', 93)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs56', 94)" onmouseover="showTip(event, 'fs56', 94)" class="id">head</span>
</code></pre>
<table class="pre"><tr><td><pre><code>([Asp; Leu; His; Pro; Leu; Leu; Thr; Ser; Leu; Pro; Thr; Lys; Pro; Gly; Ser; Ala;
  Ala; Thr; Pro; Tyr; Thr; Thr; Thr; Gln; Ser; Pro; Pro; Ser; Thr; Thr; Leu; Ser;
  Ter], 3279.687453)</code></pre></td></tr></table>
<p>However, we are only interest in possible amino acid sequences, that match the particular precursor peptide mass of our example spectrum
with 1020.523675 Da. Additionaly, we should also consider a small measurement error.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 5</span>

<span onmouseout="hideTip(event, 'fs27', 95)" onmouseover="showTip(event, 'fs27', 95)" class="id">peptideAndMasses</span>
<span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs34', 96)" onmouseover="showTip(event, 'fs34', 96)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs57', 97)" onmouseover="showTip(event, 'fs57', 97)" class="id">filter</span> <span class="pn">(</span><span class="k">fun</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs58', 98)" onmouseover="showTip(event, 'fs58', 98)" class="fn">sequence</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs59', 99)" onmouseover="showTip(event, 'fs59', 99)" class="fn">mass</span><span class="pn">)</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs59', 100)" onmouseover="showTip(event, 'fs59', 100)" class="fn">mass</span> <span class="o">&gt;</span> <span class="n">1020.52</span>  <span class="o">&amp;&amp;</span> <span onmouseout="hideTip(event, 'fs59', 101)" onmouseover="showTip(event, 'fs59', 101)" class="fn">mass</span> <span class="o">&lt;</span> <span class="n">1020.53</span><span class="pn">)</span>
</code></pre>
<table class="pre"><tr><td><pre><code>[|([Leu; Ser; Ser; Glu; Val; Pro; Phe; Ser; Arg], 1020.52401);
  ([Leu; Ser; Ser; Glu; Val; Pro; Phe; Ser; Arg], 1020.52401);
  ([Ala; Ala; Thr; Thr; Ala; Thr; Met; Leu; Asn; Lys], 1020.527381);
  ([Leu; Gly; Ala; Leu; Glu; Ala; Glu; Tyr; Arg], 1020.52401);
  ([Ser; Leu; Ile; Gly; Asp; Asp; Phe; Val; Arg], 1020.52401);
  ([Gly; Glu; Glu; Leu; Leu; Ser; Ala; Phe; Arg], 1020.52401);
  ([Thr; Thr; Pro; Tyr; Gly; Val; Val; Glu; Arg], 1020.52401);
  ([Val; Asp; Phe; Thr; Asn; Ala; Asn; Leu; Lys], 1020.52401);
  ([Leu; Tyr; Ala; Ala; Asp; Gly; Asn; Ala; Val; Lys], 1020.52401);
  ([Asp; Val; Leu; Thr; Trp; Tyr; Pro; Lys], 1020.528032);
  ([Glu; Tyr; Gln; Ala; Glu; Leu; Leu; Arg], 1020.52401);
  ([Phe; Ala; Gln; Gly; Ala; Asp; Ala; Thr; Ile; Lys], 1020.52401);
  ([Asp; Phe; Val; Glu; Val; Val; Ser; Ala; Arg], 1020.52401);
  ([Gln; Asn; Ile; Phe; Pro; Thr; Ser; Ser; Lys], 1020.52401);
  ([Ala; Val; Gln; Leu; Tyr; Asp; Gln; Gly; Lys], 1020.52401);
  ([Val; Val; Asp; Leu; Ser; Ala; Asp; Phe; Arg], 1020.52401);
  ([Gly; Ser; Thr; Tyr; Gly; Gln; Pro; Ala; Leu; Lys], 1020.52401);
  ([Asn; Tyr; Val; Glu; Asp; Leu; Leu; Arg], 1020.52401);
  ([Ala; Glu; Ala; Val; Glu; Thr; Phe; Val; Arg], 1020.52401);
  ([Val; Ser; Thr; Met; Asp; Thr; Leu; Val; Arg], 1020.527381);
  ([Ser; Ala; Ser; Pro; Pro; Pro; Pro; Gly; Pro; Ser; Lys], 1020.52401);
  ([Ser; Ala; Ser; Pro; Pro; Pro; Pro; Gly; Pro; Ser; Lys], 1020.52401);
  ([Ser; Ala; Ser; Pro; Pro; Pro; Pro; Gly; Pro; Ser; Lys], 1020.52401);
  ([Trp; Leu; Ala; Ser; Val; Ile; Phe; Glu; Gly; Ter], 1020.528032);
  ([Trp; Pro; Gly; Val; Ala; Ala; His; Gln; Arg], 1020.525347);
  ([Ala; Asp; Leu; Asp; Gly; Val; Leu; Tyr; Arg], 1020.52401);
  ([Val; Ile; Gly; Cys; Thr; Thr; Thr; Gly; Ala; Ala; Lys], 1020.527381);
  ([Val; Ile; Gly; Cys; Thr; Thr; Thr; Gly; Ala; Ala; Lys], 1020.527381);
  ([Phe; Leu; Asp; Ala; Gly; Leu; Glu; Thr; Arg], 1020.52401);
  ([Phe; Val; Thr; Leu; Ala; Glu; Gly; Glu; Arg], 1020.52401);
  ([Ala; Tyr; Gln; Gly; Ala; Glu; Ala; Ala; Leu; Lys], 1020.52401);
  ([Asp; Val; Leu; Asp; Trp; Phe; Val; Lys], 1020.528032);
  ([Leu; Glu; Gly; Tyr; Val; Glu; Gly; Val; Arg], 1020.52401);
  ([Gln; Ile; Ile; Thr; Phe; Glu; Asp; Arg], 1020.52401);
  ([Asp; Thr; Asp; Ile; Leu; Ala; Ala; Phe; Arg], 1020.52401)|]</code></pre></td></tr></table>
<p>In the previous notebook <a href="JP09_Fragmentation_for_peptide_identification.ipynb">Fragmentation for peptide identification</a>,
we used functions that generate the theoretical series of b- and y-ions from the given peptide. Combined with the function
<code>Mz.SequestLike.predictOf</code> that generates theoretical spectra that fit the Sequest scoring algorithm.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 6</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs60', 102)" onmouseover="showTip(event, 'fs60', 102)" class="fn">predictFromSequence</span> <span onmouseout="hideTip(event, 'fs61', 103)" onmouseover="showTip(event, 'fs61', 103)" class="fn">peptide</span> <span class="o">=</span>
    <span class="pn">[</span>
        <span onmouseout="hideTip(event, 'fs61', 104)" onmouseover="showTip(event, 'fs61', 104)" class="fn">peptide</span>
        <span class="o">|&gt;</span> <span class="id">Mz</span><span class="pn">.</span><span class="id">Fragmentation</span><span class="pn">.</span><span class="id">Series</span><span class="pn">.</span> <span class="id">yOfBioList</span> <span onmouseout="hideTip(event, 'fs51', 105)" onmouseover="showTip(event, 'fs51', 105)" class="id">BioItem</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs62', 106)" onmouseover="showTip(event, 'fs62', 106)" class="id">initMonoisoMassWithMemP</span>
        <span onmouseout="hideTip(event, 'fs61', 107)" onmouseover="showTip(event, 'fs61', 107)" class="fn">peptide</span>
        <span class="o">|&gt;</span> <span class="id">Mz</span><span class="pn">.</span><span class="id">Fragmentation</span><span class="pn">.</span><span class="id">Series</span><span class="pn">.</span><span class="id">bOfBioList</span> <span onmouseout="hideTip(event, 'fs51', 108)" onmouseover="showTip(event, 'fs51', 108)" class="id">BioItem</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs62', 109)" onmouseover="showTip(event, 'fs62', 109)" class="id">initMonoisoMassWithMemP</span>
    <span class="pn">]</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs8', 110)" onmouseover="showTip(event, 'fs8', 110)" class="m">List</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs63', 111)" onmouseover="showTip(event, 'fs63', 111)" class="id">concat</span>
    <span class="o">|&gt;</span> <span class="id">Mz</span><span class="pn">.</span><span class="id">SequestLike</span><span class="pn">.</span><span class="id">predictOf</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs17', 112)" onmouseover="showTip(event, 'fs17', 112)" class="id">lowerScanLimit</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs18', 113)" onmouseover="showTip(event, 'fs18', 113)" class="id">upperScanLimit</span><span class="pn">)</span> <span class="n">2.</span>

<span onmouseout="hideTip(event, 'fs60', 114)" onmouseover="showTip(event, 'fs60', 114)" class="fn">predictFromSequence</span>
</code></pre>
<h3><a name="Step-3-4-Matching-and-Scoring" class="anchor" href="#Step-3-4-Matching-and-Scoring">Step 3+4: Matching and Scoring</a></h3>
<a href="#Peptide-Identification" style="display: inline-block"><sup>&#8593;back</sup></a><br>
<p>In the matching step, we compare theoretical spectra of peptides within our precursor peptide mass range with our measured spectra.
We get a score which tells us how well the theoretical spectrum fits the given experimental spectrum.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 7</span>

<span onmouseout="hideTip(event, 'fs27', 115)" onmouseover="showTip(event, 'fs27', 115)" class="id">peptideAndMasses</span>
<span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs34', 116)" onmouseover="showTip(event, 'fs34', 116)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs57', 117)" onmouseover="showTip(event, 'fs57', 117)" class="id">filter</span> <span class="pn">(</span><span class="k">fun</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs58', 118)" onmouseover="showTip(event, 'fs58', 118)" class="fn">sequence</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs59', 119)" onmouseover="showTip(event, 'fs59', 119)" class="fn">mass</span><span class="pn">)</span> <span class="k">-&gt;</span>
    <span onmouseout="hideTip(event, 'fs59', 120)" onmouseover="showTip(event, 'fs59', 120)" class="fn">mass</span> <span class="o">&gt;</span> <span class="n">1020.52</span>  <span class="o">&amp;&amp;</span> <span onmouseout="hideTip(event, 'fs59', 121)" onmouseover="showTip(event, 'fs59', 121)" class="fn">mass</span> <span class="o">&lt;</span> <span class="n">1020.53</span>
<span class="pn">)</span>
<span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs34', 122)" onmouseover="showTip(event, 'fs34', 122)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs46', 123)" onmouseover="showTip(event, 'fs46', 123)" class="id">map</span> <span class="pn">(</span><span class="k">fun</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs58', 124)" onmouseover="showTip(event, 'fs58', 124)" class="fn">sequence</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs59', 125)" onmouseover="showTip(event, 'fs59', 125)" class="fn">mass</span><span class="pn">)</span>    <span class="k">-&gt;</span>
    <span onmouseout="hideTip(event, 'fs58', 126)" onmouseover="showTip(event, 'fs58', 126)" class="fn">sequence</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs60', 127)" onmouseover="showTip(event, 'fs60', 127)" class="fn">predictFromSequence</span> <span onmouseout="hideTip(event, 'fs58', 128)" onmouseover="showTip(event, 'fs58', 128)" class="fn">sequence</span>
<span class="pn">)</span>
<span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs34', 129)" onmouseover="showTip(event, 'fs34', 129)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs46', 130)" onmouseover="showTip(event, 'fs46', 130)" class="id">map</span> <span class="pn">(</span><span class="k">fun</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs58', 131)" onmouseover="showTip(event, 'fs58', 131)" class="fn">sequence</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs64', 132)" onmouseover="showTip(event, 'fs64', 132)" class="fn">theoSpectrum</span><span class="pn">)</span> <span class="k">-&gt;</span> 
    <span onmouseout="hideTip(event, 'fs58', 133)" onmouseover="showTip(event, 'fs58', 133)" class="fn">sequence</span><span class="pn">,</span><span class="id">Mz</span><span class="pn">.</span><span class="id">SequestLike</span><span class="pn">.</span><span class="id">scoreSingle</span> <span onmouseout="hideTip(event, 'fs64', 134)" onmouseover="showTip(event, 'fs64', 134)" class="id">theoSpectrum</span> <span onmouseout="hideTip(event, 'fs19', 135)" onmouseover="showTip(event, 'fs19', 135)" class="id">preprocessedIntesities</span>
<span class="pn">)</span>
<span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs34', 136)" onmouseover="showTip(event, 'fs34', 136)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs65', 137)" onmouseover="showTip(event, 'fs65', 137)" class="id">sortByDescending</span> <span class="pn">(</span><span class="k">fun</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs58', 138)" onmouseover="showTip(event, 'fs58', 138)" class="fn">sequence</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs66', 139)" onmouseover="showTip(event, 'fs66', 139)" class="fn">score</span><span class="pn">)</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs66', 140)" onmouseover="showTip(event, 'fs66', 140)" class="fn">score</span><span class="pn">)</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<p>Finaly, we pick the sequence with the best score and are done for now. Notice however, that in a real world we would need to
relate our score to the complete data set to get an idea of the overall quality and which numerical value we could trust.</p>
<hr>
<nav class="level is-mobile">
    <div class="level-left">
        <div class="level-item">
            <button class="button is-primary is-outlined" onclick="location.href='/JP09_Fragmentation_for_peptide_identification.html';">&#171; JP09</button>
        </div>
    </div>
    <div class="level-right">
        <div class="level-item">
            <button class="button is-primary is-outlined" onclick="location.href='/JP11_Quantification.html';">JP11 &#187;</button>
        </div>
    </div>
</nav>

                </div>
            </div>
        </div>
        <div class="fsdocs-tip" id="fs1">namespace Plotly</div>
<div class="fsdocs-tip" id="fs2">namespace Plotly.NET</div>
<div class="fsdocs-tip" id="fs3">namespace BioFSharp</div>
<div class="fsdocs-tip" id="fs4">val ms2 : IO.Mgf.MgfEntry</div>
<div class="fsdocs-tip" id="fs5">namespace BioFSharp.IO</div>
<div class="fsdocs-tip" id="fs6">module Mgf<br /><br />from BioFSharp.IO</div>
<div class="fsdocs-tip" id="fs7">val readMgf : path:string -&gt; IO.Mgf.MgfEntry list</div>
<div class="fsdocs-tip" id="fs8">Multiple items<br />module List<br /><br />from Microsoft.FSharp.Collections<br /><br />--------------------<br />type List&lt;&#39;T&gt; =<br />&#160;&#160;| ( [] )<br />&#160;&#160;| ( :: ) of Head: &#39;T * Tail: &#39;T list<br />&#160;&#160;&#160;&#160;interface IReadOnlyList&lt;&#39;T&gt;<br />&#160;&#160;&#160;&#160;interface IReadOnlyCollection&lt;&#39;T&gt;<br />&#160;&#160;&#160;&#160;interface IEnumerable<br />&#160;&#160;&#160;&#160;interface IEnumerable&lt;&#39;T&gt;<br />&#160;&#160;&#160;&#160;member GetReverseIndex : rank:int * offset:int -&gt; int<br />&#160;&#160;&#160;&#160;member GetSlice : startIndex:int option * endIndex:int option -&gt; &#39;T list<br />&#160;&#160;&#160;&#160;member Head : &#39;T<br />&#160;&#160;&#160;&#160;member IsEmpty : bool<br />&#160;&#160;&#160;&#160;member Item : index:int -&gt; &#39;T with get<br />&#160;&#160;&#160;&#160;member Length : int<br />&#160;&#160;&#160;&#160;...</div>
<div class="fsdocs-tip" id="fs9">val head : list:&#39;T list -&gt; &#39;T</div>
<div class="fsdocs-tip" id="fs10">val ms2Chart : GenericChart.GenericChart</div>
<div class="fsdocs-tip" id="fs11">type Chart =<br />&#160;&#160;static member Area : xy:seq&lt;#IConvertible * #IConvertible&gt; * ?Name:string * ?ShowMarkers:bool * ?Showlegend:bool * ?MarkerSymbol:Symbol * ?Color:string * ?Opacity:float * ?Labels:seq&lt;string&gt; * ?TextPosition:TextPosition * ?TextFont:Font * ?Dash:DrawingStyle * ?Width:&#39;a2 -&gt; GenericChart<br />&#160;&#160;static member Area : x:seq&lt;#IConvertible&gt; * y:seq&lt;#IConvertible&gt; * ?Name:string * ?ShowMarkers:bool * ?Showlegend:bool * ?MarkerSymbol:Symbol * ?Color:string * ?Opacity:float * ?Labels:seq&lt;string&gt; * ?TextPosition:TextPosition * ?TextFont:Font * ?Dash:DrawingStyle * ?Width:&#39;a2 -&gt; GenericChart<br />&#160;&#160;static member Bar : keysvalues:seq&lt;#IConvertible * #IConvertible&gt; * ?Name:string * ?Showlegend:bool * ?Color:&#39;a2 * ?Opacity:float * ?Labels:seq&lt;string&gt; * ?TextPosition:TextPosition * ?TextFont:Font * ?Marker:Marker -&gt; GenericChart<br />&#160;&#160;static member Bar : keys:seq&lt;#IConvertible&gt; * values:seq&lt;#IConvertible&gt; * ?Name:string * ?Showlegend:bool * ?Color:&#39;a2 * ?Opacity:float * ?Labels:seq&lt;string&gt; * ?TextPosition:TextPosition * ?TextFont:Font * ?Marker:Marker -&gt; GenericChart<br />&#160;&#160;static member BoxPlot : xy:seq&lt;&#39;a0 * &#39;a1&gt; * ?Name:string * ?Showlegend:bool * ?Color:string * ?Fillcolor:&#39;a2 * ?Opacity:float * ?Whiskerwidth:&#39;a3 * ?Boxpoints:Boxpoints * ?Boxmean:BoxMean * ?Jitter:&#39;a4 * ?Pointpos:&#39;a5 * ?Orientation:Orientation * ?Marker:Marker * ?Line:Line * ?Alignmentgroup:&#39;a6 * ?Offsetgroup:&#39;a7 * ?Notched:bool * ?NotchWidth:float * ?QuartileMethod:QuartileMethod -&gt; GenericChart<br />&#160;&#160;static member BoxPlot : ?x:&#39;a0 * ?y:&#39;a1 * ?Name:string * ?Showlegend:bool * ?Color:string * ?Fillcolor:&#39;a2 * ?Opacity:float * ?Whiskerwidth:&#39;a3 * ?Boxpoints:Boxpoints * ?Boxmean:BoxMean * ?Jitter:&#39;a4 * ?Pointpos:&#39;a5 * ?Orientation:Orientation * ?Marker:Marker * ?Line:Line * ?Alignmentgroup:&#39;a6 * ?Offsetgroup:&#39;a7 * ?Notched:bool * ?NotchWidth:float * ?QuartileMethod:QuartileMethod -&gt; GenericChart<br />&#160;&#160;static member Bubble : xysizes:seq&lt;#IConvertible * #IConvertible * #IConvertible&gt; * ?Name:string * ?Showlegend:bool * ?MarkerSymbol:Symbol * ?Color:string * ?Opacity:float * ?Labels:seq&lt;string&gt; * ?TextPosition:TextPosition * ?TextFont:Font * ?StackGroup:string * ?Orientation:Orientation * ?GroupNorm:GroupNorm * ?UseWebGL:bool -&gt; GenericChart<br />&#160;&#160;static member Bubble : x:seq&lt;#IConvertible&gt; * y:seq&lt;#IConvertible&gt; * sizes:seq&lt;#IConvertible&gt; * ?Name:string * ?Showlegend:bool * ?MarkerSymbol:Symbol * ?Color:string * ?Opacity:float * ?Labels:seq&lt;string&gt; * ?TextPosition:TextPosition * ?TextFont:Font * ?StackGroup:string * ?Orientation:Orientation * ?GroupNorm:GroupNorm * ?UseWebGL:bool -&gt; GenericChart<br />&#160;&#160;static member Candlestick : stockTimeSeries:seq&lt;DateTime * StockData&gt; * ?Increasing:Line * ?Decreasing:Line * ?WhiskerWidth:float * ?Line:Line * ?XCalendar:Calendar -&gt; GenericChart<br />&#160;&#160;static member Candlestick : open:seq&lt;#IConvertible&gt; * high:seq&lt;#IConvertible&gt; * low:seq&lt;#IConvertible&gt; * close:seq&lt;#IConvertible&gt; * x:seq&lt;#IConvertible&gt; * ?Increasing:Line * ?Decreasing:Line * ?WhiskerWidth:float * ?Line:Line * ?XCalendar:Calendar -&gt; GenericChart<br />&#160;&#160;...</div>
<div class="fsdocs-tip" id="fs12">static member Chart.Column : keysvalues:seq&lt;#System.IConvertible * #System.IConvertible&gt; * ?Name:string * ?Showlegend:bool * ?Color:&#39;a2 * ?Opacity:float * ?Labels:seq&lt;string&gt; * ?TextPosition:StyleParam.TextPosition * ?TextFont:Font * ?Marker:Marker -&gt; GenericChart.GenericChart<br />static member Chart.Column : keys:seq&lt;#System.IConvertible&gt; * values:seq&lt;#System.IConvertible&gt; * ?Name:string * ?Showlegend:bool * ?Color:&#39;a2 * ?Opacity:float * ?Labels:seq&lt;string&gt; * ?TextPosition:StyleParam.TextPosition * ?TextFont:Font * ?Marker:Marker -&gt; GenericChart.GenericChart</div>
<div class="fsdocs-tip" id="fs13">IO.Mgf.MgfEntry.Mass: float []</div>
<div class="fsdocs-tip" id="fs14">IO.Mgf.MgfEntry.Intensity: float []</div>
<div class="fsdocs-tip" id="fs15">module GenericChart<br /><br />from Plotly.NET</div>
<div class="fsdocs-tip" id="fs16">val toChartHTML : gChart:GenericChart.GenericChart -&gt; string</div>
<div class="fsdocs-tip" id="fs17">val lowerScanLimit : float</div>
<div class="fsdocs-tip" id="fs18">val upperScanLimit : float</div>
<div class="fsdocs-tip" id="fs19">val preprocessedIntesities : seq&lt;System.IConvertible&gt;</div>
<div class="fsdocs-tip" id="fs20">val pa : obj</div>
<div class="fsdocs-tip" id="fs21">val intensityChart : GenericChart.GenericChart</div>
<div class="fsdocs-tip" id="fs22">val peptideMass : float</div>
<div class="fsdocs-tip" id="fs23">module Mass<br /><br />from BioFSharp</div>
<div class="fsdocs-tip" id="fs24">val ofMZ : mz:float -&gt; z:float -&gt; float</div>
<div class="fsdocs-tip" id="fs25">val source : string</div>
<div class="fsdocs-tip" id="fs26">val filePath : string</div>
<div class="fsdocs-tip" id="fs27">val peptideAndMasses : (AminoAcids.AminoAcid list * float) []</div>
<div class="fsdocs-tip" id="fs28">module FastA<br /><br />from BioFSharp.IO</div>
<div class="fsdocs-tip" id="fs29">val fromFile : converter:(seq&lt;char&gt; -&gt; &#39;a) -&gt; filePath:string -&gt; seq&lt;IO.FastA.FastaItem&lt;&#39;a&gt;&gt;</div>
<div class="fsdocs-tip" id="fs30">Multiple items<br />module BioArray<br /><br />from BioFSharp.BioCollectionsExtensions<br /><br />--------------------<br />module BioArray<br /><br />from BioFSharp</div>
<div class="fsdocs-tip" id="fs31">val ofAminoAcidString : s:#seq&lt;char&gt; -&gt; BioArray.BioArray&lt;AminoAcids.AminoAcid&gt;</div>
<div class="fsdocs-tip" id="fs32">Multiple items<br />module Seq<br /><br />from Plotly.NET<br /><br />--------------------<br />module Seq<br /><br />from Microsoft.FSharp.Collections</div>
<div class="fsdocs-tip" id="fs33">val toArray : source:seq&lt;&#39;T&gt; -&gt; &#39;T []</div>
<div class="fsdocs-tip" id="fs34">module Array<br /><br />from Microsoft.FSharp.Collections</div>
<div class="fsdocs-tip" id="fs35">val mapi : mapping:(int -&gt; &#39;T -&gt; &#39;U) -&gt; array:&#39;T [] -&gt; &#39;U []</div>
<div class="fsdocs-tip" id="fs36">val i : int</div>
<div class="fsdocs-tip" id="fs37">val fastAItem : IO.FastA.FastaItem&lt;BioArray.BioArray&lt;AminoAcids.AminoAcid&gt;&gt;</div>
<div class="fsdocs-tip" id="fs38">module Digestion<br /><br />from BioFSharp</div>
<div class="fsdocs-tip" id="fs39">module BioArray<br /><br />from BioFSharp.Digestion</div>
<div class="fsdocs-tip" id="fs40">val digest : protease:Digestion.Protease -&gt; proteinID:&#39;a -&gt; aas:AminoAcids.AminoAcid [] -&gt; Digestion.DigestedPeptide&lt;&#39;a&gt; []</div>
<div class="fsdocs-tip" id="fs41">module Table<br /><br />from BioFSharp.Digestion</div>
<div class="fsdocs-tip" id="fs42">val Trypsin : Digestion.Protease</div>
<div class="fsdocs-tip" id="fs43">IO.FastA.FastaItem.Sequence: BioArray.BioArray&lt;AminoAcids.AminoAcid&gt;</div>
<div class="fsdocs-tip" id="fs44">val concernMissCleavages : minMissCleavages:int -&gt; maxMisscleavages:int -&gt; digestedPeptides:Digestion.DigestedPeptide&lt;&#39;a&gt; [] -&gt; Digestion.DigestedPeptide&lt;&#39;a&gt; [] (requires equality)</div>
<div class="fsdocs-tip" id="fs45">val concat : arrays:seq&lt;&#39;T []&gt; -&gt; &#39;T []</div>
<div class="fsdocs-tip" id="fs46">val map : mapping:(&#39;T -&gt; &#39;U) -&gt; array:&#39;T [] -&gt; &#39;U []</div>
<div class="fsdocs-tip" id="fs47">val peptide : Digestion.DigestedPeptide&lt;int&gt;</div>
<div class="fsdocs-tip" id="fs48">Digestion.DigestedPeptide.PepSequence: AminoAcids.AminoAcid list</div>
<div class="fsdocs-tip" id="fs49">Multiple items<br />module BioSeq<br /><br />from BioFSharp.BioCollectionsExtensions<br /><br />--------------------<br />module BioSeq<br /><br />from BioFSharp</div>
<div class="fsdocs-tip" id="fs50">val toMonoisotopicMassWith : state:float -&gt; bs:seq&lt;#IBioItem&gt; -&gt; float</div>
<div class="fsdocs-tip" id="fs51">module BioItem<br /><br />from BioFSharp</div>
<div class="fsdocs-tip" id="fs52">val monoisoMass&lt;&#39;a (requires &#39;a :&gt; IBioItem)&gt; : (IBioItem -&gt; float) (requires &#39;a :&gt; IBioItem)</div>
<div class="fsdocs-tip" id="fs53">module ModificationInfo<br /><br />from BioFSharp</div>
<div class="fsdocs-tip" id="fs54">module Table<br /><br />from BioFSharp.ModificationInfo</div>
<div class="fsdocs-tip" id="fs55">val H2O : ModificationInfo.Modification</div>
<div class="fsdocs-tip" id="fs56">val head : array:&#39;T [] -&gt; &#39;T</div>
<div class="fsdocs-tip" id="fs57">val filter : predicate:(&#39;T -&gt; bool) -&gt; array:&#39;T [] -&gt; &#39;T []</div>
<div class="fsdocs-tip" id="fs58">val sequence : AminoAcids.AminoAcid list</div>
<div class="fsdocs-tip" id="fs59">val mass : float</div>
<div class="fsdocs-tip" id="fs60">val predictFromSequence : peptide:&#39;a -&gt; &#39;b</div>
<div class="fsdocs-tip" id="fs61">val peptide : &#39;a</div>
<div class="fsdocs-tip" id="fs62">val initMonoisoMassWithMemP&lt;&#39;a (requires &#39;a :&gt; IBioItem)&gt; : (IBioItem -&gt; float) (requires &#39;a :&gt; IBioItem)</div>
<div class="fsdocs-tip" id="fs63">val concat : lists:seq&lt;&#39;T list&gt; -&gt; &#39;T list</div>
<div class="fsdocs-tip" id="fs64">val theoSpectrum : obj</div>
<div class="fsdocs-tip" id="fs65">val sortByDescending : projection:(&#39;T -&gt; &#39;Key) -&gt; array:&#39;T [] -&gt; &#39;T [] (requires comparison)</div>
<div class="fsdocs-tip" id="fs66">val score : System.IComparable</div>

    </div>

    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <script type="text/javascript">var fsdocs_search_baseurl = '/BIO-BTE-06-L-7/'</script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/lunr.js/2.3.8/lunr.min.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.min.js"></script>
    <script type="text/javascript" src="/BIO-BTE-06-L-7/content/fsdocs-search.js"></script>
    <!-- END SEARCH BOX: this adds support for the search box -->
</body>

</html>