<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <title>NB01b Plant Systems Biology
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="author" content="">
    <link rel="shortcut icon" type="image/x-icon" href="/BIO-BTE-06-L-7//img/favicon.ico">
    <script src="https://code.jquery.com/jquery-3.4.1.min.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <!-- <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/MathJax.js?config=TeX-MML-AM_CHTML"></script> -->

    <link type="text/css" rel="stylesheet" href="/BIO-BTE-06-L-7/content/fsdocs-default.css" />
    <link type="text/css" rel="stylesheet" href="/BIO-BTE-06-L-7/content/fsdocs-custom.css" />
    <script type="text/javascript" src="/BIO-BTE-06-L-7/content/fsdocs-tips.js"></script>
    
    <link type="text/css" rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" />
    <script defer type="text/javascript" src="https://kit.fontawesome.com/0d3e0ea7a6.js"></script>
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <!-- END SEARCH BOX: this adds support for the search box -->
    
</head> 

<body>
    <div class="columns is-fullheight m-0">
        <div id="fsdocs-menu" class="column is-2 is-paddingless">
            <aside class="menu">
                <div class="container is-paddingless is-marginless has-background-dark">
                    <a href="/BIO-BTE-06-L-7/">
                        <figure class="image is-128x128 container">
                            <img src="img/logo.svg">
                        </figure>
                    </a>
                    <h2 class="title has-text-centered has-text-white pb-4">BIO-BTE-06-L-7</h2>
                </div>
                <div class="content">
                    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
                    <div id="header">
                        <div class="field">
                            <div class="control has-icons-left has-icons-right searchbox" id="fsdocs-searchbox">
                                <input data-search-input="" id="search-by" type="search" placeholder="Search..." class="input"/>
                                <span class="icon is-small is-left">
                                    <i class="fas fa-search"></i>
                                </span>
                            </div>
                        </div>
                    </div>

                    <!-- END SEARCH BOX: this adds support for the search box -->
                    <ul class="navbar-nav">
                        <li class="nav-item">
                            <a class="nav-link" href="NB00a_FSharp_Coding_literacy_Part_I.html">
                                <span class="text">
                                    <span>NB00a FSharp Coding literacy part I</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB00b_FSharp_Coding_literacy_Part_II.html">
                                <span class="text">
                                    <span>NB00b FSharp Coding literacy part II</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB01a_Systems_Biology_FSharp_Introduction.html">
                                <span class="text">
                                    <span>NB01a Systems Biology</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB01b_Plant_Systems_Biology.html">
                                <span class="text">
                                    <span>NB01b Plant Systems Biology</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB02a_Mass_spectrometry_based_proteomics.html">
                                <span class="text">
                                    <span>NB02a Mass Spectrometry Based Proteomics</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB02b_Digestion_and_mass_calculation.html">
                                <span class="text">
                                    <span>NB02b Digestion and mass calculation</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB02c_Isotopic_distribution.html">
                                <span class="text">
                                    <span>NB02c Isotopic Distribution</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB03a_Retention_time_and_scan_time.html">
                                <span class="text">
                                    <span>NB03a Retention time and scan time</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB03b_Signal_detection_and_quantification.html">
                                <span class="text">
                                    <span>NB03b Signal detection and quantification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB03c_Centroidisation.html">
                                <span class="text">
                                    <span>NB03c Centroidisation</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB04a_Fragmentation_for_peptide_identification.html">
                                <span class="text">
                                    <span>NB04a Fragmentation for peptide identification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB04b_Peptide_Identification.html">
                                <span class="text">
                                    <span>NB04b Peptide Identification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB05a_Quantification.html">
                                <span class="text">
                                    <span>NB05a Quantification</span>
                                </span>
                            </a>
                        </li>
         <!--                <li class="nav-item">
                            <a class="nav-link" href="NB06a_Research_Data_Management.html">
                                <span class="text">
                                    <span>NB06a Research Data Management</span>
                                </span>
                            </a>
                        </li> -->
                        <li class="nav-item">
                            <a class="nav-link" href="NB07a_Working_With_Deedle.html">
                                <span class="text">
                                    <span>NB07a Working with Deedle</span>
                                </span>
                            </a>
                        </li>
                        <!--<li class="nav-item">
                            <a class="nav-link" href="NB08a_Data_Access_And_Quality_Control.html">
                                <span class="text">
                                    <span>NB08a Data Access and Quality Control</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08a_Data_Access_And_Quality_Control_SDS.html">
                                <span class="text">
                                    <span>NB08a Data Access and Quality Control (SDS)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08a_Data_Access_And_Quality_Control_BN.html">
                                <span class="text">
                                    <span>NB08a Data Access and Quality Control (BN)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08b_Label_efficiency.html">
                                <span class="text">
                                    <span>NB08b Label Efficiency</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08b_Label_efficiency_SDS.html">
                                <span class="text">
                                    <span>NB08b Label Efficiency (SDS)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08b_Label_efficiency_BN.html">
                                <span class="text">
                                    <span>NB08b Label Efficiency (BN)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08c_Absolute_Quantification.html">
                                <span class="text">
                                    <span>NB08c Absolute Quantification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08c_Absolute_Quantification_SDS.html">
                                <span class="text">
                                    <span>NB08c Absolute Quantification (SDS)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08c_Absolute_Quantification_BN.html">
                                <span class="text">
                                    <span>NB08c Absolute Quantification (BN)</span>
                                </span>
                            </a>
                        </li>-->
                        <li class="nav-header">Links</li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://csb.bio.uni-kl.de/#teaching">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fas fa-graduation-cap"></i>
                                    </span>
                                    <span>Other CSB courses</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://twitter.com/cs_biology">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fab fa-twitter"></i>
                                    </span>
                                    <span>Get in touch on twitter</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://github.com/CSBiology?type=source">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fab fa-github"></i>
                                    </span>
                                    <span>Check out our open source projects</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://github.com/CSBiology/BIO-BTE-06-L-7">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fas fa-code-branch"></i>
                                    </span>
                                    <span>Source repository</span>
                                </span>
                            </a>
                        </li>
                    </ul>
                </div>
            </aside>
        </div>
        <div class="column is-10 is-paddingless has-text-justified	">
            <div class="section pt-4">
                <div class="container" id="fsdocs-content">
                    <h1><a name="NB01b-Plant-Systems-Biology" class="anchor" href="#NB01b-Plant-Systems-Biology">NB01b Plant Systems Biology</a></h1>
<p><a href="https://mybinder.org/v2/gh/CSBiology/BIO-BTE-06-L-7/gh-pages?filepath=NB01b_Plant_Systems_Biology.ipynb"><img src="https://mybinder.org/badge_logo.svg" alt="Binder" /></a></p>
<p><a href="https://github.com/CSBiology/BIO-BTE-06-L-7/releases/download/NB01a_NB01b/NB01b_Plant_Systems_Biology.ipynb">Download Notebook</a></p>
<ol>
<li>Plant Systems Biology</li>
<li>Modeling growth</li>
<li>Insert growth data and display as chart</li>
<li>Calculation of growth rate and doubling time for cell cultures</li>
<li>
Fitting biological growth curves
<ol>
<li>Theory</li>
<li>Model selection</li>
</ol>
</li>
<li>Calculate cell doubling time</li>
<li>Questions</li>
<li>References</li>
</ol>
<h2><a name="Plant-Systems-Biology" class="anchor" href="#Plant-Systems-Biology">Plant Systems Biology</a></h2>
<p>The general paradigm of Systems Biology clearly applies to plants, as they represent complex biological systems.
The functioning of a plant as a biological system is the result of a combination of multiple intertwined and dynamic interactions between its components.
In addition, most plants are sessile systems that have to face fluctuating environmental conditions, including biotic and abiotic stresses (Ruffel et al. 2010).
The process of a biological system responding to changes in environmental conditions is termed acclimation. These molecular physiological responses represent a complex
dynamic adjustment of the interplay between genes, proteins and metabolites that allows the organism to acclimate to the changing environment.
The ability to acclimate ensures the survival of all living organisms and is therefore fundamental for the understanding of biological systems.
Detailed knowledge about how plants acclimate to a changing environment is crucial especially in times of global climate change,
as plants are of great importance for our quality of life as a key source of food, shelter, fiber, medicine, and fuel (Minorsky 2003).</p>
<p>The prominent model plant <em>Arabidopsis thaliana</em> is well suited for Plant Systems Biology studies because sophisticated experimental tools and extensive data
collections are readily available (Van Norman et al. 2009). However, the importance of a model organism is not only coined by the availability of molecular
tools to manipulate the organism, but also by its agricultural and economic impact like in the cases of tobacco, rice, maize or
barley (Pãcurar 2009). Also, microalgae are of special economic interest due to their potential as biofuel producers (Cagnon et al. 2013).
Additionally, the use of organisms with lower biological complexity facilitates the feasibility of System Biology studies and is an important factor to consider
for the choice of a suitable model organism in Systems Biology.</p>
<p>The eukaryotic green alga <em>Chlamydomonas reinhardtii</em> is particularly well suited for Plant Systems Biology approaches.
This unicellular freshwater and soil-dwelling alga has a single, cup-shaped chloroplast with a photosynthetic apparatus that is similar to
that of higher plants (Eberhard et al. 2008, Merchant et al. 2007). Hence, results gained on photosynthesis processes in <em>Chlamydomonas</em>
are likely to be transferable to higher plants. The nuclear, mitochondrial, and chloroplast genomes have been sequenced and tools for manipulating them
are available (Merchant et al. 2007). <em>Chlamydomonas</em> cells have a size of ~10 µm and grow under photo-, mixo-, and heterotrophic conditions
with a generation time of ~5-8 h (Harris, 2008). <em>Chlamydomonas</em> can be maintained under controlled conditions and environmental
changes can be applied homogeneously and rapidly to all cells in a liquid culture. In contrast to multicellular organisms there are no influences by
tissue heterogeneity. Even the influence of different cell cycle stages may be ruled out by performing experiments with asynchronous cell cultures
(Bruggeman and Westerhoff 2007, Harris 2001). Finally, gene families in <em>Chlamydomonas</em> have fewer members than those in higher plants thus facilitating the
interpretation of results involving many genes/proteins (Merchant et al. 2007).</p>
<h2><a name="Modeling-growth" class="anchor" href="#Modeling-growth">Modeling growth</a></h2>
<p>In order to solve real world tasks more convenient, F# provides a huge collection of additional programming libraries.
Anything that extends beyond the basics must be written by a user. If the chunk of code is useful to multiple different users,
it's often put into a library to make it easily reusable. A library is a collection of related pieces of code that have been compiled
and stored together in a single file and can than be used an included. The most important libraries in F# for bioinformatics are:</p>
<ul>
<li><a href="https://csbiology.github.io/BioFSharp/">BioFSharp</a>: Open source bioinformatics and computational biology toolbox written in F#</li>
<li><a href="https://github.com/fslaborg/FSharp.Stats">FSharp.Stats</a>: F# project for statistical computing</li>
<li><a href="https://github.com/plotly/Plotly.NET">Plotly.NET</a>: .NET interface for plotly.js written in F# 📈</li>
</ul>
<p>The first real world use case of F# in Systems Biology is to model growth for a defined cell number to see possible overexpression effects.
Biologists often utilize growth experiments to analyze basic properties of a given organism or cellular model. For a solid comparison of data
obtained from different experiments and to investigate the speciﬁc effect of a given experimental set up, modeling the growth is needed after recording the data.</p>
<p>This notebook introduces the most classical way to model growth of <em>Chlamydomonas reinhardtii</em> or any other growth data using F#.</p>
<p>Now, let's get started by loading the required libraries first.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="pp">#r</span> <span class="s">&quot;nuget: FSharp.Stats, 0.4.3&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: Plotly.NET, 4.2.0&quot;</span>

<span class="pp">#if</span> <span class="id">IPYNB</span>
<span class="inactive">#r</span><span class="inactive"> </span><span class="inactive">&quot;nuget:</span><span class="inactive"> </span><span class="inactive">Plotly.NET.Interactive,</span><span class="inactive"> </span><span class="inactive">4.2.0&quot;</span>
<span class="pp">#endif</span> <span class="c">// IPYNB</span>

<span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="id">System</span>
<span class="k">open</span> <span class="id">Plotly</span><span class="pn">.</span><span class="id">NET</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs2', 2)" onmouseover="showTip(event, 'fs2', 2)" class="id">FSharp</span><span class="pn">.</span><span class="id">Stats</span>
</code></pre>
<h2><a name="Insert-growth-data-and-display-as-chart" class="anchor" href="#Insert-growth-data-and-display-as-chart">Insert growth data and display as chart</a></h2>
<p>A standard cell culture experiment with cell count measurements will result in data like the following.
Multiple cell counts (<code>y_Count</code>), each related to a specific timepoint (<code>x_Hours</code>).</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 1</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs3', 3)" onmouseover="showTip(event, 'fs3', 3)" class="id">exmp_x_Hours</span> <span class="o">=</span> <span class="pn">[|</span><span class="n">0.</span><span class="pn">;</span> <span class="n">19.5</span><span class="pn">;</span> <span class="n">25.5</span><span class="pn">;</span> <span class="n">43.</span><span class="pn">;</span> <span class="n">48.5</span><span class="pn">;</span> <span class="n">67.75</span><span class="pn">|]</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs4', 4)" onmouseover="showTip(event, 'fs4', 4)" class="id">exmp_y_Count</span> <span class="o">=</span> <span class="pn">[|</span><span class="n">1659000.</span><span class="pn">;</span> <span class="n">4169000.</span><span class="pn">;</span> <span class="n">6585400.</span><span class="pn">;</span> <span class="n">16608400.</span><span class="pn">;</span> <span class="n">17257800.</span><span class="pn">;</span> <span class="n">18041000.</span><span class="pn">|]</span>

<span class="c">// Such data can easily be display with the following code block.</span>
<span class="c">// Chart.Point takes a sequence of x-axis-points and a series of y-axis-points as input</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs5', 5)" onmouseover="showTip(event, 'fs5', 5)" class="id">example_Chart_1</span> <span class="o">=</span> 
    <span class="id">Chart</span><span class="pn">.</span><span class="id">Point</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs3', 6)" onmouseover="showTip(event, 'fs3', 6)" class="id">exmp_x_Hours</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs4', 7)" onmouseover="showTip(event, 'fs4', 7)" class="id">exmp_y_Count</span><span class="pn">)</span>
    <span class="c">// some minor styling with title and axis-titles.</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withTitle</span> <span class="s">&quot;Growth curve of Chlamydomonas reinhardtii cell cultures&quot;</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withYAxisStyle</span> <span class="pn">(</span><span class="s">&quot;Number of cells&quot;</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withXAxisStyle</span> <span class="pn">(</span><span class="s">&quot;Time [hours]&quot;</span><span class="pn">)</span>

<span onmouseout="hideTip(event, 'fs5', 8)" onmouseover="showTip(event, 'fs5', 8)" class="id">example_Chart_1</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<h2><a name="Calculation-of-growth-rate-and-doubling-time-for-cell-cultures" class="anchor" href="#Calculation-of-growth-rate-and-doubling-time-for-cell-cultures">Calculation of growth rate and doubling time for cell cultures</a></h2>
<p>The standard growth of an in vitro cell culture is defined by three phases. The lag phase in which the cells still acclimate to the growth conditions, the exponential growth,
also called log phase, during which cell growth is exponential due to the iterative proliferation of cells into two daughter cells, and the stationary phase in which the growth rate and the
death rate are equal. The stationary phase is typically initiated due to limitations in growth conditions, e.g. depletion of essential nutrients or accumulation of toxic/inhibitory
excretions/products. The doubling time (or generation time) defines a time interval in which the quantity of cells doubles.</p>
<p><img src="https://raw.githubusercontent.com/CSBiology/BIO-BTE-06-L-7/main/docs/img/growthCurve.png" alt="" /><br />
<strong><em>Figure 2: An exemplary growth curve.</em></strong></p>
<p>Growth data always should be visualized in log space. Therefore the count data must be log transformed. When a log2 transform is applied,
a doubling of the original counts is achieved, when the value increase 1 unit.
Keeping that in mind, the slope of the growth function can be used to calculate the time it takes for the log transformed data to increase 1 unit.</p>
<p>The corresponding chart of the log transformed count data looks like this:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 2</span>

<span class="c">// log transform the count data with a base of 2</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs6', 10)" onmouseover="showTip(event, 'fs6', 10)" class="id">exmp_y_Count_Log</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs4', 11)" onmouseover="showTip(event, 'fs4', 11)" class="id">exmp_y_Count</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs7', 12)" onmouseover="showTip(event, 'fs7', 12)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs8', 13)" onmouseover="showTip(event, 'fs8', 13)" class="id">map</span> <span class="id">log2</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs9', 14)" onmouseover="showTip(event, 'fs9', 14)" class="id">example_Chart_2</span> <span class="o">=</span> 
    <span class="id">Chart</span><span class="pn">.</span><span class="id">Point</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs3', 15)" onmouseover="showTip(event, 'fs3', 15)" class="id">exmp_x_Hours</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs6', 16)" onmouseover="showTip(event, 'fs6', 16)" class="id">exmp_y_Count_Log</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withTitle</span> <span class="s">&quot;Growth curve of Chlamydomonas reinhardtii cell cultures&quot;</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withYAxisStyle</span> <span class="pn">(</span><span class="s">&quot;Number of cells [log2]&quot;</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withXAxisStyle</span> <span class="pn">(</span><span class="s">&quot;Time [hours]&quot;</span><span class="pn">)</span>

<span onmouseout="hideTip(event, 'fs9', 17)" onmouseover="showTip(event, 'fs9', 17)" class="id">example_Chart_2</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<p>After the log transform the exponential phase becomes linear. Since a y axis difference of 1 corresponds to a doubling of the cells the generation time can simply be estimated by determination of how many hours are required for the data to span one y axis unit.
In this case it seems, that the time required to get from y=22 to y=23 takes approximately 10 hours.</p>
<p>As you may noticed we just determined the cell doubling time per eye.
The formalization of this process is trivial.</p>
<ul>
<li>The y-value increment that is required for doubling can be calculated by log_x(2) where x defines the used base. So for log2 transformed data it is 1 (log2(2)).</li>
<li>The generation time is calculated by dividing this y-value increment by the growth rate which is the steepest slope of the log transformed data (approximately 0.1 in the given data). The steepest slope in growth curves always occurs at the inflection point of the sigmoidal function shape.</li>
</ul>
<p>So all we have to know is the performed log transform and the slope of the function at its steepest point and afterwards apply the following equation.</p>
<p><em>Equation 1: Calculation of the doubling time. Growth rate is the steepest slope of the log transformed count data.</em></p>
<p><img src="https://latex.codecogs.com/png.latex?doubling&amp;space;Time&amp;space;=&amp;space;\frac{log_x(2)}{growthRate}" alt="" /></p>
<p>For a log2 transform the numerator is 1.</p>
<h2><a name="Fitting-biological-growth-curves" class="anchor" href="#Fitting-biological-growth-curves">Fitting biological growth curves</a></h2>
<br>
<h3><a name="Theory" class="anchor" href="#Theory">Theory</a></h3>
<p>To derive the slope required for the doubling time calculation, the measured growth data points have to be modelled.
In order to obtain a continuous function with known coefficients, a suitable model function is fitted onto the existing data.
Many models exist, each one of them optimized for a specific task (Kaplan et al. 2018).</p>
<p>Linear model function example: <img src="https://latex.codecogs.com/png.latex?f(x)&amp;space;=&amp;space;mx&amp;space;&amp;plus;&amp;space;b" alt="" /></p>
<p>When a model function is fitted onto the data, there are endless possibilities to choose coefficients of the model function.
In the case above there are two coefficients to be identified: The slope m and the y-intercept b. But how can the best fitting coefficients be determined?</p>
<p>Therefore a quality measure called <strong><em>Residual Sum of Squares (RSS)</em></strong> is used. It describes the discrepancy of the measured points
and the corresponding estimation model. If the discrepancy is small, the RSS is small too.</p>
<p>In regression analysis the optimal set of coefficients (m and b) that <a href="https://mathworld.wolfram.com/LeastSquaresFitting.html">minimizes the RSS is searched</a>.</p>
<p>If there is no straightforward way to identify the RSS-minimizing coefficient set, then the problem is part of nonlinear regression.
Here, initial coefficients are guessed and the RSS is calculated. Thereafter, the coefficients are modified in tiny steps.
If the RSS decreases, the direction of the coefficient change seems to be correct.
By <a href="https://books.google.de/books?id=rs51DwAAQBAJ&amp;pg=PA422&amp;lpg=PA422&amp;dq=rss+minimizing+solver&amp;source=bl&amp;ots=qZ0Y4cYtM-&amp;sig=ACfU3U0rHGWCmTo_kv5wqYMmSo8ZKyj5Pg&amp;hl=de&amp;sa=X&amp;ved=2ahUKEwjKtdf-oaHoAhUUwsQBHX07DTwQ6AEwBHoECAkQAQ#v=onepage&amp;q=rss%20minimizing%20solver&amp;f=false">iteratively changing coefficients</a>
, the optimal coefficient set is determined when no further change leads to an decrease in RSS.
Algorithms, that perform such a 'gradient descent' methods to solve nonlinear regression tasks are called <strong><em>solver</em></strong>
(e.g. Gauss-Newton algorithm or Levenberg–Marquardt algorithm). <a href="https://www.youtube.com/watch?v=sDv4f4s2SB8">Introduction to RSS and optimization problems.</a></p>
<h3><a name="Model-selection" class="anchor" href="#Model-selection">Model selection</a></h3>
<p>Depending on the given problem, different models can be fitted to the data. Several growth models exist, each is specialized for a particular problem. See <a href="http://www.pisces-conservation.com/growthhelp/index.html">Types of growth curve</a> or <a href="https://fslab.org/FSharp.Stats/GrowthCurve.html">FSharp.Stats - Growth Curve</a> for more information.</p>
<p>The selected model should match the theoretical (time) course of the studied signal, but under
consideration of Occams razor principle. It states, that a approriate model with a low number of coefficients should be preferred over a
model with many coefficients, since the excessive use of coefficients leads to overfitting.</p>
<p>An often used growth curve model is the four parameter <a href="https://en.wikipedia.org/wiki/Gompertz_function">Gompertz model</a>.</p>
<p>The function has the form: <img src="https://latex.codecogs.com/png.latex?A+Ce^{-e^{-B(t-M)}}" alt="" /> <a href="https://www.sciencedirect.com/science/article/pii/0168160588900517?via%3Dihub">Gibson et al. 1988</a>.</p>
<p>where:</p>
<ul>
<li>A = curve minimum</li>
<li>B = relative growth rate (not to be confused with absolute)</li>
<li>C = curve maximum - curve minimum (y range)</li>
<li>M = x position of inflection point</li>
<li>t = time point</li>
</ul>
<p>In the following, we will go through the necessary steps to calculate the generation time with the help of a Gompertz model.
While the curve minimum and maximum are easy to define by eye, to estimate the remaining coefficients is a nontrivial task.</p>
<p>FSharp.Stats provides a function, that estimates the model coefficients from the data and a guess of the expected generation time. For Chlamydomonas the initial guess would be 8 hours.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 3</span>

<span class="c">// open module in FSharp.Stats to perform nonlinear regression</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs2', 19)" onmouseover="showTip(event, 'fs2', 19)" class="id">FSharp</span><span class="pn">.</span><span class="id">Stats</span><span class="pn">.</span><span class="id">Fitting</span><span class="pn">.</span><span class="id">NonLinearRegression</span>

<span class="c">// The model we need already exists in FSharp.Stats and can be taken from the &quot;Table&quot; module.</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs10', 20)" onmouseover="showTip(event, 'fs10', 20)" class="id">modelGompertz</span> <span class="o">=</span> <span class="id">Table</span><span class="pn">.</span><span class="id">GrowthModels</span><span class="pn">.</span><span class="id">gompertz</span>

<span class="c">// The solver that iteratively optimizes the coefficients requires an initial guess of the coefficients.</span>
<span class="c">// The following function was specifically designed to estimate gompertz model coefficients from the data</span>
<span class="c">// You have to provide the time data, the log transformed count data, the expected generation time, and the used log transform</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs11', 21)" onmouseover="showTip(event, 'fs11', 21)" class="id">solverOptions</span> <span class="o">=</span> <span class="id">Table</span><span class="pn">.</span><span class="id">GrowthModels</span><span class="pn">.</span><span class="id">getSolverOptionsGompertz</span> <span onmouseout="hideTip(event, 'fs3', 22)" onmouseover="showTip(event, 'fs3', 22)" class="id">exmp_x_Hours</span> <span onmouseout="hideTip(event, 'fs6', 23)" onmouseover="showTip(event, 'fs6', 23)" class="id">exmp_y_Count_Log</span> <span class="n">8.</span> <span class="id">log2</span>

<span class="c">// sequence of initial guess coefficients</span>
<span onmouseout="hideTip(event, 'fs11', 24)" onmouseover="showTip(event, 'fs11', 24)" class="id">solverOptions</span><span class="pn">.</span><span class="id">InitialParamGuess</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<p>The initial coefficient estimations match the expectations.</p>
<ul>
<li>A = 20.7</li>
<li>B = 0.099</li>
<li>C = 3.4</li>
<li>M = 19.5</li>
</ul>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 4</span>

<span class="c">// By solving the nonlinear fitting problem, the optimal model coefficients are determined. </span>
<span class="c">// The solver iteratively changes the coefficients until the model function fits the data best.</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs12', 26)" onmouseover="showTip(event, 'fs12', 26)" class="id">gompertzParams</span> <span class="o">=</span>
    <span class="id">LevenbergMarquardt</span><span class="pn">.</span><span class="id">estimatedParams</span> <span class="c">// The Levenberg Marquardt is used as solver</span>
        <span onmouseout="hideTip(event, 'fs10', 27)" onmouseover="showTip(event, 'fs10', 27)" class="id">modelGompertz</span>    <span class="c">// The gompertz model is used as growth model</span>
        <span onmouseout="hideTip(event, 'fs11', 28)" onmouseover="showTip(event, 'fs11', 28)" class="id">solverOptions</span>    <span class="c">// The initial guess of the coefficients</span>
        <span class="n">0.1</span>              <span class="c">// Parameter required from the solver</span>
        <span class="n">10.</span>              <span class="c">// Parameter required from the solver</span>
        <span onmouseout="hideTip(event, 'fs3', 29)" onmouseover="showTip(event, 'fs3', 29)" class="id">exmp_x_Hours</span>     <span class="c">// The time data</span>
        <span onmouseout="hideTip(event, 'fs6', 30)" onmouseover="showTip(event, 'fs6', 30)" class="id">exmp_y_Count_Log</span> <span class="c">// The transformed count data</span>

<span onmouseout="hideTip(event, 'fs12', 31)" onmouseover="showTip(event, 'fs12', 31)" class="id">gompertzParams</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<p>The model coefficients were determined to be:</p>
<ul>
<li>A = 20.66</li>
<li>B = 0.107</li>
<li>C = 3.52</li>
<li>M = 19.63</li>
</ul>
<p>They are pretty close to the initial estimations that were determined in Code-Block 3 With the coefficients at hand, the model function can be filled with coefficients and can be used to create a fit to the data.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 5</span>

<span class="c">// Create fitting function from optimal coefficients</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs13', 33)" onmouseover="showTip(event, 'fs13', 33)" class="fn">fittingFunction</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs10', 34)" onmouseover="showTip(event, 'fs10', 34)" class="id">modelGompertz</span><span class="pn">.</span><span class="id">GetFunctionValue</span> <span onmouseout="hideTip(event, 'fs12', 35)" onmouseover="showTip(event, 'fs12', 35)" class="id">gompertzParams</span>

<span class="c">// Fit the optimized model function to all x values from 0 to 70 hours.</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs14', 36)" onmouseover="showTip(event, 'fs14', 36)" class="id">fittedValuesGompertz</span> <span class="o">=</span>
    <span class="pn">[</span><span class="o">0.</span> <span class="o">..</span> <span class="n">0.1</span> <span class="o">..</span> <span class="n">70.</span><span class="pn">]</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs15', 37)" onmouseover="showTip(event, 'fs15', 37)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs16', 38)" onmouseover="showTip(event, 'fs16', 38)" class="id">map</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs17', 39)" onmouseover="showTip(event, 'fs17', 39)" class="fn">x</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs17', 40)" onmouseover="showTip(event, 'fs17', 40)" class="fn">x</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs13', 41)" onmouseover="showTip(event, 'fs13', 41)" class="fn">fittingFunction</span> <span onmouseout="hideTip(event, 'fs17', 42)" onmouseover="showTip(event, 'fs17', 42)" class="fn">x</span><span class="pn">)</span> 
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">Line</span>
    

<span class="c">// combine the raw data and the fit into one chart</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs18', 43)" onmouseover="showTip(event, 'fs18', 43)" class="id">fittedChartGompertz</span> <span class="o">=</span> 
    <span class="pn">[</span>
        <span onmouseout="hideTip(event, 'fs9', 44)" onmouseover="showTip(event, 'fs9', 44)" class="id">example_Chart_2</span>      <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withTraceInfo</span> <span class="s">&quot;raw data&quot;</span>
        <span onmouseout="hideTip(event, 'fs14', 45)" onmouseover="showTip(event, 'fs14', 45)" class="id">fittedValuesGompertz</span> <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withTraceInfo</span> <span class="s">&quot;gompertz model&quot;</span>
    <span class="pn">]</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">combine</span>

<span onmouseout="hideTip(event, 'fs18', 46)" onmouseover="showTip(event, 'fs18', 46)" class="id">fittedChartGompertz</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<h2><a name="Calculate-cell-doubling-time" class="anchor" href="#Calculate-cell-doubling-time">Calculate cell doubling time</a></h2>
<p>To calculate the doubling time it is necessary to determine the growth rate (gr) for <em>equation 1</em>.
As discussed above the growth rate is the maximal slope of the model function. It always occurs at the inflection point, which we know is at x=19.628.
After calculating the first derivative of the model function, we would be able to calculate the growth rate as the slope at the inflection point.
Luckily, there is a short cut when using the Gompertz model. It allows the determination of generation times from its parameters (see Gibson et al. 1988 for details).</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 6</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs19', 48)" onmouseover="showTip(event, 'fs19', 48)" class="fn">getGenTimeFromGompertz</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs20', 49)" onmouseover="showTip(event, 'fs20', 49)" class="fn">parameterVector</span><span class="pn">:</span> <span class="id">vector</span><span class="pn">)</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs21', 50)" onmouseover="showTip(event, 'fs21', 50)" class="fn">logTransform</span><span class="pn">:</span> <span onmouseout="hideTip(event, 'fs22', 51)" onmouseover="showTip(event, 'fs22', 51)" class="vt">float</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs22', 52)" onmouseover="showTip(event, 'fs22', 52)" class="vt">float</span><span class="pn">)</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs21', 53)" onmouseover="showTip(event, 'fs21', 53)" class="fn">logTransform</span> <span class="n">2.</span> <span class="o">*</span> <span onmouseout="hideTip(event, 'fs23', 54)" onmouseover="showTip(event, 'fs23', 54)" class="id">Math</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs24', 55)" onmouseover="showTip(event, 'fs24', 55)" class="id">E</span> <span class="o">/</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs20', 56)" onmouseover="showTip(event, 'fs20', 56)" class="fn">parameterVector</span><span class="pn">.</span><span class="pn">[</span><span class="n">1</span><span class="pn">]</span> <span class="o">*</span> <span onmouseout="hideTip(event, 'fs20', 57)" onmouseover="showTip(event, 'fs20', 57)" class="fn">parameterVector</span><span class="pn">.</span><span class="pn">[</span><span class="n">2</span><span class="pn">]</span><span class="pn">)</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs25', 58)" onmouseover="showTip(event, 'fs25', 58)" class="id">genTime</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs19', 59)" onmouseover="showTip(event, 'fs19', 59)" class="fn">getGenTimeFromGompertz</span> <span onmouseout="hideTip(event, 'fs12', 60)" onmouseover="showTip(event, 'fs12', 60)" class="id">gompertzParams</span> <span class="id">log2</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs26', 61)" onmouseover="showTip(event, 'fs26', 61)" class="id">gt</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs27', 62)" onmouseover="showTip(event, 'fs27', 62)" class="fn">sprintf</span> <span class="s">&quot;The generation time is </span><span class="pf">%.2f</span><span class="s"> hours.&quot;</span> <span onmouseout="hideTip(event, 'fs25', 63)" onmouseover="showTip(event, 'fs25', 63)" class="id">genTime</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<br>
<h2><a name="Questions" class="anchor" href="#Questions">Questions:</a></h2>
<ol>
<li>
<p>Why is it useful to use a log2 transform rather than a ln, log10, or any other log transform?<br />
<em>Hint:</em> Define your own exponentially growing cell counts with a generation time of 1 and transform them using different log transforms.</p>
</li>
<li>
<p>Why is it not sufficient to fit the (raw or transformed) data using the possibilities Excel offers?<br />
<em>Hint:</em> Which models are available and why are these not always appropriate?</p>
</li>
<li>Calculate the generation time of the following data. Compare the time points of maximal slope of the raw data and the transformed data by eye. Without the log transform you are blind for the actual point of maximal growth.</li>
</ol>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs28', 65)" onmouseover="showTip(event, 'fs28', 65)" class="id">rawX_hours</span> <span class="o">=</span> <span class="pn">[|</span><span class="n">0.</span> <span class="o">..</span> <span class="n">12.</span><span class="pn">|]</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs29', 66)" onmouseover="showTip(event, 'fs29', 66)" class="id">rawY_count</span> <span class="o">=</span> <span class="pn">[|</span><span class="n">2.</span><span class="pn">;</span><span class="n">2.2</span><span class="pn">;</span><span class="n">2.9</span><span class="pn">;</span><span class="n">5.</span><span class="pn">;</span><span class="n">9.5</span><span class="pn">;</span><span class="n">19.</span><span class="pn">;</span><span class="n">38.</span><span class="pn">;</span><span class="n">65.</span><span class="pn">;</span><span class="n">85.</span><span class="pn">;</span><span class="n">90.</span><span class="pn">;</span><span class="n">91.</span><span class="pn">;</span><span class="n">91.</span><span class="pn">;</span><span class="n">91.</span><span class="pn">;</span><span class="pn">|]</span>
</code></pre>
<h2><a name="References" class="anchor" href="#References">References</a></h2>
<ol>
<li>Ruffel, S., Krouk, G. &amp; Coruzzi, G. M. A systems view of responses to nutritional cues in Arabidopsis: toward a paradigm shift for predictive network modeling. Plant physiology 152, 445–452; 10.1104/pp.109.148502 (2010).</li>
<li>Minorsky, P. V. Achieving the in Silico Plant. Systems Biology and the Future of Plant Biological Research. Plant physiology 132, 404–409; 10.1104/pp.900076 (2003).</li>
<li>Van Norman, Jaimie M &amp; Benfey, P. N. Arabidopsis thaliana as a model organism in systems biology. Wiley interdisciplinary reviews. Systems biology and medicine 1, 372–379; 10.1002/wsbm.25 (2009).</li>
<li>Pãcurar, D. I. Model organisms - a journey from the dawn of biological research to the post-genomic era. Romanian Society of Biological Sciences, 4087–4094 (2009).</li>
<li>Cagnon, C. et al. Development of a forward genetic screen to isolate oil mutants in the green microalga Chlamydomonas reinhardtii. Biotechnology for biofuels 6, 178; 10.1186/1754-6834-6-178 (2013).</li>
<li>Eberhard, S., Finazzi, G. &amp; Wollman, F.-A. The dynamics of photosynthesis. Annual review of genetics 42, 463–515; 10.1146/annurev.genet.42.110807.091452 (2008).</li>
<li>Merchant, S. S. et al. The Chlamydomonas genome reveals the evolution of key animal and plant functions. Science (New York, N.Y.) 318, 245–250; 10.1126/science.1143609 (2007).</li>
<li>Harris, E. H. The chlamydomonas sourcebook. 2nd ed. (Academic, London, 2008).</li>
<li>Bruggeman, F. J. &amp; Westerhoff, H. V. The nature of systems biology. Trends in microbiology 15, 45–50; 10.1016/j.tim.2006.11.003 (2007).</li>
<li>Harris, E. H. CHLAMYDOMONAS AS A MODEL ORGANISM. Annual review of plant physiology and plant molecular biology 52, 363–406; 10.1146/annurev.arplant.52.1.363 (2001).</li>
<li>Kaplan, S. et al. Comparison of growth curves using non-linear regression function in Japanese squail. Journal of Applied Animal Research 46, 112-117; 10.1080/09712119.2016.1268965 (2018).</li>
<li>Gibson, A., Bratchell, N., Roberts, T.A., Predicting microbial growth: growth responses of salmonellae in a laboratory medium as affected by pH, sodium chloride and storage temperature, International Journal of Food Microbiology, Volume 6, Issue 2,  <a href="https://doi.org/10.1016/0168-1605(88)90051-7">https://doi.org/10.1016/0168-1605(88)90051-7</a> (1988).</li>
</ol>

                </div>
            </div>
        </div>
        <div class="fsdocs-tip" id="fs1">namespace System</div>
<div class="fsdocs-tip" id="fs2">namespace Microsoft.FSharp</div>
<div class="fsdocs-tip" id="fs3">val exmp_x_Hours : float []</div>
<div class="fsdocs-tip" id="fs4">val exmp_y_Count : float []</div>
<div class="fsdocs-tip" id="fs5">val example_Chart_1 : obj</div>
<div class="fsdocs-tip" id="fs6">val exmp_y_Count_Log : obj []</div>
<div class="fsdocs-tip" id="fs7">type Array =
  interface ICollection
  interface IEnumerable
  interface IList
  interface IStructuralComparable
  interface IStructuralEquatable
  interface ICloneable
  new : unit -&gt; unit
  member Clone : unit -&gt; obj
  member CopyTo : array: Array * index: int -&gt; unit + 1 overload
  member GetEnumerator : unit -&gt; IEnumerator
  ...<br /><em>&lt;summary&gt;Provides methods for creating, manipulating, searching, and sorting arrays, thereby serving as the base class for all arrays in the common language runtime.&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs8">val map : mapping:(&#39;T -&gt; &#39;U) -&gt; array:&#39;T [] -&gt; &#39;U []<br /><em>&lt;summary&gt;Builds a new array whose elements are the results of applying the given function
 to each of the elements of the array.&lt;/summary&gt;<br />&lt;param name=&quot;mapping&quot;&gt;The function to transform elements of the array.&lt;/param&gt;<br />&lt;param name=&quot;array&quot;&gt;The input array.&lt;/param&gt;<br />&lt;returns&gt;The array of transformed elements.&lt;/returns&gt;<br />&lt;exception cref=&quot;T:System.ArgumentNullException&quot;&gt;Thrown when the input array is null.&lt;/exception&gt;</em></div>
<div class="fsdocs-tip" id="fs9">val example_Chart_2 : obj</div>
<div class="fsdocs-tip" id="fs10">val modelGompertz : obj</div>
<div class="fsdocs-tip" id="fs11">val solverOptions : obj</div>
<div class="fsdocs-tip" id="fs12">val gompertzParams : obj</div>
<div class="fsdocs-tip" id="fs13">val fittingFunction : (float -&gt; obj)</div>
<div class="fsdocs-tip" id="fs14">val fittedValuesGompertz : obj</div>
<div class="fsdocs-tip" id="fs15">module Seq

from Microsoft.FSharp.Collections<br /><em>&lt;summary&gt;Contains operations for working with values of type &lt;see cref=&quot;T:Microsoft.FSharp.Collections.seq`1&quot; /&gt;.&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs16">val map : mapping:(&#39;T -&gt; &#39;U) -&gt; source:seq&lt;&#39;T&gt; -&gt; seq&lt;&#39;U&gt;<br /><em>&lt;summary&gt;Builds a new collection whose elements are the results of applying the given function
 to each of the elements of the collection.  The given function will be applied
 as elements are demanded using the &lt;c&gt;MoveNext&lt;/c&gt; method on enumerators retrieved from the
 object.&lt;/summary&gt;<br />&lt;remarks&gt;The returned sequence may be passed between threads safely. However, 
 individual IEnumerator values generated from the returned sequence should not be accessed concurrently.&lt;/remarks&gt;<br />&lt;param name=&quot;mapping&quot;&gt;A function to transform items from the input sequence.&lt;/param&gt;<br />&lt;param name=&quot;source&quot;&gt;The input sequence.&lt;/param&gt;<br />&lt;returns&gt;The result sequence.&lt;/returns&gt;<br />&lt;exception cref=&quot;T:System.ArgumentNullException&quot;&gt;Thrown when the input sequence is null.&lt;/exception&gt;</em></div>
<div class="fsdocs-tip" id="fs17">val x : float</div>
<div class="fsdocs-tip" id="fs18">val fittedChartGompertz : obj</div>
<div class="fsdocs-tip" id="fs19">val getGenTimeFromGompertz : parameterVector:&#39;a -&gt; logTransform:(float -&gt; float) -&gt; float</div>
<div class="fsdocs-tip" id="fs20">val parameterVector : &#39;a</div>
<div class="fsdocs-tip" id="fs21">val logTransform : (float -&gt; float)</div>
<div class="fsdocs-tip" id="fs22">Multiple items<br />val float : value:&#39;T -&gt; float (requires member op_Explicit)<br /><em>&lt;summary&gt;Converts the argument to 64-bit float. This is a direct conversion for all 
 primitive numeric types. For strings, the input is converted using &lt;c&gt;Double.Parse()&lt;/c&gt;  
 with InvariantCulture settings. Otherwise the operation requires an appropriate
 static conversion method on the input type.&lt;/summary&gt;<br />&lt;param name=&quot;value&quot;&gt;The input value.&lt;/param&gt;<br />&lt;returns&gt;The converted float&lt;/returns&gt;</em><br /><br />--------------------<br />[&lt;Struct&gt;]
type float = Double<br /><em>&lt;summary&gt;An abbreviation for the CLI type &lt;see cref=&quot;T:System.Double&quot; /&gt;.&lt;/summary&gt;<br />&lt;category&gt;Basic Types&lt;/category&gt;</em><br /><br />--------------------<br />type float&lt;&#39;Measure&gt; =
  float<br /><em>&lt;summary&gt;The type of double-precision floating point numbers, annotated with a unit of measure. 
 The unit of measure is erased in compiled code and when values of this type
 are analyzed using reflection. The type is representationally equivalent to 
 &lt;see cref=&quot;T:System.Double&quot; /&gt;.&lt;/summary&gt;<br />&lt;category index=&quot;6&quot;&gt;Basic Types with Units of Measure&lt;/category&gt;</em></div>
<div class="fsdocs-tip" id="fs23">type Math =
  static member Abs : value: decimal -&gt; decimal + 6 overloads
  static member Acos : d: float -&gt; float
  static member Acosh : d: float -&gt; float
  static member Asin : d: float -&gt; float
  static member Asinh : d: float -&gt; float
  static member Atan : d: float -&gt; float
  static member Atan2 : y: float * x: float -&gt; float
  static member Atanh : d: float -&gt; float
  static member BigMul : a: int * b: int -&gt; int64 + 2 overloads
  static member BitDecrement : x: float -&gt; float
  ...<br /><em>&lt;summary&gt;Provides constants and static methods for trigonometric, logarithmic, and other common mathematical functions.&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs24">field Math.E: float = 2.71828182846<br /><em>&lt;summary&gt;Represents the natural logarithmic base, specified by the constant, &lt;see langword=&quot;e&quot; /&gt;.&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs25">val genTime : float</div>
<div class="fsdocs-tip" id="fs26">val gt : string</div>
<div class="fsdocs-tip" id="fs27">val sprintf : format:Printf.StringFormat&lt;&#39;T&gt; -&gt; &#39;T<br /><em>&lt;summary&gt;Print to a string using the given format.&lt;/summary&gt;<br />&lt;param name=&quot;format&quot;&gt;The formatter.&lt;/param&gt;<br />&lt;returns&gt;The formatted result.&lt;/returns&gt;</em></div>
<div class="fsdocs-tip" id="fs28">val rawX_hours : float []</div>
<div class="fsdocs-tip" id="fs29">val rawY_count : float []</div>

    </div>

    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <script type="text/javascript">var fsdocs_search_baseurl = '/BIO-BTE-06-L-7/'</script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/lunr.js/2.3.8/lunr.min.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.min.js"></script>
    <script type="text/javascript" src="/BIO-BTE-06-L-7/content/fsdocs-search.js"></script>
    <!-- END SEARCH BOX: this adds support for the search box -->
</body>

</html>