<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <title>NB02c Isotopic Distribution
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="author" content="">
    <link rel="shortcut icon" type="image/x-icon" href="/BIO-BTE-06-L-7//img/favicon.ico">
    <script src="https://code.jquery.com/jquery-3.4.1.min.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <!-- <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/MathJax.js?config=TeX-MML-AM_CHTML"></script> -->

    <link type="text/css" rel="stylesheet" href="/BIO-BTE-06-L-7/content/fsdocs-default.css" />
    <link type="text/css" rel="stylesheet" href="/BIO-BTE-06-L-7/content/fsdocs-custom.css" />
    <script type="text/javascript" src="/BIO-BTE-06-L-7/content/fsdocs-tips.js"></script>
    
    <link type="text/css" rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" />
    <script defer type="text/javascript" src="https://kit.fontawesome.com/0d3e0ea7a6.js"></script>
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <!-- END SEARCH BOX: this adds support for the search box -->
    
</head> 

<body>
    <div class="columns is-fullheight m-0">
        <div id="fsdocs-menu" class="column is-2 is-paddingless">
            <aside class="menu">
                <div class="container is-paddingless is-marginless has-background-dark">
                    <a href="/BIO-BTE-06-L-7/">
                        <figure class="image is-128x128 container">
                            <img src="img/logo.svg">
                        </figure>
                    </a>
                    <h2 class="title has-text-centered has-text-white pb-4">BIO-BTE-06-L-7</h2>
                </div>
                <div class="content">
                    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
                    <div id="header">
                        <div class="field">
                            <div class="control has-icons-left has-icons-right searchbox" id="fsdocs-searchbox">
                                <input data-search-input="" id="search-by" type="search" placeholder="Search..." class="input"/>
                                <span class="icon is-small is-left">
                                    <i class="fas fa-search"></i>
                                </span>
                            </div>
                        </div>
                    </div>

                    <!-- END SEARCH BOX: this adds support for the search box -->
                    <ul class="navbar-nav">
                        <li class="nav-item">
                            <a class="nav-link" href="NB00a_FSharp_Coding_literacy_Part_I.html">
                                <span class="text">
                                    <span>NB00a FSharp Coding literacy part I</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB00b_FSharp_Coding_literacy_Part_II.html">
                                <span class="text">
                                    <span>NB00b FSharp Coding literacy part II</span>
                                </span>
                            </a>
                        </li>
                        <!--<li class="nav-item">
                            <a class="nav-link" href="NB01a_Systems_Biology_FSharp_Introduction.html">
                                <span class="text">
                                    <span>NB01a Systems Biology</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB01b_Plant_Systems_Biology.html">
                                <span class="text">
                                    <span>NB01b Plant Systems Biology</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB02a_Mass_spectrometry_based_proteomics.html">
                                <span class="text">
                                    <span>NB02a Mass Spectrometry Based Proteomics</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB02b_Digestion_and_mass_calculation.html">
                                <span class="text">
                                    <span>NB02b Digestion and mass calculation</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB02c_Isotopic_distribution.html">
                                <span class="text">
                                    <span>NB02c Isotopic Distribution</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB03a_Retention_time_and_scan_time.html">
                                <span class="text">
                                    <span>NB03a Retention time and scan time</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB03b_Signal_detection_and_quantification.html">
                                <span class="text">
                                    <span>NB03b Signal detection and quantification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB03c_Centroidisation.html">
                                <span class="text">
                                    <span>NB03c Centroidisation</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB04a_Fragmentation_for_peptide_identification.html">
                                <span class="text">
                                    <span>NB04a Fragmentation for peptide identification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB04b_Peptide_Identification.html">
                                <span class="text">
                                    <span>NB04b Peptide Identification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB05a_Quantification.html">
                                <span class="text">
                                    <span>NB05a Quantification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB06a_Research_Data_Management.html">
                                <span class="text">
                                    <span>NB06a Research Data Management</span>
                                </span>
                            </a>
                        </li>-->
                        <li class="nav-item">
                            <a class="nav-link" href="NB07a_Working_With_Deedle.html">
                                <span class="text">
                                    <span>NB07a Working with Deedle</span>
                                </span>
                            </a>
                        </li>
                        <!--<li class="nav-item">
                            <a class="nav-link" href="NB08a_Data_Access_And_Quality_Control.html">
                                <span class="text">
                                    <span>NB08a Data Access and Quality Control</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08a_Data_Access_And_Quality_Control_SDS.html">
                                <span class="text">
                                    <span>NB08a Data Access and Quality Control (SDS)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08a_Data_Access_And_Quality_Control_BN.html">
                                <span class="text">
                                    <span>NB08a Data Access and Quality Control (BN)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08b_Label_efficiency.html">
                                <span class="text">
                                    <span>NB08b Label Efficiency</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08b_Label_efficiency_SDS.html">
                                <span class="text">
                                    <span>NB08b Label Efficiency (SDS)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08b_Label_efficiency_BN.html">
                                <span class="text">
                                    <span>NB08b Label Efficiency (BN)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08c_Absolute_Quantification.html">
                                <span class="text">
                                    <span>NB08c Absolute Quantification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08c_Absolute_Quantification_SDS.html">
                                <span class="text">
                                    <span>NB08c Absolute Quantification (SDS)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08c_Absolute_Quantification_BN.html">
                                <span class="text">
                                    <span>NB08c Absolute Quantification (BN)</span>
                                </span>
                            </a>
                        </li>-->
                        <li class="nav-header">Links</li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://csb.bio.uni-kl.de/#teaching">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fas fa-graduation-cap"></i>
                                    </span>
                                    <span>Other CSB courses</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://twitter.com/cs_biology">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fab fa-twitter"></i>
                                    </span>
                                    <span>Get in touch on twitter</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://github.com/CSBiology?type=source">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fab fa-github"></i>
                                    </span>
                                    <span>Check out our open source projects</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://github.com/CSBiology/BIO-BTE-06-L-7">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fas fa-code-branch"></i>
                                    </span>
                                    <span>Source repository</span>
                                </span>
                            </a>
                        </li>
                    </ul>
                </div>
            </aside>
        </div>
        <div class="column is-10 is-paddingless has-text-justified	">
            <div class="section pt-4">
                <div class="container" id="fsdocs-content">
                    <h1><a name="NB02c-Isotopic-Distribution" class="anchor" href="#NB02c-Isotopic-Distribution">NB02c Isotopic Distribution</a></h1>
<p><a href="https://mybinder.org/v2/gh/CSBiology/BIO-BTE-06-L-7/gh-pages?filepath=NB02c_Isotopic_distribution.ipynb"><img src="https://mybinder.org/badge_logo.svg" alt="Binder" /></a></p>
<p><a href="https://github.com/CSBiology/BIO-BTE-06-L-7/releases/download/NB02a_NB02b_NB02c/NB02c_Isotopic_distribution.ipynb">Download Notebook</a></p>
<ol>
<li>
Isotopic Distribution
<ol>
<li>Simulating Isotopic Clusters for peptides</li>
<li>Simulating Isotopic Clusters for peptides with stable isotope labeled variant</li>
</ol>
</li>
<li>References</li>
</ol>
<h2><a name="Isotopic-Distribution" class="anchor" href="#Isotopic-Distribution">Isotopic Distribution</a></h2>
<p>Peptide signals exhibit a characteristic shape in the mass spectrum that depend on their isotopic profile, which is defined by
the number of naturally occurring isotopes in the peptide. The occurrence probabilities of natural isotopes are reflected in the mass
spectrum by the relative heights of the peak series belonging to the respective peptide. The frequency at which natural isotopes occur
is known and can be used to compute the isotope distribution of a molecule. The isotopic distribution for a given peptide molecule
C(v)H(w)N(x)O(y)S(z) is described by the following product of polynomials:</p>
<p><img src="https://latex.codecogs.com/png.latex?\large&amp;space;\newline(&amp;space;{}^{12}\textrm{C}&amp;space;&amp;plus;&amp;space;{}^{13}\textrm{C})^{v}&amp;space;\times&amp;space;({}^{1}\textrm{H}&amp;plus;{}^{2}\textrm{H})^{w}&amp;space;\times&amp;space;({}^{14}\textrm{N}&amp;plus;{}^{15}\textrm{N})^{x}\times({}^{16}\textrm{O}&amp;plus;{}^{17}\textrm{O}&amp;space;&amp;plus;&amp;space;{}^{18}\textrm{O})^{y}\newline\times({}^{32}\textrm{S}&amp;plus;{}^{33}\textrm{S}&amp;plus;{}^{34}\textrm{S}&amp;plus;{}^{36}\textrm{S})^{z}" alt="" /></p>
<p>Symbolic expansion of the polynomials results in many product terms, which correspond to different isotopic variants of a molecule.
Even for molecules of a medium size, the straightforward expansion of the polynomials leads to an explosion regarding the number of product terms.
Due to this complexity, there was a need to develop algorithms for efficient computation. The different strategies comprise pruning the
polynomials to discard terms with coefficients below a threshold (Yergey 1983) combined with a recursive
computation (Claesen et al. 2012), and Fourier Transformation for a more efficient convolution of the isotope distributions of
individual elements (Rockwood et al. 1995), or rely on dynamic programming (Snider 2007).</p>
<blockquote>
<p>MIDAs (Alves and Yu 2005) is one of the more elaborate algorithms to predict an isotope cluster based on a given peptide sequence.
Simulate the isotopic cluster of the peptide sequence ‘PEPTIDES’ and ‘PEPTIDEPEPTIDEPEPTIDEPEPTIDES’ with natural occurring isotope abundances.</p>
</blockquote>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="pp">#r</span> <span class="s">&quot;nuget: BioFSharp, 2.0.0-beta5&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: BioFSharp.IO, 2.0.0-beta5&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: Plotly.NET, 4.2.0&quot;</span>

<span class="pp">#if</span> <span class="id">IPYNB</span>
<span class="inactive">#r</span><span class="inactive"> </span><span class="inactive">&quot;nuget:</span><span class="inactive"> </span><span class="inactive">Plotly.NET.Interactive,</span><span class="inactive"> </span><span class="inactive">4.2.0&quot;</span>
<span class="pp">#endif</span> <span class="c">// IPYNB</span>

<span class="k">open</span> <span class="id">Plotly</span><span class="pn">.</span><span class="id">NET</span>
<span class="k">open</span> <span class="id">BioFSharp</span>
</code></pre>
<h2><a name="Simulating-Isotopic-Clusters-for-peptides" class="anchor" href="#Simulating-Isotopic-Clusters-for-peptides">Simulating Isotopic Clusters for peptides</a></h2>
<p>We will use two artificial peptide sequences and translate them into their elemental composition to simulate their isotopic clusters.
Therefore, we first define a function that maps from a peptide sequence to its formula:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 1</span>

<span class="c">// create chemical formula for amino acid and add water to reflect hydrolysed state in mass spectrometer</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="fn">toFormula</span> <span onmouseout="hideTip(event, 'fs2', 2)" onmouseover="showTip(event, 'fs2', 2)" class="fn">bioseq</span> <span class="o">=</span>  
    <span onmouseout="hideTip(event, 'fs2', 3)" onmouseover="showTip(event, 'fs2', 3)" class="fn">bioseq</span>
    <span class="o">|&gt;</span> <span class="id">BioSeq</span><span class="pn">.</span><span class="id">toFormula</span>
    <span class="c">// peptides are hydrolysed in the mass spectrometer, so we add H2O</span>
    <span class="o">|&gt;</span> <span class="id">Formula</span><span class="pn">.</span><span class="id">add</span> <span class="id">Formula</span><span class="pn">.</span><span class="id">Table</span><span class="pn">.</span><span class="id">H2O</span>
</code></pre>
<p>Next, we will apply our function to receive the elemental composition or chemical formula of the peptides.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 2</span>

<span class="c">// translate single letter code into amino acids and create chemical formula of it.</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs3', 4)" onmouseover="showTip(event, 'fs3', 4)" class="id">peptide_short</span> <span class="o">=</span> 
    <span class="s">&quot;PEPTIDES&quot;</span> 
    <span class="o">|&gt;</span> <span class="id">BioSeq</span><span class="pn">.</span><span class="id">ofAminoAcidString</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs1', 5)" onmouseover="showTip(event, 'fs1', 5)" class="fn">toFormula</span>
    
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs4', 6)" onmouseover="showTip(event, 'fs4', 6)" class="id">peptide_long</span>  <span class="o">=</span> 
    <span class="s">&quot;PEPTIDEPEPTIDEPEPTIDEPEPTIDES&quot;</span> 
    <span class="o">|&gt;</span> <span class="id">BioSeq</span><span class="pn">.</span><span class="id">ofAminoAcidString</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs1', 7)" onmouseover="showTip(event, 'fs1', 7)" class="fn">toFormula</span>
    
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs5', 8)" onmouseover="showTip(event, 'fs5', 8)" class="id">peptide_shortString</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs3', 9)" onmouseover="showTip(event, 'fs3', 9)" class="id">peptide_short</span> 
    <span class="o">|&gt;</span> <span class="id">Formula</span><span class="pn">.</span><span class="id">toString</span>


<span class="k">let</span> <span onmouseout="hideTip(event, 'fs6', 10)" onmouseover="showTip(event, 'fs6', 10)" class="id">peptide_longString</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs4', 11)" onmouseover="showTip(event, 'fs4', 11)" class="id">peptide_long</span> 
    <span class="o">|&gt;</span> <span class="id">Formula</span><span class="pn">.</span><span class="id">toString</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<p>Additionally, we need a function that maps from Formula (and charge) to the isotopic distribution. Here, we
can use <code>IsotopicDistribution.MIDA.ofFormula</code> from the BioFSharp library. However, for convenience
(to use the same parameter twice), we define our function <code>generateIsotopicDistribution</code>:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 3</span>

<span class="c">// Predicts an isotopic distribution of the given formula at the given charge, </span>
<span class="c">// normalized by the sum of probabilities, using the MIDAs algorithm</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs7', 12)" onmouseover="showTip(event, 'fs7', 12)" class="fn">generateIsotopicDistribution</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs8', 13)" onmouseover="showTip(event, 'fs8', 13)" class="fn">charge</span><span class="pn">:</span> <span onmouseout="hideTip(event, 'fs9', 14)" onmouseover="showTip(event, 'fs9', 14)" class="vt">int</span><span class="pn">)</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs10', 15)" onmouseover="showTip(event, 'fs10', 15)" class="fn">f</span><span class="pn">:</span> <span class="id">Formula</span><span class="pn">.</span><span class="id">Formula</span><span class="pn">)</span> <span class="o">=</span>
    <span class="id">IsotopicDistribution</span><span class="pn">.</span><span class="id">MIDA</span><span class="pn">.</span><span class="id">ofFormula</span> 
        <span class="id">IsotopicDistribution</span><span class="pn">.</span><span class="id">MIDA</span><span class="pn">.</span><span class="id">normalizeByMaxProb</span>
        <span class="n">0.01</span>
        <span class="n">0.005</span>
        <span onmouseout="hideTip(event, 'fs8', 16)" onmouseover="showTip(event, 'fs8', 16)" class="id">charge</span>
        <span onmouseout="hideTip(event, 'fs10', 17)" onmouseover="showTip(event, 'fs10', 17)" class="id">f</span>

<span class="c">// create pattern for peptide_short</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs11', 18)" onmouseover="showTip(event, 'fs11', 18)" class="id">isoPattern_peptide_short</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs7', 19)" onmouseover="showTip(event, 'fs7', 19)" class="fn">generateIsotopicDistribution</span> <span class="n">1</span> <span onmouseout="hideTip(event, 'fs3', 20)" onmouseover="showTip(event, 'fs3', 20)" class="id">peptide_short</span>

<span class="c">// create pattern for peptide_long</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs12', 21)" onmouseover="showTip(event, 'fs12', 21)" class="id">isoPattern_peptide_long</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs7', 22)" onmouseover="showTip(event, 'fs7', 22)" class="fn">generateIsotopicDistribution</span> <span class="n">1</span> <span onmouseout="hideTip(event, 'fs4', 23)" onmouseover="showTip(event, 'fs4', 23)" class="id">peptide_long</span>
    
<span class="c">//isoPattern_peptide_long</span>
<span onmouseout="hideTip(event, 'fs11', 24)" onmouseover="showTip(event, 'fs11', 24)" class="id">isoPattern_peptide_short</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 4</span>

<span class="c">// create one chart for both, short and long peptide isotopic patterns.</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs13', 25)" onmouseover="showTip(event, 'fs13', 25)" class="id">isoPatternChart</span> <span class="o">=</span> 
    <span class="pn">[</span>
        <span class="id">Chart</span><span class="pn">.</span><span class="id">Column</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs11', 26)" onmouseover="showTip(event, 'fs11', 26)" class="id">isoPattern_peptide_short</span><span class="pn">,</span> <span class="id">Name</span> <span class="o">=</span> <span class="s">&quot;peptide_short&quot;</span> <span class="pn">)</span>
        <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withXAxisStyle</span> <span class="pn">(</span><span class="s">&quot;m/z&quot;</span><span class="pn">,</span> <span class="id">MinMax</span> <span class="o">=</span> <span class="pn">(</span><span class="n">885.</span><span class="pn">,</span> <span class="n">895.</span><span class="pn">)</span><span class="pn">)</span>
        <span class="id">Chart</span><span class="pn">.</span><span class="id">Column</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs12', 27)" onmouseover="showTip(event, 'fs12', 27)" class="id">isoPattern_peptide_long</span><span class="pn">,</span> <span class="id">Name</span> <span class="o">=</span> <span class="s">&quot;peptide_long&quot;</span> <span class="pn">)</span>
        <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withXAxisStyle</span> <span class="pn">(</span><span class="s">&quot;m/z&quot;</span><span class="pn">,</span> <span class="id">MinMax</span> <span class="o">=</span> <span class="pn">(</span><span class="n">3230.</span><span class="pn">,</span> <span class="n">3240.</span><span class="pn">)</span><span class="pn">)</span>
    <span class="pn">]</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">Grid</span> <span class="pn">(</span><span class="n">1</span><span class="pn">,</span><span class="n">2</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withSize</span> <span class="pn">(</span><span class="n">900.</span><span class="pn">,</span><span class="n">600.</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withTitle</span> <span class="s">&quot;Isotopeclusters&quot;</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withYAxisStyle</span> <span class="s">&quot;intensity&quot;</span>

<span onmouseout="hideTip(event, 'fs13', 28)" onmouseover="showTip(event, 'fs13', 28)" class="id">isoPatternChart</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<h2><a name="Simulating-Isotopic-Clusters-for-peptides-with-stable-isotope-labeled-variant" class="anchor" href="#Simulating-Isotopic-Clusters-for-peptides-with-stable-isotope-labeled-variant">Simulating Isotopic Clusters for peptides with stable isotope labeled variant</a></h2>
<p>In addition to the natural occurring isotopic distribution, the field of proteomics has benefited greatly from the ability to
introduce stable isotopes into peptide sequences. So called isotopic labeling refers to the introduction of a naturally low-abundance
isotope of carbon, nitrogen, hydrogen and, in some cases, oxygen, into a peptide sequence. The isotopes commonly used are 13C,
15N, 2H (deuterium) and 18O with natural abundances of 1.10 %, 0.366 %, 0.015 % and 0.200 %,
respectively (Becker 2008). Therefore, the introduction of these isotopes into a peptide sequence can be detected by
most modern mass spectrometers leading to a respective mass shift and the ability to separate the same peptide species within the same run.</p>
<blockquote>
<p>MIDAs (Alves and Yu 2005) is also able to predict isotope clusters with altered isotope abundances. Simulate the isotopic cluster
of the peptide sequence ‘PEPTIDES’ and ‘PEPTIDEPEPTIDEPEPTIDEPEPTIDES’ with stable isotopes 15N labeling.</p>
</blockquote>
<p>Therefore, we define a function called <code>label</code>. The function maps from a formula to a formula with exchangen nitrogen isotopes.
(Attention: Don't get confused a formula is just an FSharpMap.)</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 5</span>

<span class="c">/// returns a function that replaces the nitrogen atoms in a formula</span>
<span class="c">/// with the 15N isotope</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs14', 30)" onmouseover="showTip(event, 'fs14', 30)" class="fn">label</span> <span onmouseout="hideTip(event, 'fs15', 31)" onmouseover="showTip(event, 'fs15', 31)" class="fn">formula</span> <span class="o">=</span>
    <span class="id">Formula</span><span class="pn">.</span><span class="id">replaceElement</span> <span onmouseout="hideTip(event, 'fs15', 32)" onmouseover="showTip(event, 'fs15', 32)" class="id">formula</span> <span class="id">Elements</span><span class="pn">.</span><span class="id">Table</span><span class="pn">.</span><span class="id">N</span> <span class="id">Elements</span><span class="pn">.</span><span class="id">Table</span><span class="pn">.</span><span class="id">Heavy</span><span class="pn">.</span><span class="id">N15</span>

<span class="c">// Code-Block 6</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs16', 33)" onmouseover="showTip(event, 'fs16', 33)" class="id">N15_peptide_short</span> <span class="o">=</span> 
    <span class="s">&quot;PEPTIDES&quot;</span> 
    <span class="o">|&gt;</span> <span class="id">BioSeq</span><span class="pn">.</span><span class="id">ofAminoAcidString</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs1', 34)" onmouseover="showTip(event, 'fs1', 34)" class="fn">toFormula</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs14', 35)" onmouseover="showTip(event, 'fs14', 35)" class="fn">label</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs17', 36)" onmouseover="showTip(event, 'fs17', 36)" class="id">N15_peptide_long</span>  <span class="o">=</span> 
    <span class="s">&quot;PEPTIDEPEPTIDEPEPTIDEPEPTIDES&quot;</span> 
    <span class="o">|&gt;</span> <span class="id">BioSeq</span><span class="pn">.</span><span class="id">ofAminoAcidString</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs1', 37)" onmouseover="showTip(event, 'fs1', 37)" class="fn">toFormula</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs14', 38)" onmouseover="showTip(event, 'fs14', 38)" class="fn">label</span>

<span class="c">//result: N15_peptide_short</span>
<span onmouseout="hideTip(event, 'fs16', 39)" onmouseover="showTip(event, 'fs16', 39)" class="id">N15_peptide_short</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//result: N15_peptide_long</span>
<span onmouseout="hideTip(event, 'fs17', 40)" onmouseover="showTip(event, 'fs17', 40)" class="id">N15_peptide_long</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 7</span>

<span class="c">// create pattern for N15_peptide_short</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs18', 41)" onmouseover="showTip(event, 'fs18', 41)" class="id">N15_isoPattern_peptide_short</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs7', 42)" onmouseover="showTip(event, 'fs7', 42)" class="fn">generateIsotopicDistribution</span> <span class="n">1</span> <span onmouseout="hideTip(event, 'fs16', 43)" onmouseover="showTip(event, 'fs16', 43)" class="id">N15_peptide_short</span>

<span class="c">// create pattern for N15_peptide_long</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs19', 44)" onmouseover="showTip(event, 'fs19', 44)" class="id">N15_isoPattern_peptid_long</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs7', 45)" onmouseover="showTip(event, 'fs7', 45)" class="fn">generateIsotopicDistribution</span> <span class="n">1</span> <span onmouseout="hideTip(event, 'fs17', 46)" onmouseover="showTip(event, 'fs17', 46)" class="id">N15_peptide_long</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 8</span>

<span class="c">// Create two charts. Each with the related 14N and 15N isotopic clusters. Then stack them to one unit.</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs20', 47)" onmouseover="showTip(event, 'fs20', 47)" class="id">isoPatternChart2</span> <span class="o">=</span> 
    <span class="pn">[</span>
        <span class="pn">[</span>
            <span class="id">Chart</span><span class="pn">.</span><span class="id">Column</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs11', 48)" onmouseover="showTip(event, 'fs11', 48)" class="id">isoPattern_peptide_short</span><span class="pn">,</span> <span class="id">Name</span> <span class="o">=</span> <span class="s">&quot;peptide_short&quot;</span> <span class="pn">)</span>
            <span class="id">Chart</span><span class="pn">.</span><span class="id">Column</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs18', 49)" onmouseover="showTip(event, 'fs18', 49)" class="id">N15_isoPattern_peptide_short</span><span class="pn">,</span> <span class="id">Name</span> <span class="o">=</span> <span class="s">&quot;N15_peptide_short&quot;</span> <span class="pn">)</span>
        <span class="pn">]</span> 
        <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">combine</span> 
        <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withXAxisStyle</span> <span class="pn">(</span><span class="s">&quot;m/z&quot;</span><span class="pn">,</span> <span class="id">MinMax</span> <span class="o">=</span> <span class="pn">(</span><span class="n">885.</span><span class="pn">,</span> <span class="n">905.0</span><span class="pn">)</span><span class="pn">)</span>

        <span class="pn">[</span>
            <span class="id">Chart</span><span class="pn">.</span><span class="id">Column</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs12', 50)" onmouseover="showTip(event, 'fs12', 50)" class="id">isoPattern_peptide_long</span><span class="pn">,</span><span class="id">Name</span> <span class="o">=</span> <span class="s">&quot;peptide_long&quot;</span> <span class="pn">)</span>
            <span class="id">Chart</span><span class="pn">.</span><span class="id">Column</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs19', 51)" onmouseover="showTip(event, 'fs19', 51)" class="id">N15_isoPattern_peptid_long</span><span class="pn">,</span> <span class="id">Name</span> <span class="o">=</span> <span class="s">&quot;N15_peptide_long&quot;</span> <span class="pn">)</span>
        <span class="pn">]</span> 
        <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">combine</span> 
        <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withXAxisStyle</span> <span class="pn">(</span><span class="s">&quot;m/z&quot;</span><span class="pn">,</span> <span class="id">MinMax</span> <span class="o">=</span> <span class="pn">(</span><span class="n">3230.0</span><span class="pn">,</span> <span class="n">3270.0</span><span class="pn">)</span><span class="pn">)</span>
    <span class="pn">]</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">Grid</span> <span class="pn">(</span><span class="n">1</span><span class="pn">,</span> <span class="n">2</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withTitle</span> <span class="s">&quot;Isotopeclusters&quot;</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withYAxisStyle</span> <span class="s">&quot;intensity&quot;</span>

<span onmouseout="hideTip(event, 'fs20', 52)" onmouseover="showTip(event, 'fs20', 52)" class="id">isoPatternChart2</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<h2><a name="Questions" class="anchor" href="#Questions">Questions</a></h2>
<ol>
<li>Does the isotopic distribution tend to vary more or less the larger a peptide sequence becomes?</li>
<li>How does the isotopic distribution of the amino acid-sequence "SEQUENCE" looks like with natural and heavy N-labelling? Plot it via using the functions provided above.</li>
<li>Which information can be drawn from the distance of two values of adjacent signals (of one isotopic cluster) regarding m/z?</li>
</ol>
<h2><a name="References" class="anchor" href="#References">References</a></h2>
<ol>
<li>Yergey, J. A. A General-Approach to Calculating Isotopic Distributions for Mass-Spectrometry. Int J Mass Spectrom 52, 337–349; 10.1016/0020-7381(83)85053-0 (1983).</li>
<li>Claesen, J., Dittwald, P., Burzykowski, T. &amp; Valkenborg, D. An efficient method to calculate the aggregated isotopic distribution and exact center-masses. Journal of the American Society for Mass Spectrometry 23, 753–763; 10.1007/s13361-011-0326-2 (2012).</li>
<li>Rockwood, A. L., Vanorden, S. L. &amp; Smith, R. D. Rapid Calculation of Isotope Distributions. Anal Chem 67, 2699–2704; 10.1021/Ac00111a031 (1995).</li>
<li>Snider, R. K. Efficient calculation of exact mass isotopic distributions. Journal of the American Society for Mass Spectrometry 18, 1511–1515; 10.1016/j.jasms.2007.05.016 (2007).</li>
<li>Alves, G. &amp; Yu, Y. K. Robust accurate identification of peptides (RAId). deciphering MS2 data using a structured library search with de novo based statistics. Bioinformatics 21, 3726–3732; 10.1093/bioinformatics/bti620 (2005).</li>
<li>Becker, G. W. Stable isotopic labeling of proteins for quantitative proteomic applications. Brief Funct Genomic Proteomic 7, 371–382; 10.1093/bfgp/eln047 (2008).</li>
</ol>

                </div>
            </div>
        </div>
        <div class="fsdocs-tip" id="fs1">val toFormula : bioseq:&#39;a -&gt; &#39;b</div>
<div class="fsdocs-tip" id="fs2">val bioseq : &#39;a</div>
<div class="fsdocs-tip" id="fs3">val peptide_short : obj</div>
<div class="fsdocs-tip" id="fs4">val peptide_long : obj</div>
<div class="fsdocs-tip" id="fs5">val peptide_shortString : obj</div>
<div class="fsdocs-tip" id="fs6">val peptide_longString : obj</div>
<div class="fsdocs-tip" id="fs7">val generateIsotopicDistribution : charge:int -&gt; f:&#39;a -&gt; &#39;b</div>
<div class="fsdocs-tip" id="fs8">val charge : int</div>
<div class="fsdocs-tip" id="fs9">Multiple items<br />val int : value:&#39;T -&gt; int (requires member op_Explicit)<br /><em>&lt;summary&gt;Converts the argument to signed 32-bit integer. This is a direct conversion for all 
 primitive numeric types. For strings, the input is converted using &lt;c&gt;Int32.Parse()&lt;/c&gt;  
 with InvariantCulture settings. Otherwise the operation requires an appropriate
 static conversion method on the input type.&lt;/summary&gt;<br />&lt;param name=&quot;value&quot;&gt;The input value.&lt;/param&gt;<br />&lt;returns&gt;The converted int&lt;/returns&gt;</em><br /><br />--------------------<br />[&lt;Struct&gt;]
type int = int32<br /><em>&lt;summary&gt;An abbreviation for the CLI type &lt;see cref=&quot;T:System.Int32&quot; /&gt;.&lt;/summary&gt;<br />&lt;category&gt;Basic Types&lt;/category&gt;</em><br /><br />--------------------<br />type int&lt;&#39;Measure&gt; =
  int<br /><em>&lt;summary&gt;The type of 32-bit signed integer numbers, annotated with a unit of measure. The unit
 of measure is erased in compiled code and when values of this type
 are analyzed using reflection. The type is representationally equivalent to 
 &lt;see cref=&quot;T:System.Int32&quot; /&gt;.&lt;/summary&gt;<br />&lt;category&gt;Basic Types with Units of Measure&lt;/category&gt;</em></div>
<div class="fsdocs-tip" id="fs10">val f : &#39;a</div>
<div class="fsdocs-tip" id="fs11">val isoPattern_peptide_short : obj</div>
<div class="fsdocs-tip" id="fs12">val isoPattern_peptide_long : obj</div>
<div class="fsdocs-tip" id="fs13">val isoPatternChart : obj</div>
<div class="fsdocs-tip" id="fs14">val label : formula:&#39;a -&gt; &#39;b<br /><em>&#160;returns a function that replaces the nitrogen atoms in a formula<br />&#160;with the 15N isotope</em></div>
<div class="fsdocs-tip" id="fs15">val formula : &#39;a</div>
<div class="fsdocs-tip" id="fs16">val N15_peptide_short : obj</div>
<div class="fsdocs-tip" id="fs17">val N15_peptide_long : obj</div>
<div class="fsdocs-tip" id="fs18">val N15_isoPattern_peptide_short : obj</div>
<div class="fsdocs-tip" id="fs19">val N15_isoPattern_peptid_long : obj</div>
<div class="fsdocs-tip" id="fs20">val isoPatternChart2 : obj</div>

    </div>

    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <script type="text/javascript">var fsdocs_search_baseurl = '/BIO-BTE-06-L-7/'</script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/lunr.js/2.3.8/lunr.min.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.min.js"></script>
    <script type="text/javascript" src="/BIO-BTE-06-L-7/content/fsdocs-search.js"></script>
    <!-- END SEARCH BOX: this adds support for the search box -->
</body>

</html>