<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <title>NB04a Fragmentation for peptide identification
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="author" content="">
    <link rel="shortcut icon" type="image/x-icon" href="/BIO-BTE-06-L-7//img/favicon.ico">
    <script src="https://code.jquery.com/jquery-3.4.1.min.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <!-- <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/MathJax.js?config=TeX-MML-AM_CHTML"></script> -->

    <link type="text/css" rel="stylesheet" href="/BIO-BTE-06-L-7/content/fsdocs-default.css" />
    <link type="text/css" rel="stylesheet" href="/BIO-BTE-06-L-7/content/fsdocs-custom.css" />
    <script type="text/javascript" src="/BIO-BTE-06-L-7/content/fsdocs-tips.js"></script>
    
    <link type="text/css" rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" />
    <script defer type="text/javascript" src="https://kit.fontawesome.com/0d3e0ea7a6.js"></script>
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <!-- END SEARCH BOX: this adds support for the search box -->
    
</head> 

<body>
    <div class="columns is-fullheight m-0">
        <div id="fsdocs-menu" class="column is-2 is-paddingless">
            <aside class="menu">
                <div class="container is-paddingless is-marginless has-background-dark">
                    <a href="/BIO-BTE-06-L-7/">
                        <figure class="image is-128x128 container">
                            <img src="img/logo.svg">
                        </figure>
                    </a>
                    <h2 class="title has-text-centered has-text-white pb-4">BIO-BTE-06-L-7</h2>
                </div>
                <div class="content">
                    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
                    <div id="header">
                        <div class="field">
                            <div class="control has-icons-left has-icons-right searchbox" id="fsdocs-searchbox">
                                <input data-search-input="" id="search-by" type="search" placeholder="Search..." class="input"/>
                                <span class="icon is-small is-left">
                                    <i class="fas fa-search"></i>
                                </span>
                            </div>
                        </div>
                    </div>

                    <!-- END SEARCH BOX: this adds support for the search box -->
                    <ul class="navbar-nav">
                        <li class="nav-item">
                            <a class="nav-link" href="NB00a_FSharp_Coding_literacy_Part_I.html">
                                <span class="text">
                                    <span>NB00a FSharp Coding literacy part I</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB00b_FSharp_Coding_literacy_Part_II.html">
                                <span class="text">
                                    <span>NB00b FSharp Coding literacy part II</span>
                                </span>
                            </a>
                        </li>
                        <!--<li class="nav-item">
                            <a class="nav-link" href="NB01a_Systems_Biology_FSharp_Introduction.html">
                                <span class="text">
                                    <span>NB01a Systems Biology</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB01b_Plant_Systems_Biology.html">
                                <span class="text">
                                    <span>NB01b Plant Systems Biology</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB02a_Mass_spectrometry_based_proteomics.html">
                                <span class="text">
                                    <span>NB02a Mass Spectrometry Based Proteomics</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB02b_Digestion_and_mass_calculation.html">
                                <span class="text">
                                    <span>NB02b Digestion and mass calculation</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB02c_Isotopic_distribution.html">
                                <span class="text">
                                    <span>NB02c Isotopic Distribution</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB03a_Retention_time_and_scan_time.html">
                                <span class="text">
                                    <span>NB03a Retention time and scan time</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB03b_Signal_detection_and_quantification.html">
                                <span class="text">
                                    <span>NB03b Signal detection and quantification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB03c_Centroidisation.html">
                                <span class="text">
                                    <span>NB03c Centroidisation</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB04a_Fragmentation_for_peptide_identification.html">
                                <span class="text">
                                    <span>NB04a Fragmentation for peptide identification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB04b_Peptide_Identification.html">
                                <span class="text">
                                    <span>NB04b Peptide Identification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB05a_Quantification.html">
                                <span class="text">
                                    <span>NB05a Quantification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB06a_Research_Data_Management.html">
                                <span class="text">
                                    <span>NB06a Research Data Management</span>
                                </span>
                            </a>
                        </li>-->
                        <li class="nav-item">
                            <a class="nav-link" href="NB07a_Working_With_Deedle.html">
                                <span class="text">
                                    <span>NB07a Working with Deedle</span>
                                </span>
                            </a>
                        </li>
                        <!--<li class="nav-item">
                            <a class="nav-link" href="NB08a_Data_Access_And_Quality_Control.html">
                                <span class="text">
                                    <span>NB08a Data Access and Quality Control</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08a_Data_Access_And_Quality_Control_SDS.html">
                                <span class="text">
                                    <span>NB08a Data Access and Quality Control (SDS)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08a_Data_Access_And_Quality_Control_BN.html">
                                <span class="text">
                                    <span>NB08a Data Access and Quality Control (BN)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08b_Label_efficiency.html">
                                <span class="text">
                                    <span>NB08b Label Efficiency</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08b_Label_efficiency_SDS.html">
                                <span class="text">
                                    <span>NB08b Label Efficiency (SDS)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08b_Label_efficiency_BN.html">
                                <span class="text">
                                    <span>NB08b Label Efficiency (BN)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08c_Absolute_Quantification.html">
                                <span class="text">
                                    <span>NB08c Absolute Quantification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08c_Absolute_Quantification_SDS.html">
                                <span class="text">
                                    <span>NB08c Absolute Quantification (SDS)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08c_Absolute_Quantification_BN.html">
                                <span class="text">
                                    <span>NB08c Absolute Quantification (BN)</span>
                                </span>
                            </a>
                        </li>-->
                        <li class="nav-header">Links</li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://csb.bio.uni-kl.de/#teaching">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fas fa-graduation-cap"></i>
                                    </span>
                                    <span>Other CSB courses</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://twitter.com/cs_biology">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fab fa-twitter"></i>
                                    </span>
                                    <span>Get in touch on twitter</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://github.com/CSBiology?type=source">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fab fa-github"></i>
                                    </span>
                                    <span>Check out our open source projects</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://github.com/CSBiology/BIO-BTE-06-L-7">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fas fa-code-branch"></i>
                                    </span>
                                    <span>Source repository</span>
                                </span>
                            </a>
                        </li>
                    </ul>
                </div>
            </aside>
        </div>
        <div class="column is-10 is-paddingless has-text-justified	">
            <div class="section pt-4">
                <div class="container" id="fsdocs-content">
                    <h1><a name="NB04a-Fragmentation-for-peptide-identification" class="anchor" href="#NB04a-Fragmentation-for-peptide-identification">NB04a Fragmentation for peptide identification</a></h1>
<p><a href="https://mybinder.org/v2/gh/CSBiology/BIO-BTE-06-L-7/gh-pages?filepath=NB04a_Fragmentation_for_peptide_identification.ipynb"><img src="https://mybinder.org/badge_logo.svg" alt="Binder" /></a></p>
<p><a href="https://github.com/CSBiology/BIO-BTE-06-L-7/releases/download/NB04a_NB04b/NB04a_Fragmentation_for_peptide_identification.ipynb">Download Notebook</a></p>
<ol>
<li>Understanding MS2 spectra: From peptide to fragment</li>
<li>Simulate MS2 Fragmentation</li>
<li>Questions</li>
<li>References</li>
</ol>
<h2><a name="Understanding-MS2-spectra-From-peptide-to-fragment" class="anchor" href="#Understanding-MS2-spectra-From-peptide-to-fragment">Understanding MS2 spectra: From peptide to fragment</a></h2>
<p>The currency of information for identification in MS-based proteomics is the fragment ion spectrum (MS/MS spectrum) that is typically
derived from the fragmentation of a specific peptide in the collision cell of a mass spectrometer. Peptides produce fragments that provide
information on their amino acid sequence. The correct assignment of such a spectrum to a peptide sequence is the central step to link
m/z values and ion intensities to biology (Nesvizhskii et al. 2007).</p>
<p><img src="https://raw.githubusercontent.com/CSBiology/BIO-BTE-06-L-7/main/docs/img/FragmentIonNomenclature.PNG" alt="" /><br />
<strong><em>Figure 5: The Roepstorff-Fohlmann-Biemann nomenclature of fragment ions.</em></strong>
N-terminal and C-terminal peptide fragments result of dissociation of electron bonds along the peptide backbone.</p>
<p>During the unimolecular peptide ion dissociation processes, different chemical reactions can lead to different types
of product ions. The types of ions observed in MS/MS experiments depend on the physicochemical properties of the amino
acids and their sequence, on the amount of internal energy, and on the peptide’s charge state. In addition, product ion formation
is strongly influenced by the fragmentation method (Medzihradszky 2005). The most widely used fragmentation methods today
are low-energy collision-induced dissociation (CID) (Johnson et al. 1987) and electron transfer dissociation
(ETD) (Mikesh et al. 2006). These methods favor fragmentation along the peptide backbone and result in an N-terminal prefix
fragment and a C-terminal suffix fragment. The standard nomenclature for the C-terminal fragments is x, y and z whereas the corresponding
N-terminal fragments are denoted as a, b and c depending on the position where the breakage occurs at the peptide backbone level. The numbering
of each fragment starts from the N-terminus for a,b,c series and from the C-terminus for x,y,z series (Figure 4).
One should keep in mind that during parent ion selection many of the same peptide ions are selected and dissociated into fragments, with the
resulting fragment ions having different relative abundances according to the preferred fragmentation reaction. In addition to the
fragmentation along the peptide backbone, fragment ions containing the amino acids R, K, N, or Q can lose ammonia (-17 Da) and are then
denoted a<em>, b</em> and y*. Fragments containing the amino acids S, T, E, or D may lose water (-18 Da) and are then denoted a°, b° and y°.
These losses do not change the charge of the ions and are observable as natural losses (Forner et al. 2007, Steen and Mann 2004).</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="pp">#r</span> <span class="s">&quot;nuget: BioFSharp, 2.0.0-beta5&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: BioFSharp.IO, 2.0.0-beta5&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: Plotly.NET, 2.0.0-preview.16&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: BioFSharp.Mz, 0.1.5-beta&quot;</span>

<span class="pp">#if</span> <span class="id">IPYNB</span>
<span class="inactive">#r</span><span class="inactive"> </span><span class="inactive">&quot;nuget:</span><span class="inactive"> </span><span class="inactive">Plotly.NET.Interactive,</span><span class="inactive"> </span><span class="inactive">2.0.0-preview.16&quot;</span>
<span class="pp">#endif</span> <span class="c">// IPYNB</span>

<span class="k">open</span> <span class="id">Plotly</span><span class="pn">.</span><span class="id">NET</span>
<span class="k">open</span> <span class="id">BioFSharp</span>
</code></pre>
<h2><a name="Simulate-MS2-Fragmentation" class="anchor" href="#Simulate-MS2-Fragmentation">Simulate MS2 Fragmentation</a></h2>
<p>For the simulation we first define a short peptide. The peptide we take for this example is from rbcL.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 1</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="id">peptide</span> <span class="o">=</span> 
    <span class="s">&quot;DTDILAAFR&quot;</span>
    <span class="o">|&gt;</span> <span class="id">BioList</span><span class="pn">.</span><span class="id">ofAminoAcidString</span>

<span onmouseout="hideTip(event, 'fs1', 2)" onmouseover="showTip(event, 'fs1', 2)" class="id">peptide</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<p>In the <code>Mz</code> namespace of <a href="https://csbiology.github.io/BioFSharp/">BioFSharp</a>, we can find a function that can
generate the theoretical series of y-ions from the given peptide. This function provides a lot of information, but we are only interested
in the mass. Notice, that we do not know the intesity of the fragment ions and just use '1.' for simulation.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 2</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs2', 3)" onmouseover="showTip(event, 'fs2', 3)" class="id">ionSeriesY</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs1', 4)" onmouseover="showTip(event, 'fs1', 4)" class="id">peptide</span>
    <span class="o">|&gt;</span> <span class="id">Mz</span><span class="pn">.</span><span class="id">Fragmentation</span><span class="pn">.</span><span class="id">Series</span><span class="pn">.</span><span class="id">yOfBioList</span> <span class="id">BioItem</span><span class="pn">.</span><span class="id">initMonoisoMassWithMemP</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs3', 5)" onmouseover="showTip(event, 'fs3', 5)" class="m">List</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs4', 6)" onmouseover="showTip(event, 'fs4', 6)" class="id">map</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs5', 7)" onmouseover="showTip(event, 'fs5', 7)" class="fn">aac</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs5', 8)" onmouseover="showTip(event, 'fs5', 8)" class="fn">aac</span><span class="pn">.</span><span class="id">MainPeak</span><span class="pn">.</span><span class="id">Mass</span><span class="pn">,</span><span class="n">1.</span><span class="pn">)</span>
    
<span onmouseout="hideTip(event, 'fs2', 9)" onmouseover="showTip(event, 'fs2', 9)" class="id">ionSeriesY</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<p>Similarly, we can simulate the b-ion series.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 3</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs6', 10)" onmouseover="showTip(event, 'fs6', 10)" class="id">ionSeriesB</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs1', 11)" onmouseover="showTip(event, 'fs1', 11)" class="id">peptide</span>
    <span class="o">|&gt;</span> <span class="id">Mz</span><span class="pn">.</span><span class="id">Fragmentation</span><span class="pn">.</span><span class="id">Series</span><span class="pn">.</span><span class="id">bOfBioList</span> <span class="id">BioItem</span><span class="pn">.</span><span class="id">initMonoisoMassWithMemP</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs3', 12)" onmouseover="showTip(event, 'fs3', 12)" class="m">List</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs4', 13)" onmouseover="showTip(event, 'fs4', 13)" class="id">map</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs5', 14)" onmouseover="showTip(event, 'fs5', 14)" class="fn">aac</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs5', 15)" onmouseover="showTip(event, 'fs5', 15)" class="fn">aac</span><span class="pn">.</span><span class="id">MainPeak</span><span class="pn">.</span><span class="id">Mass</span><span class="pn">,</span><span class="n">1.</span><span class="pn">)</span>

<span onmouseout="hideTip(event, 'fs6', 16)" onmouseover="showTip(event, 'fs6', 16)" class="id">ionSeriesB</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<p>Now, we can just plot the simulated data and look at our theoretical spectrum.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 4</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs7', 17)" onmouseover="showTip(event, 'fs7', 17)" class="id">ionChart</span> <span class="o">=</span>
    <span class="pn">[</span>    
        <span class="id">Chart</span><span class="pn">.</span><span class="id">Column</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs6', 18)" onmouseover="showTip(event, 'fs6', 18)" class="id">ionSeriesB</span><span class="pn">,</span> <span class="id">Name</span> <span class="o">=</span> <span class="s">&quot;b ions&quot;</span><span class="pn">)</span>
        <span class="id">Chart</span><span class="pn">.</span><span class="id">Column</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs2', 19)" onmouseover="showTip(event, 'fs2', 19)" class="id">ionSeriesY</span><span class="pn">,</span> <span class="id">Name</span> <span class="o">=</span> <span class="s">&quot;y ions&quot;</span><span class="pn">)</span>
    <span class="pn">]</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">combine</span>

<span onmouseout="hideTip(event, 'fs7', 20)" onmouseover="showTip(event, 'fs7', 20)" class="id">ionChart</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<h2><a name="Questions" class="anchor" href="#Questions">Questions:</a></h2>
<ol>
<li>Why are ms1 spectra not sufficent for peptide identification?</li>
<li>How can fragmentation help with this?</li>
<li>For an oligopeptide consisting of 3 amino acids, roughly estimate the number of possible fragments if only cosidering b/y fragments or abc/xyz fragments. What advantages and disadvantages might only considering b/y fragments have?</li>
</ol>
<h2><a name="References" class="anchor" href="#References">References</a></h2>
<ol>
<li>Nesvizhskii, A. I., Vitek, O. &amp; Aebersold, R. Analysis and validation of proteomic data generated by tandem mass spectrometry. Nature methods 4, 787–797; 10.1038/nmeth1088 (2007).</li>
<li>Medzihradszky, K. F. Peptide sequence analysis. Method Enzymol 402, 209–244; 10.1016/S0076-6879(05)02007-0 (2005).</li>
<li>Johnson, R. S., Martin, S. A., Biemann, K., Stults, J. T. &amp; Watson, J. T. Novel fragmentation process of peptides by collision-induced decomposition in a tandem mass spectrometer: differentiation of leucine and isoleucine. Anal. Chem. 59, 2621–2625; 10.1021/Ac00148a019 (1987).</li>
<li>Mikesh, L. M. et al. The utility of ETD mass spectrometry in proteomic analysis. Biochimica et biophysica acta 1764, 1811–1822; 10.1016/j.bbapap.2006.10.003 (2006).</li>
<li>Forner, F., Foster, L. J. &amp; Toppo, S. Mass spectrometry data analysis in the proteomics era. Curr Bioinform 2, 63–93; 10.2174/157489307779314285 (2007).</li>
<li>Steen, H. &amp; Mann, M. The ABC's (and XYZ's) of peptide sequencing. Nat. Rev. Mol. Cell Biol. 5, 699–711; 10.1038/nrm1468 (2004).</li>
</ol>

                </div>
            </div>
        </div>
        <div class="fsdocs-tip" id="fs1">val peptide : obj</div>
<div class="fsdocs-tip" id="fs2">val ionSeriesY : (obj * float) list</div>
<div class="fsdocs-tip" id="fs3">Multiple items<br />module List

from Microsoft.FSharp.Collections<br /><em>&lt;summary&gt;Contains operations for working with values of type &lt;see cref=&quot;T:Microsoft.FSharp.Collections.list`1&quot; /&gt;.&lt;/summary&gt;<br />&lt;namespacedoc&gt;&lt;summary&gt;Operations for collections such as lists, arrays, sets, maps and sequences. See also 
    &lt;a href=&quot;https://docs.microsoft.com/dotnet/fsharp/language-reference/fsharp-collection-types&quot;&gt;F# Collection Types&lt;/a&gt; in the F# Language Guide.
 &lt;/summary&gt;&lt;/namespacedoc&gt;</em><br /><br />--------------------<br />type List&lt;&#39;T&gt; =
  | ( [] )
  | ( :: ) of Head: &#39;T * Tail: &#39;T list
    interface IReadOnlyList&lt;&#39;T&gt;
    interface IReadOnlyCollection&lt;&#39;T&gt;
    interface IEnumerable
    interface IEnumerable&lt;&#39;T&gt;
    member GetReverseIndex : rank:int * offset:int -&gt; int
    member GetSlice : startIndex:int option * endIndex:int option -&gt; &#39;T list
    static member Cons : head:&#39;T * tail:&#39;T list -&gt; &#39;T list
    member Head : &#39;T
    member IsEmpty : bool
    member Item : index:int -&gt; &#39;T with get
    ...<br /><em>&lt;summary&gt;The type of immutable singly-linked lists.&lt;/summary&gt;<br />&lt;remarks&gt;Use the constructors &lt;c&gt;[]&lt;/c&gt; and &lt;c&gt;::&lt;/c&gt; (infix) to create values of this type, or
 the notation &lt;c&gt;[1;2;3]&lt;/c&gt;. Use the values in the &lt;c&gt;List&lt;/c&gt; module to manipulate 
 values of this type, or pattern match against the values directly.
 &lt;/remarks&gt;<br />&lt;exclude /&gt;</em></div>
<div class="fsdocs-tip" id="fs4">val map : mapping:(&#39;T -&gt; &#39;U) -&gt; list:&#39;T list -&gt; &#39;U list<br /><em>&lt;summary&gt;Builds a new collection whose elements are the results of applying the given function
 to each of the elements of the collection.&lt;/summary&gt;<br />&lt;param name=&quot;mapping&quot;&gt;The function to transform elements from the input list.&lt;/param&gt;<br />&lt;param name=&quot;list&quot;&gt;The input list.&lt;/param&gt;<br />&lt;returns&gt;The list of transformed elements.&lt;/returns&gt;</em></div>
<div class="fsdocs-tip" id="fs5">val aac : obj</div>
<div class="fsdocs-tip" id="fs6">val ionSeriesB : (obj * float) list</div>
<div class="fsdocs-tip" id="fs7">val ionChart : obj</div>

    </div>

    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <script type="text/javascript">var fsdocs_search_baseurl = '/BIO-BTE-06-L-7/'</script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/lunr.js/2.3.8/lunr.min.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.min.js"></script>
    <script type="text/javascript" src="/BIO-BTE-06-L-7/content/fsdocs-search.js"></script>
    <!-- END SEARCH BOX: this adds support for the search box -->
</body>

</html>