<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <title>NB04b Peptide Identification
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="author" content="">
    <link rel="shortcut icon" type="image/x-icon" href="/BIO-BTE-06-L-7//img/favicon.ico">
    <script src="https://code.jquery.com/jquery-3.4.1.min.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <!-- <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/MathJax.js?config=TeX-MML-AM_CHTML"></script> -->

    <link type="text/css" rel="stylesheet" href="/BIO-BTE-06-L-7/content/fsdocs-default.css" />
    <link type="text/css" rel="stylesheet" href="/BIO-BTE-06-L-7/content/fsdocs-custom.css" />
    <script type="text/javascript" src="/BIO-BTE-06-L-7/content/fsdocs-tips.js"></script>
    
    <link type="text/css" rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" />
    <script defer type="text/javascript" src="https://kit.fontawesome.com/0d3e0ea7a6.js"></script>
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <!-- END SEARCH BOX: this adds support for the search box -->
    
</head> 

<body>
    <div class="columns is-fullheight m-0">
        <div id="fsdocs-menu" class="column is-2 is-paddingless">
            <aside class="menu">
                <div class="container is-paddingless is-marginless has-background-dark">
                    <a href="/BIO-BTE-06-L-7/">
                        <figure class="image is-128x128 container">
                            <img src="img/logo.svg">
                        </figure>
                    </a>
                    <h2 class="title has-text-centered has-text-white pb-4">BIO-BTE-06-L-7</h2>
                </div>
                <div class="content">
                    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
                    <div id="header">
                        <div class="field">
                            <div class="control has-icons-left has-icons-right searchbox" id="fsdocs-searchbox">
                                <input data-search-input="" id="search-by" type="search" placeholder="Search..." class="input"/>
                                <span class="icon is-small is-left">
                                    <i class="fas fa-search"></i>
                                </span>
                            </div>
                        </div>
                    </div>

                    <!-- END SEARCH BOX: this adds support for the search box -->
                    <ul class="navbar-nav">
                        <li class="nav-item">
                            <a class="nav-link" href="NB00a_FSharp_Coding_literacy_Part_I.html">
                                <span class="text">
                                    <span>NB00a FSharp Coding literacy part I</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB00b_FSharp_Coding_literacy_Part_II.html">
                                <span class="text">
                                    <span>NB00b FSharp Coding literacy part II</span>
                                </span>
                            </a>
                        </li>
                        <!--<li class="nav-item">
                            <a class="nav-link" href="NB01a_Systems_Biology_FSharp_Introduction.html">
                                <span class="text">
                                    <span>NB01a Systems Biology</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB01b_Plant_Systems_Biology.html">
                                <span class="text">
                                    <span>NB01b Plant Systems Biology</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB02a_Mass_spectrometry_based_proteomics.html">
                                <span class="text">
                                    <span>NB02a Mass Spectrometry Based Proteomics</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB02b_Digestion_and_mass_calculation.html">
                                <span class="text">
                                    <span>NB02b Digestion and mass calculation</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB02c_Isotopic_distribution.html">
                                <span class="text">
                                    <span>NB02c Isotopic Distribution</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB03a_Retention_time_and_scan_time.html">
                                <span class="text">
                                    <span>NB03a Retention time and scan time</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB03b_Signal_detection_and_quantification.html">
                                <span class="text">
                                    <span>NB03b Signal detection and quantification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB03c_Centroidisation.html">
                                <span class="text">
                                    <span>NB03c Centroidisation</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB04a_Fragmentation_for_peptide_identification.html">
                                <span class="text">
                                    <span>NB04a Fragmentation for peptide identification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB04b_Peptide_Identification.html">
                                <span class="text">
                                    <span>NB04b Peptide Identification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB05a_Quantification.html">
                                <span class="text">
                                    <span>NB05a Quantification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB06a_Research_Data_Management.html">
                                <span class="text">
                                    <span>NB06a Research Data Management</span>
                                </span>
                            </a>
                        </li>-->
                        <li class="nav-item">
                            <a class="nav-link" href="NB07a_Working_With_Deedle.html">
                                <span class="text">
                                    <span>NB07a Working with Deedle</span>
                                </span>
                            </a>
                        </li>
                        <!--<li class="nav-item">
                            <a class="nav-link" href="NB08a_Data_Access_And_Quality_Control.html">
                                <span class="text">
                                    <span>NB08a Data Access and Quality Control</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08a_Data_Access_And_Quality_Control_SDS.html">
                                <span class="text">
                                    <span>NB08a Data Access and Quality Control (SDS)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08a_Data_Access_And_Quality_Control_BN.html">
                                <span class="text">
                                    <span>NB08a Data Access and Quality Control (BN)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08b_Label_efficiency.html">
                                <span class="text">
                                    <span>NB08b Label Efficiency</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08b_Label_efficiency_SDS.html">
                                <span class="text">
                                    <span>NB08b Label Efficiency (SDS)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08b_Label_efficiency_BN.html">
                                <span class="text">
                                    <span>NB08b Label Efficiency (BN)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08c_Absolute_Quantification.html">
                                <span class="text">
                                    <span>NB08c Absolute Quantification</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08c_Absolute_Quantification_SDS.html">
                                <span class="text">
                                    <span>NB08c Absolute Quantification (SDS)</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="NB08c_Absolute_Quantification_BN.html">
                                <span class="text">
                                    <span>NB08c Absolute Quantification (BN)</span>
                                </span>
                            </a>
                        </li>-->
                        <li class="nav-header">Links</li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://csb.bio.uni-kl.de/#teaching">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fas fa-graduation-cap"></i>
                                    </span>
                                    <span>Other CSB courses</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://twitter.com/cs_biology">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fab fa-twitter"></i>
                                    </span>
                                    <span>Get in touch on twitter</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://github.com/CSBiology?type=source">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fab fa-github"></i>
                                    </span>
                                    <span>Check out our open source projects</span>
                                </span>
                            </a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="https://github.com/CSBiology/BIO-BTE-06-L-7">
                                <span class="icon-text">
                                    <span class="icon">
                                        <i class="fas fa-code-branch"></i>
                                    </span>
                                    <span>Source repository</span>
                                </span>
                            </a>
                        </li>
                    </ul>
                </div>
            </aside>
        </div>
        <div class="column is-10 is-paddingless has-text-justified	">
            <div class="section pt-4">
                <div class="container" id="fsdocs-content">
                    <h1><a name="NB04b-Peptide-Identification" class="anchor" href="#NB04b-Peptide-Identification">NB04b Peptide Identification</a></h1>
<p><a href="https://mybinder.org/v2/gh/CSBiology/BIO-BTE-06-L-7/gh-pages?filepath=NB04b_Peptide_Identification.ipynb"><img src="https://mybinder.org/badge_logo.svg" alt="Binder" /></a></p>
<p><a href="https://github.com/CSBiology/BIO-BTE-06-L-7/releases/download/NB04a_NB04b/NB04b_Peptide_Identification.ipynb">Download Notebook</a></p>
<ol>
<li>Understanding peptide identification from MS2 spectra</li>
<li>
Matching and scoring of Tandem MS peptide identification
<ol>
<li>Step 1: Data acquisition and preprocessing</li>
<li>Step 2: Preselecting the peptides of interest</li>
<li>Step 3+4: Matching and Scoring</li>
</ol>
</li>
</ol>
<h2><a name="Understanding-peptide-identification-from-MS2-spectra" class="anchor" href="#Understanding-peptide-identification-from-MS2-spectra">Understanding peptide identification from MS2 spectra</a></h2>
<p>Under low-energy fragmentation conditions, peptide fragment patterns are reproducible and, in general, predictable, which allows for an
amino acid sequence identification according to a fragmentation expectation. Algorithms for peptide identification perform in principle
three basic tasks:</p>
<p><strong>(i)</strong> a raw data preprocessing step is applied to the MS/MS spectra to obtain clean peak information. The same signal filtering
and background subtraction methods are used as discussed in the section of low-level processing. Peak detection, however, may be performed
differently. Preprocessing of MS/MS spectra focuses on the extraction of the precise m/z of the peak rather than the accurate peak areas.
The conversion of a peak profile into the corresponding m/z and intensity values reduces the complexity, its representation is termed centroiding.
To extract the masses for identification in a simple and fast way, peak fitting approaches are used. These approaches take either the most intense
point of the peak profile, fit a Lorentzian distribution to the profile, or use a quadratic fit.</p>
<p><strong>(ii)</strong> Spectrum information and possible amino acid sequences assignments are evaluated.</p>
<p><strong>(iii)</strong> The quality of the match between spectrum and possible sequences is scored.</p>
<p>Even though the three steps roughly describe the basic principle of algorithms used for peptide sequence identification, most implementations
show differences in individual steps which can lead to major changes in the outcome. Therefore, it has been proven useful to utilize more than
one algorithm for a robust and thorough identification. Due to their major difference in identification strategies and prerequisites,
identification algorithms are normally classified into three categories: (i) <em>de novo</em> peptide sequencing, (ii) peptide sequence-tag (PST)
searching, and (iii) uninterpreted sequence searching. However, in this notebook we focus on the explanation of (iii) uninterpreted sequence
searching, the most frequently used methods.</p>
<h2><a name="Matching-and-scoring-of-Tandem-MS-peptide-identification" class="anchor" href="#Matching-and-scoring-of-Tandem-MS-peptide-identification">Matching and scoring of Tandem MS peptide identification</a></h2>
<p><img src="https://raw.githubusercontent.com/CSBiology/BIO-BTE-06-L-7/main/docs/img/ComputationalProteinIdentification.png" alt="" /></p>
<p><strong>Figure 6: Process of computational identification of peptides from their fragment spectra</strong></p>
<p>Previously we learned, that peptides fragment to create patterns characteristic of a specific amino acid sequence. These patterns are reproducible and, in general,
predictable taking the applied fragmentation method into account. This can be used for computational identification of peptides from their fragment spectra.
This process can be subdivided into 5 main steps: spectrum preprocessing, selection of possible sequences, generating theoretical spectra, matching and scoring
(Figure 6). The first step is a preprocessing of the experimental spectra and is done to reduce noise. Secondly, all possible amino acid
sequences are selected which match the particular precursor peptide mass. The possible peptides can but do not need to be restricted to a particular organism.
A theoretical spectrum is predicted for each of these amino acid sequences. Matching and scoring is performed by comparing experimental spectra to their predicted
corresponding theoretical spectra. The score function measures the closeness of fit between the experimental acquired and theoretical spectrum. There are many
different score functions, which can be considered as the main reason of different identifications considering different identification algorithm. The most
natural score function is the cross correlation score (xcorr) used by the commercially available search tool SEQUEST. One of the reasons the xcorr is so
sensitive is because it involves a correction factor that assesses the background correlation for each acquired spectrum and the theoretically predicted
spectrum from sequences within a database. To compute this correction factor, a measure of similarity is calculated at different offsets between a preprocessed
mass spectrum and a theoretical spectrum. The final xcorr is then the correlation at zero offset minus the mean correlation from all the individual offsets.</p>
<p>Thus, the correlation function measures the coherence of two signals by, in effect, translating one signal across the other. This can be mathematically
achieved using the formula for cross-correlation in the form for discrete input signals:</p>
<p><strong><em>Equation 5</em></strong></p>
<p><img src="https://latex.codecogs.com/png.latex?\large&amp;space;R_{\tau}&amp;space;=&amp;space;\sum_{i=0}^{n-1}x_{i}\cdot%20y_{i&amp;plus;\tau}" alt="" /></p>
<p>where x(i) is a peak of the reconstructed spectrum at position i and y(i) is a peak of the experimental spectrum. The displacement value ùúè
is the amount by which the signal is offset during the translation and is varied over a range of values. If two signals are the same, the correlation
function should have its maxima at ùúè=0, where there is no offset. The average of the offset-correlation is seen as the average background correlation
and needs to be subtracted from the correlation at ùúè=0. It follows:</p>
<p><strong><em>Equation 6</em></strong></p>
<p><img src="https://latex.codecogs.com/png.latex?xcorr&amp;space;=&amp;space;R_{0}&amp;space;-&amp;space;\frac{(\sum&amp;space;\begin{matrix}&amp;space;\tau=&amp;plus;offeset\\&amp;space;\tau=-offeset\end{matrix}R_{\tau})}{2*offset&amp;plus;1}" alt="" /></p>
<p>In practice many theoretical spectra have to be matched again a single experimental spectrum. Therefore, the calculation can be speed up by reformulating Equation 5 and Equation 6 and introduce a preprocessing step, which is independent of the predicted spectra.</p>
<p><strong><em>Equation 7</em></strong></p>
<p><img src="https://latex.codecogs.com/png.latex?xcorr&amp;space;=&amp;space;x_{0}\cdot&amp;space;y_{0}&amp;space;-&amp;space;\frac{(\sum&amp;space;\begin{matrix}&amp;space;\tau=&amp;plus;offeset\\&amp;space;\tau=-offeset\end{matrix}x_{0}\cdot&amp;space;y_{\tau})}{2*offset&amp;plus;1}" alt="" /></p>
<p>For the preprocessed experimental spectrum y' it follows:</p>
<p><strong><em>Equation 8</em></strong></p>
<p><img src="https://latex.codecogs.com/png.latex?xcorr&amp;space;=&amp;space;x_{0}\cdot&amp;space;y`" alt="" /></p>
<p>where:</p>
<p><strong><em>Equation 9</em></strong></p>
<p><img src="https://latex.codecogs.com/png.latex?y'&amp;space;=&amp;space;y_{0}&amp;space;-&amp;space;\frac{(\sum&amp;space;\begin{matrix}&amp;space;\tau=&amp;plus;offeset\\&amp;space;\tau=-offeset\end{matrix}x_{0}\cdot&amp;space;y_{\tau})}{2*offset&amp;plus;1}" alt="" /></p>
<p>Matching a measured spectrum against chlamy database</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="pp">#r</span> <span class="s">&quot;nuget: BioFSharp, 2.0.0-beta4&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: BioFSharp.IO, 2.0.0-beta4&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: Plotly.NET, 2.0.0-preview.16&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: BioFSharp.Mz, 0.1.5-beta&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: BIO-BTE-06-L-7_Aux, 0.0.9&quot;</span>

<span class="pp">#if</span> <span class="id">IPYNB</span>
<span class="inactive">#r</span><span class="inactive"> </span><span class="inactive">&quot;nuget:</span><span class="inactive"> </span><span class="inactive">Plotly.NET.Interactive,</span><span class="inactive"> </span><span class="inactive">2.0.0-preview.16&quot;</span>
<span class="pp">#endif</span> <span class="c">// IPYNB</span>

<span class="k">open</span> <span class="id">Plotly</span><span class="pn">.</span><span class="id">NET</span>
<span class="k">open</span> <span class="id">BioFSharp</span>
<span class="k">open</span> <span class="id">BioFSharp</span><span class="pn">.</span><span class="id">Mz</span>
<span class="k">open</span> <span class="id">BIO_BTE_06_L_7_Aux</span><span class="pn">.</span><span class="id">FS3_Aux</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="id">System</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs2', 2)" onmouseover="showTip(event, 'fs2', 2)" class="id">IO</span>
</code></pre>
<h3><a name="Step-1-Data-acquisition-and-preprocessing" class="anchor" href="#Step-1-Data-acquisition-and-preprocessing">Step 1: Data acquisition and preprocessing</a></h3>
<p>We load a single MS2 spectrum that is saved in a mgf file.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 1</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs3', 3)" onmouseover="showTip(event, 'fs3', 3)" class="id">directory</span> <span class="o">=</span> <span class="k">__SOURCE_DIRECTORY__</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs4', 4)" onmouseover="showTip(event, 'fs4', 4)" class="id">path</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 5)" onmouseover="showTip(event, 'fs5', 5)" class="rt">Path</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs6', 6)" onmouseover="showTip(event, 'fs6', 6)" class="id">Combine</span><span class="pn">[|</span><span onmouseout="hideTip(event, 'fs3', 7)" onmouseover="showTip(event, 'fs3', 7)" class="id">directory</span><span class="pn">;</span><span class="s">&quot;downloads/ms2sample.mgf&quot;</span><span class="pn">|]</span>
<span class="id">downloadFile</span> <span onmouseout="hideTip(event, 'fs4', 8)" onmouseover="showTip(event, 'fs4', 8)" class="id">path</span> <span class="s">&quot;ms2sample.mgf&quot;</span> <span class="s">&quot;bio-bte-06-l-7&quot;</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs7', 9)" onmouseover="showTip(event, 'fs7', 9)" class="id">ms2</span> <span class="o">=</span> 
    <span class="id">BioFSharp</span><span class="pn">.</span><span class="id">IO</span><span class="pn">.</span><span class="id">Mgf</span><span class="pn">.</span><span class="id">readMgf</span> <span onmouseout="hideTip(event, 'fs4', 10)" onmouseover="showTip(event, 'fs4', 10)" class="id">path</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs8', 11)" onmouseover="showTip(event, 'fs8', 11)" class="m">List</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs9', 12)" onmouseover="showTip(event, 'fs9', 12)" class="id">head</span>
    
<span onmouseout="hideTip(event, 'fs7', 13)" onmouseover="showTip(event, 'fs7', 13)" class="id">ms2</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<p>Here, the spectrum is already centroidized as shown in <em>NB03c_Centroidisation.ipynb</em> using the function
<code>msPeakPicking</code>. So we just visualize mass and intensity:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 2</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs10', 14)" onmouseover="showTip(event, 'fs10', 14)" class="id">ms2Chart</span> <span class="o">=</span> 
    <span class="id">Chart</span><span class="pn">.</span><span class="id">Column</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs7', 15)" onmouseover="showTip(event, 'fs7', 15)" class="id">ms2</span><span class="pn">.</span><span class="id">Intensity</span><span class="pn">,</span> <span onmouseout="hideTip(event, 'fs7', 16)" onmouseover="showTip(event, 'fs7', 16)" class="id">ms2</span><span class="pn">.</span><span class="id">Mass</span><span class="pn">)</span> 
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withTemplate</span> <span class="id">ChartTemplates</span><span class="pn">.</span><span class="id">light</span>


<span onmouseout="hideTip(event, 'fs10', 17)" onmouseover="showTip(event, 'fs10', 17)" class="id">ms2Chart</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<p>Now, we will preprocess the experimental spectrum from our example. This is on the one hand to reduce noise, but also to make
the measured spectrum more like the once we are able to simulate.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 3</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs11', 19)" onmouseover="showTip(event, 'fs11', 19)" class="id">lowerScanLimit</span> <span class="o">=</span> <span class="n">150.</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs12', 20)" onmouseover="showTip(event, 'fs12', 20)" class="id">upperScanLimit</span> <span class="o">=</span> <span class="n">1000.</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs13', 21)" onmouseover="showTip(event, 'fs13', 21)" class="id">preprocessedIntesities</span> <span class="o">=</span>
    <span class="id">Mz</span><span class="pn">.</span><span class="id">PeakArray</span><span class="pn">.</span><span class="id">zip</span> <span onmouseout="hideTip(event, 'fs7', 22)" onmouseover="showTip(event, 'fs7', 22)" class="id">ms2</span><span class="pn">.</span><span class="id">Mass</span> <span onmouseout="hideTip(event, 'fs7', 23)" onmouseover="showTip(event, 'fs7', 23)" class="id">ms2</span><span class="pn">.</span><span class="id">Intensity</span>
    <span class="o">|&gt;</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs14', 24)" onmouseover="showTip(event, 'fs14', 24)" class="fn">pa</span> <span class="k">-&gt;</span> <span class="id">Mz</span><span class="pn">.</span><span class="id">PeakArray</span><span class="pn">.</span><span class="id">peaksToNearestUnitDaltonBinVector</span> <span onmouseout="hideTip(event, 'fs14', 25)" onmouseover="showTip(event, 'fs14', 25)" class="id">pa</span> <span onmouseout="hideTip(event, 'fs11', 26)" onmouseover="showTip(event, 'fs11', 26)" class="id">lowerScanLimit</span> <span onmouseout="hideTip(event, 'fs12', 27)" onmouseover="showTip(event, 'fs12', 27)" class="id">upperScanLimit</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs14', 28)" onmouseover="showTip(event, 'fs14', 28)" class="fn">pa</span> <span class="k">-&gt;</span> <span class="id">Mz</span><span class="pn">.</span><span class="id">SequestLike</span><span class="pn">.</span><span class="id">windowNormalizeIntensities</span> <span onmouseout="hideTip(event, 'fs14', 29)" onmouseover="showTip(event, 'fs14', 29)" class="id">pa</span> <span class="n">10</span><span class="pn">)</span>
    
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs15', 30)" onmouseover="showTip(event, 'fs15', 30)" class="id">intensityChart</span> <span class="o">=</span> 
    <span class="id">Chart</span><span class="pn">.</span><span class="id">Column</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs13', 31)" onmouseover="showTip(event, 'fs13', 31)" class="id">preprocessedIntesities</span><span class="pn">,</span> <span class="pn">[</span><span onmouseout="hideTip(event, 'fs11', 32)" onmouseover="showTip(event, 'fs11', 32)" class="id">lowerScanLimit</span> <span class="o">..</span> <span onmouseout="hideTip(event, 'fs12', 33)" onmouseover="showTip(event, 'fs12', 33)" class="id">upperScanLimit</span><span class="pn">]</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span class="id">Chart</span><span class="pn">.</span><span class="id">withTemplate</span> <span class="id">ChartTemplates</span><span class="pn">.</span><span class="id">light</span>

<span onmouseout="hideTip(event, 'fs15', 34)" onmouseover="showTip(event, 'fs15', 34)" class="id">intensityChart</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<h3><a name="Step-2-Preselecting-the-peptides-of-interest" class="anchor" href="#Step-2-Preselecting-the-peptides-of-interest">Step 2: Preselecting the peptides of interest</a></h3>
<p>Every MS2 spectrum is accompanied by a m/z ratio reported by the instrument. Additionally, we can estimate the charge looking
at the isotopic cluster. We take the peptide "DTDILAAFR" from our previous notebook again. Our example has a measured m/z = 511.2691141 and
a charge of z=2.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs16', 36)" onmouseover="showTip(event, 'fs16', 36)" class="id">peptideMass</span> <span class="o">=</span> 
    <span class="id">Mass</span><span class="pn">.</span><span class="id">ofMZ</span> <span class="n">511.2691141</span> <span class="n">2.</span>

<span onmouseout="hideTip(event, 'fs16', 37)" onmouseover="showTip(event, 'fs16', 37)" class="id">peptideMass</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<p>From our previos notebook <em>NB02b_Digestion_and_mass_calculation.ipynb</em>, we know how to
calculate all peptide masses that we can expect to be present in <em>Chlamydomonas reinhardtii</em>.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 4</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs17', 38)" onmouseover="showTip(event, 'fs17', 38)" class="id">path2</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 39)" onmouseover="showTip(event, 'fs5', 39)" class="rt">Path</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs6', 40)" onmouseover="showTip(event, 'fs6', 40)" class="id">Combine</span><span class="pn">[|</span><span onmouseout="hideTip(event, 'fs3', 41)" onmouseover="showTip(event, 'fs3', 41)" class="id">directory</span><span class="pn">;</span><span class="s">&quot;downloads/Chlamy_JGI5_5(Cp_Mp).fasta&quot;</span><span class="pn">|]</span>
<span class="id">downloadFile</span> <span onmouseout="hideTip(event, 'fs17', 42)" onmouseover="showTip(event, 'fs17', 42)" class="id">path2</span> <span class="s">&quot;Chlamy_JGI5_5(Cp_Mp).fasta&quot;</span> <span class="s">&quot;bio-bte-06-l-7&quot;</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs18', 43)" onmouseover="showTip(event, 'fs18', 43)" class="id">peptideAndMasses</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs17', 44)" onmouseover="showTip(event, 'fs17', 44)" class="id">path2</span>
    <span class="o">|&gt;</span> <span class="id">IO</span><span class="pn">.</span><span class="id">FastA</span><span class="pn">.</span><span class="id">fromFile</span> <span class="id">BioArray</span><span class="pn">.</span><span class="id">ofAminoAcidString</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs19', 45)" onmouseover="showTip(event, 'fs19', 45)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs20', 46)" onmouseover="showTip(event, 'fs20', 46)" class="id">toArray</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs21', 47)" onmouseover="showTip(event, 'fs21', 47)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs22', 48)" onmouseover="showTip(event, 'fs22', 48)" class="id">mapi</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs23', 49)" onmouseover="showTip(event, 'fs23', 49)" class="fn">i</span> <span onmouseout="hideTip(event, 'fs24', 50)" onmouseover="showTip(event, 'fs24', 50)" class="fn">fastAItem</span> <span class="k">-&gt;</span>
        <span class="id">Digestion</span><span class="pn">.</span><span class="id">BioArray</span><span class="pn">.</span><span class="id">digest</span> <span class="id">Digestion</span><span class="pn">.</span><span class="id">Table</span><span class="pn">.</span><span class="id">Trypsin</span> <span onmouseout="hideTip(event, 'fs23', 51)" onmouseover="showTip(event, 'fs23', 51)" class="id">i</span> <span onmouseout="hideTip(event, 'fs24', 52)" onmouseover="showTip(event, 'fs24', 52)" class="id">fastAItem</span><span class="pn">.</span><span class="id">Sequence</span>
        <span class="o">|&gt;</span> <span class="id">Digestion</span><span class="pn">.</span><span class="id">BioArray</span><span class="pn">.</span><span class="id">concernMissCleavages</span> <span class="n">0</span> <span class="n">0</span>
    <span class="pn">)</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs21', 53)" onmouseover="showTip(event, 'fs21', 53)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs25', 54)" onmouseover="showTip(event, 'fs25', 54)" class="id">concat</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs21', 55)" onmouseover="showTip(event, 'fs21', 55)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs26', 56)" onmouseover="showTip(event, 'fs26', 56)" class="id">map</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs27', 57)" onmouseover="showTip(event, 'fs27', 57)" class="fn">peptide</span> <span class="k">-&gt;</span>
        <span class="c">// calculate mass for each peptide</span>
        <span onmouseout="hideTip(event, 'fs27', 58)" onmouseover="showTip(event, 'fs27', 58)" class="fn">peptide</span><span class="pn">.</span><span class="id">PepSequence</span><span class="pn">,</span> <span class="id">BioSeq</span><span class="pn">.</span><span class="id">toMonoisotopicMassWith</span> <span class="pn">(</span><span class="id">BioItem</span><span class="pn">.</span><span class="id">monoisoMass</span> <span class="id">ModificationInfo</span><span class="pn">.</span><span class="id">Table</span><span class="pn">.</span><span class="id">H2O</span><span class="pn">)</span> <span onmouseout="hideTip(event, 'fs27', 59)" onmouseover="showTip(event, 'fs27', 59)" class="id">peptide</span><span class="pn">.</span><span class="id">PepSequence</span>
    <span class="pn">)</span>

<span onmouseout="hideTip(event, 'fs18', 60)" onmouseover="showTip(event, 'fs18', 60)" class="id">peptideAndMasses</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs21', 61)" onmouseover="showTip(event, 'fs21', 61)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs28', 62)" onmouseover="showTip(event, 'fs28', 62)" class="id">head</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<p>However, we are only interest in possible amino acid sequences, that match the particular precursor peptide mass of our example spectrum
with 1020.523675 Da. Additionaly, we should also consider a small measurement error.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 5</span>

<span onmouseout="hideTip(event, 'fs18', 63)" onmouseover="showTip(event, 'fs18', 63)" class="id">peptideAndMasses</span>
<span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs21', 64)" onmouseover="showTip(event, 'fs21', 64)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs29', 65)" onmouseover="showTip(event, 'fs29', 65)" class="id">filter</span> <span class="pn">(</span><span class="k">fun</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs30', 66)" onmouseover="showTip(event, 'fs30', 66)" class="fn">sequence</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs31', 67)" onmouseover="showTip(event, 'fs31', 67)" class="fn">mass</span><span class="pn">)</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs31', 68)" onmouseover="showTip(event, 'fs31', 68)" class="fn">mass</span> <span class="o">&gt;</span> <span class="n">1020.52</span>  <span class="o">&amp;&amp;</span> <span onmouseout="hideTip(event, 'fs31', 69)" onmouseover="showTip(event, 'fs31', 69)" class="fn">mass</span> <span class="o">&lt;</span> <span class="n">1020.53</span><span class="pn">)</span>
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<p>In the previous notebook <em>NB04a_Fragmentation_for_peptide_identification.ipynb</em>,
we used functions that generate the theoretical series of b- and y-ions from the given peptide. Combined with the function
<code>Mz.SequestLike.predictOf</code> that generates theoretical spectra that fit the Sequest scoring algorithm.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 6</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs32', 70)" onmouseover="showTip(event, 'fs32', 70)" class="fn">predictFromSequence</span> <span onmouseout="hideTip(event, 'fs33', 71)" onmouseover="showTip(event, 'fs33', 71)" class="fn">peptide</span> <span class="o">=</span>
    <span class="pn">[</span>
        <span onmouseout="hideTip(event, 'fs33', 72)" onmouseover="showTip(event, 'fs33', 72)" class="fn">peptide</span>
        <span class="o">|&gt;</span> <span class="id">Mz</span><span class="pn">.</span><span class="id">Fragmentation</span><span class="pn">.</span><span class="id">Series</span><span class="pn">.</span> <span class="id">yOfBioList</span> <span class="id">BioItem</span><span class="pn">.</span><span class="id">initMonoisoMassWithMemP</span>
        <span onmouseout="hideTip(event, 'fs33', 73)" onmouseover="showTip(event, 'fs33', 73)" class="fn">peptide</span>
        <span class="o">|&gt;</span> <span class="id">Mz</span><span class="pn">.</span><span class="id">Fragmentation</span><span class="pn">.</span><span class="id">Series</span><span class="pn">.</span><span class="id">bOfBioList</span> <span class="id">BioItem</span><span class="pn">.</span><span class="id">initMonoisoMassWithMemP</span>
    <span class="pn">]</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs8', 74)" onmouseover="showTip(event, 'fs8', 74)" class="m">List</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs34', 75)" onmouseover="showTip(event, 'fs34', 75)" class="id">concat</span>
    <span class="o">|&gt;</span> <span class="id">Mz</span><span class="pn">.</span><span class="id">SequestLike</span><span class="pn">.</span><span class="id">predictOf</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs11', 76)" onmouseover="showTip(event, 'fs11', 76)" class="id">lowerScanLimit</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs12', 77)" onmouseover="showTip(event, 'fs12', 77)" class="id">upperScanLimit</span><span class="pn">)</span> <span class="n">2.</span>
</code></pre>
<h3><a name="Step-3-4-Matching-and-Scoring" class="anchor" href="#Step-3-4-Matching-and-Scoring">Step 3+4: Matching and Scoring</a></h3>
<p>In the matching step, we compare theoretical spectra of peptides within our precursor peptide mass range with our measured spectra.
We get a score which tells us how well the theoretical spectrum fits the given experimental spectrum.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">// Code-Block 7</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs35', 78)" onmouseover="showTip(event, 'fs35', 78)" class="id">sortedScores</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs18', 79)" onmouseover="showTip(event, 'fs18', 79)" class="id">peptideAndMasses</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs21', 80)" onmouseover="showTip(event, 'fs21', 80)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs29', 81)" onmouseover="showTip(event, 'fs29', 81)" class="id">filter</span> <span class="pn">(</span><span class="k">fun</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs30', 82)" onmouseover="showTip(event, 'fs30', 82)" class="fn">sequence</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs31', 83)" onmouseover="showTip(event, 'fs31', 83)" class="fn">mass</span><span class="pn">)</span> <span class="k">-&gt;</span>
        <span onmouseout="hideTip(event, 'fs31', 84)" onmouseover="showTip(event, 'fs31', 84)" class="fn">mass</span> <span class="o">&gt;</span> <span class="n">1020.52</span>  <span class="o">&amp;&amp;</span> <span onmouseout="hideTip(event, 'fs31', 85)" onmouseover="showTip(event, 'fs31', 85)" class="fn">mass</span> <span class="o">&lt;</span> <span class="n">1020.53</span>
    <span class="pn">)</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs21', 86)" onmouseover="showTip(event, 'fs21', 86)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs26', 87)" onmouseover="showTip(event, 'fs26', 87)" class="id">map</span> <span class="pn">(</span><span class="k">fun</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs30', 88)" onmouseover="showTip(event, 'fs30', 88)" class="fn">sequence</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs31', 89)" onmouseover="showTip(event, 'fs31', 89)" class="fn">mass</span><span class="pn">)</span>    <span class="k">-&gt;</span>
        <span onmouseout="hideTip(event, 'fs30', 90)" onmouseover="showTip(event, 'fs30', 90)" class="fn">sequence</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs32', 91)" onmouseover="showTip(event, 'fs32', 91)" class="fn">predictFromSequence</span> <span onmouseout="hideTip(event, 'fs30', 92)" onmouseover="showTip(event, 'fs30', 92)" class="fn">sequence</span>
    <span class="pn">)</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs21', 93)" onmouseover="showTip(event, 'fs21', 93)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs26', 94)" onmouseover="showTip(event, 'fs26', 94)" class="id">map</span> <span class="pn">(</span><span class="k">fun</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs30', 95)" onmouseover="showTip(event, 'fs30', 95)" class="fn">sequence</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs36', 96)" onmouseover="showTip(event, 'fs36', 96)" class="fn">theoSpectrum</span><span class="pn">)</span> <span class="k">-&gt;</span> 
        <span onmouseout="hideTip(event, 'fs30', 97)" onmouseover="showTip(event, 'fs30', 97)" class="fn">sequence</span><span class="pn">,</span><span class="id">BioFSharp</span><span class="pn">.</span><span class="id">Mz</span><span class="pn">.</span><span class="id">SequestLike</span><span class="pn">.</span><span class="id">scoreSingle</span> <span onmouseout="hideTip(event, 'fs36', 98)" onmouseover="showTip(event, 'fs36', 98)" class="id">theoSpectrum</span> <span onmouseout="hideTip(event, 'fs13', 99)" onmouseover="showTip(event, 'fs13', 99)" class="id">preprocessedIntesities</span>
    <span class="pn">)</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs21', 100)" onmouseover="showTip(event, 'fs21', 100)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs37', 101)" onmouseover="showTip(event, 'fs37', 101)" class="id">sortByDescending</span> <span class="pn">(</span><span class="k">fun</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs30', 102)" onmouseover="showTip(event, 'fs30', 102)" class="fn">sequence</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs38', 103)" onmouseover="showTip(event, 'fs38', 103)" class="fn">score</span><span class="pn">)</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs38', 104)" onmouseover="showTip(event, 'fs38', 104)" class="fn">score</span><span class="pn">)</span>

<span onmouseout="hideTip(event, 'fs35', 105)" onmouseover="showTip(event, 'fs35', 105)" class="id">sortedScores</span> 
</code></pre>
<table class="pre"><tr><td><pre><code>No value returned by any evaluator</code></pre></td></tr></table>
<p>Finally, we pick the sequence with the best score and are done for now. Notice however, that in a real world we would need to
relate our score to the complete data set to get an idea of the overall quality and which numerical value we could trust.</p>
<h2><a name="Questions" class="anchor" href="#Questions">Questions</a></h2>
<ol>
<li>
How does the Chart change, when you change the value of 'numberofwindows' from 10 to 20?
What does this parameter specify? (Code-Block 3)
</li>
<li>What is the rational behind the normalization of measured spectra?</li>
<li>Why are we adding the mass of water to the peptide sequence? (BioItem.monoisoMass ModificationInfo.Table.H2O) (Code-Block 4)</li>
<li>
In code block 6 you get a raw estimate on how many peptide sequences are candidates for a match, when given a certain mz.
Given that one MS run can consist of up to 120.000 ms2 spectra, how many peptide spectrum matches do you have to perform?
What does that mean for the performance of the application? Which parameters influence the size of the "search space"? (Code-Block 6)
</li>
<li>What happens when you choose different lower and upper scan limits?</li>
<li>Visualize the distribution of scores using a histogram. (Code-Block 7)</li>
</ol>

                </div>
            </div>
        </div>
        <div class="fsdocs-tip" id="fs1">namespace System</div>
<div class="fsdocs-tip" id="fs2">namespace System.IO</div>
<div class="fsdocs-tip" id="fs3">val directory : string</div>
<div class="fsdocs-tip" id="fs4">val path : string</div>
<div class="fsdocs-tip" id="fs5">type Path =
  static member ChangeExtension : path: string * extension: string -&gt; string
  static member Combine : path1: string * path2: string -&gt; string + 3 overloads
  static member EndsInDirectorySeparator : path: ReadOnlySpan&lt;char&gt; -&gt; bool + 1 overload
  static member GetDirectoryName : path: ReadOnlySpan&lt;char&gt; -&gt; ReadOnlySpan&lt;char&gt; + 1 overload
  static member GetExtension : path: ReadOnlySpan&lt;char&gt; -&gt; ReadOnlySpan&lt;char&gt; + 1 overload
  static member GetFileName : path: ReadOnlySpan&lt;char&gt; -&gt; ReadOnlySpan&lt;char&gt; + 1 overload
  static member GetFileNameWithoutExtension : path: ReadOnlySpan&lt;char&gt; -&gt; ReadOnlySpan&lt;char&gt; + 1 overload
  static member GetFullPath : path: string -&gt; string + 1 overload
  static member GetInvalidFileNameChars : unit -&gt; char []
  static member GetInvalidPathChars : unit -&gt; char []
  ...<br /><em>&lt;summary&gt;Performs operations on &lt;see cref=&quot;T:System.String&quot; /&gt; instances that contain file or directory path information. These operations are performed in a cross-platform manner.&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs6">Path.Combine([&lt;System.ParamArray&gt;] paths: string []) : string<br />Path.Combine(path1: string, path2: string) : string<br />Path.Combine(path1: string, path2: string, path3: string) : string<br />Path.Combine(path1: string, path2: string, path3: string, path4: string) : string</div>
<div class="fsdocs-tip" id="fs7">val ms2 : obj</div>
<div class="fsdocs-tip" id="fs8">Multiple items<br />module List

from Microsoft.FSharp.Collections<br /><em>&lt;summary&gt;Contains operations for working with values of type &lt;see cref=&quot;T:Microsoft.FSharp.Collections.list`1&quot; /&gt;.&lt;/summary&gt;<br />&lt;namespacedoc&gt;&lt;summary&gt;Operations for collections such as lists, arrays, sets, maps and sequences. See also 
    &lt;a href=&quot;https://docs.microsoft.com/dotnet/fsharp/language-reference/fsharp-collection-types&quot;&gt;F# Collection Types&lt;/a&gt; in the F# Language Guide.
 &lt;/summary&gt;&lt;/namespacedoc&gt;</em><br /><br />--------------------<br />type List&lt;&#39;T&gt; =
  | ( [] )
  | ( :: ) of Head: &#39;T * Tail: &#39;T list
    interface IReadOnlyList&lt;&#39;T&gt;
    interface IReadOnlyCollection&lt;&#39;T&gt;
    interface IEnumerable
    interface IEnumerable&lt;&#39;T&gt;
    member GetReverseIndex : rank:int * offset:int -&gt; int
    member GetSlice : startIndex:int option * endIndex:int option -&gt; &#39;T list
    static member Cons : head:&#39;T * tail:&#39;T list -&gt; &#39;T list
    member Head : &#39;T
    member IsEmpty : bool
    member Item : index:int -&gt; &#39;T with get
    ...<br /><em>&lt;summary&gt;The type of immutable singly-linked lists.&lt;/summary&gt;<br />&lt;remarks&gt;Use the constructors &lt;c&gt;[]&lt;/c&gt; and &lt;c&gt;::&lt;/c&gt; (infix) to create values of this type, or
 the notation &lt;c&gt;[1;2;3]&lt;/c&gt;. Use the values in the &lt;c&gt;List&lt;/c&gt; module to manipulate 
 values of this type, or pattern match against the values directly.
 &lt;/remarks&gt;<br />&lt;exclude /&gt;</em></div>
<div class="fsdocs-tip" id="fs9">val head : list:&#39;T list -&gt; &#39;T<br /><em>&lt;summary&gt;Returns the first element of the list.&lt;/summary&gt;<br />&lt;param name=&quot;list&quot;&gt;The input list.&lt;/param&gt;<br />&lt;exception cref=&quot;T:System.ArgumentException&quot;&gt;Thrown when the list is empty.&lt;/exception&gt;<br />&lt;returns&gt;The first element of the list.&lt;/returns&gt;</em></div>
<div class="fsdocs-tip" id="fs10">val ms2Chart : obj</div>
<div class="fsdocs-tip" id="fs11">val lowerScanLimit : float</div>
<div class="fsdocs-tip" id="fs12">val upperScanLimit : float</div>
<div class="fsdocs-tip" id="fs13">val preprocessedIntesities : obj</div>
<div class="fsdocs-tip" id="fs14">val pa : obj</div>
<div class="fsdocs-tip" id="fs15">val intensityChart : obj</div>
<div class="fsdocs-tip" id="fs16">val peptideMass : obj</div>
<div class="fsdocs-tip" id="fs17">val path2 : string</div>
<div class="fsdocs-tip" id="fs18">val peptideAndMasses : (obj * float) []</div>
<div class="fsdocs-tip" id="fs19">module Seq

from Microsoft.FSharp.Collections<br /><em>&lt;summary&gt;Contains operations for working with values of type &lt;see cref=&quot;T:Microsoft.FSharp.Collections.seq`1&quot; /&gt;.&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs20">val toArray : source:seq&lt;&#39;T&gt; -&gt; &#39;T []<br /><em>&lt;summary&gt;Builds an array from the given collection.&lt;/summary&gt;<br />&lt;param name=&quot;source&quot;&gt;The input sequence.&lt;/param&gt;<br />&lt;returns&gt;The result array.&lt;/returns&gt;<br />&lt;exception cref=&quot;T:System.ArgumentNullException&quot;&gt;Thrown when the input sequence is null.&lt;/exception&gt;</em></div>
<div class="fsdocs-tip" id="fs21">module Array

from Microsoft.FSharp.Collections<br /><em>&lt;summary&gt;Contains operations for working with arrays.&lt;/summary&gt;<br />&lt;remarks&gt;
  See also &lt;a href=&quot;https://docs.microsoft.com/dotnet/fsharp/language-reference/arrays&quot;&gt;F# Language Guide - Arrays&lt;/a&gt;.
 &lt;/remarks&gt;</em></div>
<div class="fsdocs-tip" id="fs22">val mapi : mapping:(int -&gt; &#39;T -&gt; &#39;U) -&gt; array:&#39;T [] -&gt; &#39;U []<br /><em>&lt;summary&gt;Builds a new array whose elements are the results of applying the given function
 to each of the elements of the array. The integer index passed to the
 function indicates the index of element being transformed.&lt;/summary&gt;<br />&lt;param name=&quot;mapping&quot;&gt;The function to transform elements and their indices.&lt;/param&gt;<br />&lt;param name=&quot;array&quot;&gt;The input array.&lt;/param&gt;<br />&lt;returns&gt;The array of transformed elements.&lt;/returns&gt;<br />&lt;exception cref=&quot;T:System.ArgumentNullException&quot;&gt;Thrown when the input array is null.&lt;/exception&gt;</em></div>
<div class="fsdocs-tip" id="fs23">val i : int</div>
<div class="fsdocs-tip" id="fs24">val fastAItem : obj</div>
<div class="fsdocs-tip" id="fs25">val concat : arrays:seq&lt;&#39;T []&gt; -&gt; &#39;T []<br /><em>&lt;summary&gt;Builds a new array that contains the elements of each of the given sequence of arrays.&lt;/summary&gt;<br />&lt;param name=&quot;arrays&quot;&gt;The input sequence of arrays.&lt;/param&gt;<br />&lt;returns&gt;The concatenation of the sequence of input arrays.&lt;/returns&gt;<br />&lt;exception cref=&quot;T:System.ArgumentNullException&quot;&gt;Thrown when the input sequence is null.&lt;/exception&gt;</em></div>
<div class="fsdocs-tip" id="fs26">val map : mapping:(&#39;T -&gt; &#39;U) -&gt; array:&#39;T [] -&gt; &#39;U []<br /><em>&lt;summary&gt;Builds a new array whose elements are the results of applying the given function
 to each of the elements of the array.&lt;/summary&gt;<br />&lt;param name=&quot;mapping&quot;&gt;The function to transform elements of the array.&lt;/param&gt;<br />&lt;param name=&quot;array&quot;&gt;The input array.&lt;/param&gt;<br />&lt;returns&gt;The array of transformed elements.&lt;/returns&gt;<br />&lt;exception cref=&quot;T:System.ArgumentNullException&quot;&gt;Thrown when the input array is null.&lt;/exception&gt;</em></div>
<div class="fsdocs-tip" id="fs27">val peptide : obj</div>
<div class="fsdocs-tip" id="fs28">val head : array:&#39;T [] -&gt; &#39;T<br /><em>&lt;summary&gt;Returns the first element of the array.&lt;/summary&gt;<br />&lt;param name=&quot;array&quot;&gt;The input array.&lt;/param&gt;<br />&lt;returns&gt;The first element of the array.&lt;/returns&gt;<br />&lt;exception cref=&quot;T:System.ArgumentNullException&quot;&gt;Thrown when the input array is null.&lt;/exception&gt;<br />&lt;exception cref=&quot;T:System.ArgumentException&quot;&gt;Thrown when the input array is empty.&lt;/exception&gt;</em></div>
<div class="fsdocs-tip" id="fs29">val filter : predicate:(&#39;T -&gt; bool) -&gt; array:&#39;T [] -&gt; &#39;T []<br /><em>&lt;summary&gt;Returns a new collection containing only the elements of the collection
 for which the given predicate returns &quot;true&quot;.&lt;/summary&gt;<br />&lt;param name=&quot;predicate&quot;&gt;The function to test the input elements.&lt;/param&gt;<br />&lt;param name=&quot;array&quot;&gt;The input array.&lt;/param&gt;<br />&lt;returns&gt;An array containing the elements for which the given predicate returns true.&lt;/returns&gt;<br />&lt;exception cref=&quot;T:System.ArgumentNullException&quot;&gt;Thrown when the input array is null.&lt;/exception&gt;</em></div>
<div class="fsdocs-tip" id="fs30">val sequence : obj</div>
<div class="fsdocs-tip" id="fs31">val mass : float</div>
<div class="fsdocs-tip" id="fs32">val predictFromSequence : peptide:&#39;a -&gt; &#39;b</div>
<div class="fsdocs-tip" id="fs33">val peptide : &#39;a</div>
<div class="fsdocs-tip" id="fs34">val concat : lists:seq&lt;&#39;T list&gt; -&gt; &#39;T list<br /><em>&lt;summary&gt;Returns a new list that contains the elements of each the lists in order.&lt;/summary&gt;<br />&lt;param name=&quot;lists&quot;&gt;The input sequence of lists.&lt;/param&gt;<br />&lt;returns&gt;The resulting concatenated list.&lt;/returns&gt;</em></div>
<div class="fsdocs-tip" id="fs35">val sortedScores : (obj * System.IComparable) []</div>
<div class="fsdocs-tip" id="fs36">val theoSpectrum : obj</div>
<div class="fsdocs-tip" id="fs37">val sortByDescending : projection:(&#39;T -&gt; &#39;Key) -&gt; array:&#39;T [] -&gt; &#39;T [] (requires comparison)<br /><em>&lt;summary&gt;Sorts the elements of an array, in descending order, using the given projection for the keys and returning a new array. 
 Elements are compared using &lt;see cref=&quot;M:Microsoft.FSharp.Core.Operators.compare&quot; /&gt;.&lt;/summary&gt;<br />&lt;remarks&gt;This is not a stable sort, i.e. the original order of equal elements is not necessarily preserved. 
 For a stable sort, consider using &lt;see cref=&quot;M:Microsoft.FSharp.Collections.SeqModule.Sort&quot; /&gt;.&lt;/remarks&gt;<br />&lt;param name=&quot;projection&quot;&gt;The function to transform array elements into the type that is compared.&lt;/param&gt;<br />&lt;param name=&quot;array&quot;&gt;The input array.&lt;/param&gt;<br />&lt;returns&gt;The sorted array.&lt;/returns&gt;</em></div>
<div class="fsdocs-tip" id="fs38">val score : System.IComparable</div>

    </div>

    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <script type="text/javascript">var fsdocs_search_baseurl = '/BIO-BTE-06-L-7/'</script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/lunr.js/2.3.8/lunr.min.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.min.js"></script>
    <script type="text/javascript" src="/BIO-BTE-06-L-7/content/fsdocs-search.js"></script>
    <!-- END SEARCH BOX: this adds support for the search box -->
</body>

</html>