
        {
            "cells": [
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 1, "outputs": [], 
           "source": ["#r \"nuget: FSharp.Stats, 0.4.3\"\n",
"#r \"nuget: BioFSharp, 2.0.0-beta5\"\n",
"#r \"nuget: BioFSharp.IO, 2.0.0-beta5\"\n",
"#r \"nuget: Plotly.NET, 2.0.0-preview.16\"\n",
"#r \"nuget: BIO-BTE-06-L-7_Aux, 0.0.9\"\n",
"#r \"nuget: Deedle, 2.5.0\"\n",
"#r \"nuget: ISADotNet, 0.4.0-preview.4\"\n",
"#r \"nuget: ISADotNet.XLSX, 0.4.0-preview.4\"\n",
"#r \"nuget: ISADotNet.IO, 0.0.2\"\n",
"\n",
"#r \"nuget: Plotly.NET.Interactive, 2.0.0-preview.16\"\n",
"\n",
"open System.IO\n",
"open ISADotNet\n",
"open ISADotNet.API\n",
"open Deedle\n",
"open BioFSharp\n",
"open FSharpAux\n",
"open FSharp.Stats\n",
"open Plotly.NET\n",
"open arcIO.NET\n",
"open BIO_BTE_06_L_7_Aux.Deedle_Aux\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["# NB08a Data Access and Quality Control (for BN-PAGE results)\n",
"\n",
"[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/CSBiology/BIO-BTE-06-L-7/gh-pages?filepath=NB08a_Data_Access_And_Quality_Control_BN.ipynb)\n",
"\n",
"[Download Notebook](https://github.com/CSBiology/BIO-BTE-06-L-7/releases/download/NB08a/NB08a_Data_Access_And_Quality_Control_BN.ipynb)\n",
"\n",
"With this notebook, we want to converge the threads of computational and experimental proteomics by analyzing the data measured by you during the practical course.\n",
"Behind the scenes there was already a lot going on! While you were going through the hands-on tutorials addressing single steps of the computation proteomics pipeline, we executed\n",
"a pipeline combining all the steps. In this tutorial, we will provide you with the output of the pipeline, your sample description and a lot of code, which should help you explore the \n",
"data set generated by you! Since the experiment is of a non neglectable depth - so is the code needed to analyze it. Do not hesitate if you do not understand every step right away, this is no\n",
"trivial task and nontrivial are the means to achieve it! We included some questions along the way. The questions aim to check your understanding, but they can also serve as food for thought.\n",
"\n",
"For the explorative data analysis, we are using [Deedle](http://bluemountaincapital.github.io/Deedle/tutorial.html).\n",
"Deedle is an easy to use library for data and time series manipulation and for scientific programming. \n",
"It supports working with structured data frames, ordered and unordered data, as well as time series. Deedle is designed to work well for exploratory programming using F#.\n",
"\n",
"## I. Reading the sample description\n",
"\n",
"Before we analyze our data, we will download and read the sample description provided by the experimentalist.\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 2, "outputs": [], 
           "source": ["let path2 = @\"..\\assays\\VP21_WC\\isa.assay.xlsx\"\n",
"\n",
"let _,_,_,myAssayFile = XLSX.AssayFile.Assay.fromFile path2\n",
"let inOutMap = ISADotNet.createInOutMap myAssayFile\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["Next, we will prepare functions to look up parameters, which might be needed for further calculations. \n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 3, "outputs": [], 
           "source": ["let normalizeFileName (f : string) = if Path.HasExtension f then f else Path.ChangeExtension(f, \"wiff\")\n",
"\n",
"//        \n",
"let getStrain (fileName : string) =\n",
"    let fN = fileName |\u003e normalizeFileName\n",
"    ISADotNet.tryGetCharacteristic inOutMap \"Cultivation\" \"strain\" fN myAssayFile\n",
"    |\u003e Option.defaultValue \"\"\n",
"\n",
"//\n",
"let getExpressionLevel (fileName : string) =\n",
"    let fN = fileName |\u003e normalizeFileName \n",
"    ISADotNet.tryGetCharacteristic inOutMap \"Cultivation\" \"gene expression\" fN myAssayFile \n",
"    |\u003e Option.defaultValue \"Wt-Like\"\n",
"\n",
"//\n",
"let get15N_PS_Amount (fileName : string) =\n",
"    let fN = fileName |\u003e normalizeFileName\n",
"    ISADotNet.tryGetParameter inOutMap \"Protein extraction\" \"15N Photosynthesis QconCAT mass #4\" fN myAssayFile\n",
"    |\u003e Option.defaultValue \"\"\n",
"    |\u003e String.split \u0027 \u0027\n",
"    |\u003e Array.head\n",
"    |\u003e float \n",
"\n",
"//\n",
"let getGroupID (fileName : string) =\n",
"    let fN = fileName |\u003e normalizeFileName\n",
"    ISADotNet.tryGetParameter inOutMap \"Protein extraction\" \"Group name\" fN myAssayFile\n",
"    |\u003e Option.defaultValue \"\"\n",
"    |\u003e int\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["A quick execution to test the retrieval of data from the isa sample table:\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 4, "outputs": [], 
           "source": ["getStrain \"20210312BN2_U1.wiff\"\n",
"getExpressionLevel \"20210312BN2_U1.wiff\"\n",
"get15N_PS_Amount \"20210312BN2_U1.wiff\"\n",
"getGroupID \"20210312BN2_U1.wiff\"\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["Now that we have the sample sheet, all that is missing is the data to be analyzed:\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 5, "outputs": [], 
           "source": ["let path = @\"..assays\\VP21_WC\\dataset\\WCAnnotated.txt\"\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["## II. Raw data access using Deedle:\n",
"As teasered in the primer, we want to work with our tabular data using Deedle. Luckily, Deedle does not only deliver data frame and series\n",
"manipulation, but also allows us to quickly read the recently downloaded data into the memory:\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 6, "outputs": [], 
           "source": ["let rawData = Frame.ReadCsv(path, separators = \"\\t\")\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["To visualize the data, we can call the \"formatAsTable\" function. The preview of visual studio code does not allow\n",
"for the charts to be scrollable, so we pipe the output into \"Chart.Show\", to visualize the data in your browser.\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": null, "outputs": [
          {
           "data": {
            "text/plain": [""]
        },
           "execution_count": null,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["rawData\n",
"|\u003e Frame.take 10\n",
"|\u003e formatAsTable 1500.\n",
"\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["Looking at the raw data, we can see that each row contains a different quantification of a peptide ion, with the columns containing \n",
"a single ion feature each, such as peptide ion charge, sequence or a quantification value reported for a file (e.g. light, heavy or ratio).\n",
"Since the columns ProteinGroup, StringSequence, PepSequenceID and Charge uniquely identify a row, we can use these to index the rows.\n",
"For this, we use a language feature called [\"anonymous record type\"](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/anonymous-records). \n",
"Here we create a tuple like structure, with the additional feature that each element of the tuple is named (e.g.: Proteingroup).\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 8, "outputs": [], 
           "source": ["let indexedData =\n",
"    rawData\n",
"    // StringSequence is the peptide sequence\n",
"    |\u003e Frame.indexRowsUsing (fun os -\u003e \n",
"        {|\n",
"            ProteinGroup    = os.GetAs\u003cstring\u003e(\"ProteinGroup\"); \n",
"            Synonyms        = os.GetAs\u003cstring\u003e(\"Synonym\")\n",
"            StringSequence  = os.GetAs\u003cstring\u003e(\"StringSequence\");\n",
"            PepSequenceID   = os.GetAs\u003cint\u003e(\"PepSequenceID\");\n",
"            Charge          = os.GetAs\u003cint\u003e(\"Charge\")\n",
"        |}\n",
"    )\n",
"    \n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": null, "outputs": [
          {
           "data": {
            "text/plain": [""]
        },
           "execution_count": null,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// The effect of our frame manipulation can be observed:\n",
"indexedData\n",
"|\u003e Frame.take 10\n",
"|\u003e formatAsTable 1500.\n",
"\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["## III. Augmenting and filtering the data frame \n",
"The data frame already contains all information needed to perform the analysis, but it could still benefit from \n",
"some quality-of-life upgrades. Say, we want to encode the specific qConcat protein as a separate feature:\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 10, "outputs": [], 
           "source": ["// Why does it make sense to model Qprots using this type, why do we not simply use a string?\n",
"\n",
"let finalRaw = \n",
"    indexedData\n",
"    // What does this line filter for? Why does this make sense for our analysis?\n",
"    // How many peptide ions did the filter remove? \n",
"    |\u003e Frame.filterRows (fun k s -\u003e k.Synonyms \u003c\u003e \"\" \u0026\u0026 k.ProteinGroup |\u003e String.contains \"QProt_newPS\")\n",
"\n",
"finalRaw\n",
"|\u003e Frame.filterRows (fun x s -\u003e x.StringSequence = \"DTDILAAFR\")\n",
"|\u003e Frame.getCol \"20210312BN2_U1.Ratio\"\n",
"|\u003e Series.mapValues string\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": null, "outputs": [
          {
           "data": {
            "text/plain": [""]
        },
           "execution_count": null,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["finalRaw\n",
"|\u003e Frame.take 10\n",
"|\u003e formatAsTable 1500.\n",
"\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["## IV. Global quality control.\n",
"\n",
"With our data frame prepared, we want to see on a global scale if our experiment worked.\n",
"We plot the overall mean of the 14N and 15N quantifications and observe if we can recover our dilution series (15N),\n",
"while keeping the analyte to be quantified at a constant level (14N).\n",
"\n",
"Since it comes in handy to simplify the data frame, we will only keep columns that contain a specific identifier, \n",
"such as, \"Ratio\", \"Light\" or \"Heavy\". \n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 12, "outputs": [], 
           "source": ["let sliceQuantColumns quantColID frame = \n",
"    frame\n",
"    |\u003e Frame.filterCols (fun ck os -\u003e ck |\u003e String.contains (\".\" + quantColID))\n",
"    |\u003e Frame.mapColKeys (fun ck -\u003e ck.Split(\u0027.\u0027) |\u003e Array.item 0)\n",
"\n",
"// How did the data frame change, how did the column headers change?\n",
"let ratios = sliceQuantColumns \"Ratio\" finalRaw\n",
"let light = sliceQuantColumns \"Quant_Light\" finalRaw\n",
"let heavy = sliceQuantColumns \"Quant_Heavy\" finalRaw\n",
"\n",
"ratios\n",
"|\u003e Frame.filterRows (fun x s -\u003e x.StringSequence = \"DTDILAAFR\")\n",
"|\u003e Frame.getCol \"20210312BN2_U1\"\n",
"|\u003e Series.mapValues string\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": null, "outputs": [
          {
           "data": {
            "text/plain": [""]
        },
           "execution_count": null,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["ratios\n",
"|\u003e Frame.take 10\n",
"|\u003e formatAsTable 1500.\n",
"\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["A nice tool to explore and compare distributions of different populations is the representation as a boxplot!\n",
"To use this tool, we will define a function which creates a boxplot for every column (file) of our data set:\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 14, "outputs": [], 
           "source": ["let createBoxPlot f = \n",
"    f\n",
"    |\u003e Frame.getNumericCols\n",
"    |\u003e Series.map (fun k s -\u003e \n",
"         let x,y =\n",
"            s\n",
"            |\u003e Series.values \n",
"            |\u003e Seq.map (fun values -\u003e k,values)\n",
"            |\u003e Seq.unzip\n",
"         Chart.BoxPlot(x, y, Orientation = StyleParam.Orientation.Vertical)\n",
"         )\n",
"    |\u003e Series.values\n",
"    |\u003e Chart.combine\n",
"    |\u003e Chart.withYAxisStyle \"Ion intensity\"\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["The function applied to the N14 values: \n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 15, "outputs": [
          {
           "data": {
            "text/plain": ["No value returned by any evaluator"]
        },
           "execution_count": 15,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// How is the data distributed?\n",
"light\n",
"|\u003e createBoxPlot\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["The function applied to the N15 values:\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 17, "outputs": [
          {
           "data": {
            "text/plain": ["No value returned by any evaluator"]
        },
           "execution_count": 17,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// Can you recover the dilution series?\n",
"heavy\n",
"|\u003e createBoxPlot\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["The following function performs a normalization which accounts for a specific effect. Can you \n",
"determine what the function accounts for?\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 19, "outputs": [], 
           "source": ["let normalizePeptides f =\n",
"    f\n",
"    |\u003e Frame.transpose\n",
"    |\u003e Frame.getNumericCols\n",
"    |\u003e Series.mapValues (fun s -\u003e \n",
"        let m = Stats.median s\n",
"        s / m \n",
"        )\n",
"    |\u003e Frame.ofColumns\n",
"    |\u003e Frame.transpose\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 20, "outputs": [
          {
           "data": {
            "text/plain": ["No value returned by any evaluator"]
        },
           "execution_count": 20,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["// How does the distribution of the date change, when the normalization is applied? \n",
"light\n",
"|\u003e normalizePeptides\n",
"|\u003e createBoxPlot\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 22, "outputs": [
          {
           "data": {
            "text/plain": ["No value returned by any evaluator"]
        },
           "execution_count": 22,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["heavy\n",
"|\u003e normalizePeptides\n",
"|\u003e createBoxPlot \n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["Finally we have a look at the ratios. \n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 24, "outputs": [], 
           "source": ["// Does it make sense to normalize the ratios the same way?\n",
"ratios\n",
"|\u003e createBoxPlot \n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["## Local quality control\n",
"\n",
"Now that we know on a global scale how our experiment worked, it might be time to have a look at the details.\n",
"First, we want to write a function that allows us to plot all peptides of a protein vs. the dilution used. This way we can identify peptides that\n",
"we want to use and those that seem to be prone to error and should thus be discarded. \n",
"To keep things simple, we apply a filter step at the beginning, which only keeps peptides belonging to one protein and samples measured by one group\n",
"in the data frame. What are the sources of error? Which peptides do you think should be discarded and why? Which proteins need to be analyzed with extra care?\n",
"Hint: you can hover over the data points to get insight into the file name and gene expression pattern of the corresponding strain.\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 25, "outputs": [], 
           "source": ["// With this type we create an alias to our row key, this allows us to write functions, which operate on data frames such as \u0027plotPeptidesOf\u0027,\u0027discardPeptideIonInFile\u0027 and \u0027discardPeptideIon\u0027\n",
"type PeptideIon = \n",
"    {|\n",
"        ProteinGroup    : string  \n",
"        Synonyms        : string\n",
"        StringSequence  : string\n",
"        PepSequenceID   : int\n",
"        Charge          : int\n",
"    |}\n",
"\n",
"\n",
"// Given a frame, a prot-ID and a group-ID this function creates an xy plot for every peptide ion belonging to the protein/proteingroup.\n",
"// The parameter \u0027prot\u0027 can either be given a valid Cre-ID or a synonym.\n",
"// What is the unit of the x-Axis? How is the ratio calculated? \n",
"let plotPeptidesOf (ratios : Frame\u003cPeptideIon,string\u003e) (prot : string) (groupID : int) = \n",
"    try \n",
"        ratios\n",
"        |\u003e Frame.filterRows (fun k s -\u003e k.Synonyms.Contains prot || k.ProteinGroup.Contains prot)\n",
"        |\u003e Frame.filterCols (fun k s -\u003e getGroupID k = groupID)   \n",
"        |\u003e Frame.transpose\n",
"        |\u003e Frame.getNumericCols\n",
"        |\u003e Series.map (fun pep (values) -\u003e \n",
"            let fileNames,ratios,fileLabel =\n",
"                values\n",
"                |\u003e Series.map (fun fileName (ratio) -\u003e \n",
"                        let expressionLevel = getExpressionLevel fileName\n",
"                        fileName, ratio, (sprintf \"%s %s\" fileName expressionLevel)         \n",
"                    )\n",
"                |\u003e Series.values\n",
"                |\u003e Seq.unzip3\n",
"            Chart.Point(fileNames, ratios, MultiText = fileLabel)\n",
"            |\u003e Chart.withTraceName (sprintf \"S:%s_C:%i\" pep.StringSequence pep.Charge)\n",
"            |\u003e Chart.withXAxisStyle(\"File name\")\n",
"            |\u003e Chart.withYAxisStyle(\"Ratio\")\n",
"        )\n",
"        |\u003e Series.values\n",
"        |\u003e Chart.combine\n",
"        |\u003e Chart.withTitle \"BN-PAGE extraction\"\n",
"        |\u003e Chart.withSize (600.,700.)\n",
"        |\u003e Chart.withMarginSize (Bottom = 175.)\n",
"    with :? System.ArgumentException -\u003e failwith \"ERROR: Input protein was not found.\"\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["First we get an overview of available protein ids.\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 26, "outputs": [], 
           "source": ["ratios.RowKeys\n",
"|\u003e Array.ofSeq \n",
"|\u003e Array.map (fun k -\u003e k.Synonyms)\n",
"|\u003e Array.distinct\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["Then we can start to visualizes our results:\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": null, "outputs": [
          {
           "data": {
            "text/plain": ["No value returned by any evaluator"]
        },
           "execution_count": null,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["plotPeptidesOf ratios \"rbcL\" 1\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": null, "outputs": [
          {
           "data": {
            "text/plain": ["No value returned by any evaluator"]
        },
           "execution_count": null,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["plotPeptidesOf ratios \"RBCS1;RBCS2\" 2\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": null, "outputs": [
          {
           "data": {
            "text/plain": ["No value returned by any evaluator"]
        },
           "execution_count": null,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["plotPeptidesOf ratios \"SEBP1\" 4\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": null, "outputs": [
          {
           "data": {
            "text/plain": ["No value returned by any evaluator"]
        },
           "execution_count": null,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["plotPeptidesOf ratios \"TRK1\" 1 \n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": null, "outputs": [
          {
           "data": {
            "text/plain": ["No value returned by any evaluator"]
        },
           "execution_count": null,
           "metadata": {},
           "output_type": "execute_result"
          }], 
           "source": ["plotPeptidesOf ratios \"PRK1\" 1\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 32, "outputs": [], 
           "source": ["// Describe what happened with the last 3 plots.\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["Since we want to save our result and use it for the next notebook, where we will have a look at the isotopic labeling efficiency and finally calculate absolute protein amounts, we \n",
"need to save the filtered frame. Additionally, we want to keep information which was dropped along the way: isotopic patterns. In order to do so, we perform a join operation, which keeps only those rows \n",
"present in both files:\n",
"\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 33, "outputs": [], 
           "source": ["//  Are there redundant columns in the result frame? Why?\n",
"let frameComplete = \n",
"    Frame.join JoinKind.Inner finalRaw ratios\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["With the plots at hand, we can use the following functions to manipulate the data frame and discard peptides and/or whole files which we do not want to use for \n",
"an absolute protein quantification e.g.:\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 34, "outputs": [], 
           "source": ["let discardPeptideIonInFile stringsequence charge filename (ratios:Frame\u003cPeptideIon,string\u003e) = \n",
"    ratios\n",
"    |\u003e Frame.map (fun r c value -\u003e \n",
"        let cFileName = String.split \u0027.\u0027 c |\u003e Array.head\n",
"        if r.StringSequence = stringsequence \u0026\u0026 r.Charge = charge \u0026\u0026 cFileName = filename then nan else value\n",
"    )\n",
"\n",
"let discardPeptideIon stringsequence charge (ratios:Frame\u003cPeptideIon,string\u003e) = \n",
"    ratios\n",
"    |\u003e Frame.filterRows (fun r s -\u003e (r.StringSequence = stringsequence \u0026\u0026 r.Charge = charge) |\u003e not)\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["These functions can then be used to create an updated version of the frame, containing only the values we want to use for quantification e.g.:\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 35, "outputs": [], 
           "source": ["let filtered = \n",
"    frameComplete\n",
"    |\u003e discardPeptideIonInFile \"IYSFNEGNYGLWDDSVK\" 3 \"Gr2rbcL2_5\" \n",
"    |\u003e discardPeptideIon \"IYSFNEGNYGLWDDSVK\" 2\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["Of course, it is possible to apply very strict additional filters onto the previously filtered frame:\n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 36, "outputs": [], 
           "source": ["let ratiosFiltered = \n",
"    filtered\n",
"    |\u003e Frame.filterCols (fun k s -\u003e \n",
"        let kFileName = String.split \u0027.\u0027 k |\u003e Array.head\n",
"        try\n",
"            get15N_PS_Amount kFileName \u003e 0.1 \n",
"        with\n",
"        | _ -\u003e true\n",
"    )\n"]
          }
,
          {
           "cell_type": "markdown",
           "metadata": {},
           
           "source": ["This frame can then be saved locally using the following pattern:    \n",
"\n"]
          }
,
          {
           "cell_type": "code",
           "metadata": {},
            "execution_count": 37, "outputs": [], 
           "source": ["let frameToSave = \n",
"    ratiosFiltered\n",
"    |\u003e Frame.indexRowsOrdinally\n",
"\n",
"// frameToSave.SaveCsv(@\"C:\\YourPath\\testOut.txt\", separator = \u0027\\t\u0027, includeRowKeys = false)\n",
"frameToSave.SaveCsv(@\"C:\\YourPath\\testOut.txt\", separator = \u0027\\t\u0027, includeRowKeys = false)\n"]
          }],
            "metadata": {
            "kernelspec": {"display_name": ".NET (F#)", "language": "F#", "name": ".net-fsharp"},
            "langauge_info": {
        "file_extension": ".fs",
        "mimetype": "text/x-fsharp",
        "name": "C#",
        "pygments_lexer": "fsharp",
        "version": "4.5"
        }
        },
            "nbformat": 4,
            "nbformat_minor": 1
        }
        

